# Input Files used for testing

All the files in this directory were either randomly generated or sourced from publicly available genomic data. Many of the files have been downsampled to make them a manageable size for storing in a Git repository (using `git-lfs`) and to enable quick test execution in our CI. For the most part, these files are not representative of any meaningful scientific data and should be considered as mocked data only suitable for the purpose of generic testing.

The following list is sorted alphabetically:

1scattered.interval_list
: TODO

201533520001_R01C01.beta_swan_norm_unfiltered.genomic.csv
: TODO

201533520001_R03C01_Grn.idat
: TODO

201533520001_R03C01_Red.idat
: TODO

201533520001_R03C01.beta_swan_norm_unfiltered.genomic.csv
: TODO

Aligned.sortedByCoord.chr9_chr22.bam
: Coordinate sorted BAM file with alignments on `chr9` and `chr22`. Generated with STAR and parameters suitable for `arriba` variant calling. Has a single read group with ID `rg1`.

Aligned.sortedByCoord.chr9_chr22.bam.bai
: BAM index for `Aligned.sortedByCoord.chr9_chr22.bam`.

bwa_db.tar.gz
: TODO

chr1.interval_list
: TODO

combined_beta.csv
: TODO

filtered_beta.csv
: TODO

fusions.BCR-ABL1.tsv
: A single called fusion from `arriba`.

gencode.v31.chr9_chr22.gtf.gz
: Gencode v31 release GTF (or GFFv2) file, subsetted to only contain annotations for `chr9` and `chr22`. Gzipped.

gencode.v31.chrY_chrM.gene.bed
: A 3 column BED file containing (unlabelled) intervals of `gene` features from `gencode.v31.chrY_chrM.gtf.gz`. Generated by the `util.make_coverage_regions_bed` WDL task.

gencode.v31.chrY_chrM.genelengths.txt
: Unheadered 2 column TSV file containing genelengths with `gene_name` labels from `gencode.v31.chrY_chrM.gtf.gz`. Generated by the `util.calc_gene_lengths` WDL task.

gencode.v31.chrY_chrM.gtf.gz
: Gencode v31 release GTF (or GFFv2) file, subsetted to only contain annotations for `chrY` and `chrM`. Gzipped.

GRCh38.chr1_chr19.dict
: TODO

GRCh38.chr1_chr19.fa
: TODO

GRCh38.chr1_chr19.fa.fai
: TODO

GRCh38.chr9_chr22.fa.gz
: TODO

GRCh38.chrY_chrM.dict
: TODO

GRCh38.chrY_chrM.fa
: TODO

GRCh38.chrY_chrM.fa.fai
: TODO

GRCh38.tar.gz
: TODO

Homo_sapiens_assembly38.dbsnp138.top5000.vcf
: TODO

Homo_sapiens_assembly38.dbsnp138.top5000.vcf.idx
: TODO

kraken2_C_elegans_library.tar.gz
: A Kraken "custom library" created from the C. elegans genome. This is an intermediate file that is not meant to be human readable. Generated by the `kraken2.create_library_from_fastas` WDL task.

kraken2_db.mini.tar.gz
: A mini Kraken database. Suitable for testing, but should **not** be used in production.

kraken2_taxonomy.tar.gz
: The NCBI taxonomy in a format ingestible by Kraken2 during the DB build step.

Mills_and_1000G_gold_standard.indels.hg38.vcf.gz
: TODO

Mills_and_1000G_gold_standard.indels.hg38.vcf.gz.tbi
: TODO

star_db.chrY_chrM.tar.gz
: STAR database containing only the GRCh38 `chrY` and `chrM` contigs.

test_R1_copy.fq.gz
: An exact copy of `test_R1.gz.gz`.

test_R1.fq.gz
: 10,000 reads in FASTQ format. Can be used by itself to represent a Single-End sample or used with `test_R2.fq.gz` to represent a Paired-End sample.

test_R2_copy.fq.gz
: An exact copy of `test_R2.gz.gz`.

test_R2.fq.gz
: 10,000 reads in FASTQ format. Can be used by itself to represent a Single-End sample or used with `test_R1.fq.gz` to represent a Paired-End sample.

test_rnaseq_variant.bam
: Subset of reads spanning the interval `chr1:16015740-16058682` from a fully aligned `HG002_GM24385_replicate1_totalRNAseq` sample. Generated with STAR and parameters suitable for `arriba` variant calling. One read group with ID `GM24385_replicate1` and 2,534 primary reads and 48 secondary reads (2582 reads in total).

test_rnaseq_variant.bam.bai
: BAM index corresponding to `test_rnaseq_variant.bam`.

test_rnaseq_variant.recal.txt
: TODO

test.bam
: Coordinate sorted BAM with 2 read groups (IDs of `1` and `2`) and 633 total reads mapped to `chr1` and `chr19`.

test.bam.bai
: BAM index corresponding to `test.bam`.

test.bwa_aln_pe.chrY_chrM.bam
: Coordinate sorted BAM with 1 read group (`ID:test`). Aligned using BWA `aln` (using the `bwa.bwa_aln_pe` WDL task). 20,000 Paired-End reads mapped to GRCh38 `chrY` and `chrM` from `test_R1.fq.gz` and `test_R2.fq.gz`.

test.bwa_aln_pe.chrY_chrM.bam.bai
: BAM index corresponding to `test.bwa_aln_pe.chrY_chrM.bam`.

test.bwa_aln_pe.chrY_chrM.feature-counts.txt
: HTSeq Count output for `test.bwa_aln_pe.chrY_chrM.bam`.

test.bwa_aln_pe.chrY_chrM.readlength.txt
: `ngsderive readlen` output file for `test.bwa_aln_pe.chrY_chrM.bam`. Used as input for `multiqc` test.

test.bwa_aln_pe.chrY_chrM.TPM.txt
: Transcripts Per Million file generated by `htseq.calc_tpm` using `test.bwa_aln_pe.chrY_chrM.feature-counts.txt` as input.

test.extra_RG.bam
: A duplicate of `test.bam` with an added `@RG` entry with the ID `no_match`. There are no reads corresponding to the `no_match` RG entry.

test.fa
: TODO

test.SE.2_RGs.Aligned.out.bam
: Single-End BAM generated with STAR. There are two read groups, `rg1` and `rg2` which correspond to `test_R1.fq.gz` and `test_R1_copy.fq.gz`. Contains 20,000 reads mapped to the contigs `chrY` and `chrM`.

test.tar.gz
: A tarball with contents of `test_file_a` and `test_file_b` files.

test.unaccounted_read.bam
: A duplicate of `test.bam` with the last read in the file manually edited to have an `RG` tag of `no_match`. This RG tag is not present in the header.

test1.vcf.gz
: TODO

test1.vcf.gz.tbi
: TODO

test2.bam
: TODO

test2.vcf.gz
: TODO

test2.vcf.gz.tbi
: TODO

umap.csv
: TODO

wgs_calling_regions.hg38.interval_list
: TODO
