<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="author" content="St. Jude Cloud" />
      <link rel="shortcut icon" href="../../img/favicon.ico" />
    <title>Scrnaseq standard - St. Jude Cloud Bioinformatics Workflows</title>
    <link rel="stylesheet" href="../../css/theme.css" />
    <link rel="stylesheet" href="../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Scrnaseq standard";
        var mkdocs_page_input_path = "workflows/scrnaseq-standard.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../.." class="icon icon-home"> St. Jude Cloud Bioinformatics Workflows
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../build_for_dnanexus/">Building WDL workflows for DNAnexus</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Tasks</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/bwa/">Bwa</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/cellranger/">Cellranger</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/deeptools/">Deeptools</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/estimate/">Estimate</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/fastqc/">Fastqc</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/fq/">Fq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/htseq/">Htseq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/kraken2/">Kraken2</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/librarian/">librarian</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/md5sum/">Md5sum</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/mosdepth/">Mosdepth</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/multiqc/">Multiqc</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/ngsderive/">Ngsderive</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/picard/">Picard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/qualimap/">Qualimap</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/samtools/">Samtools</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/star/">Star</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/util/">Utilities</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Workflows</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../10x-bam-to-fastqs/">Cell Ranger BAM to FASTQs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ESTIMATE/">ESTIMATE</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../bam-to-fastqs/">Bam to fastqs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../bwa-db-build/">Bwa db build</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../dnaseq-standard/">Dnaseq standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../flag_filter/">FlagFilter</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../make-qc-reference/">Make qc reference</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../markdups-post/">MarkDuplicates Post</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../quality-check-standard/">Quality check standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../samtools-merge/">Samtools merge</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="./">Scrnaseq standard</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#mit-license">MIT License</a>
    </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../star-db-build/">Star db build</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../..">St. Jude Cloud Bioinformatics Workflows</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../.." class="icon icon-home" aria-label="Docs"></a></li>
          <li class="breadcrumb-item">Workflows</li>
      <li class="breadcrumb-item active">Scrnaseq standard</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/stjudecloud/workflows/edit/master/docs/workflows/scrnaseq-standard.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <p>scRNA-Seq Standard</p>
<p>This WDL workflow runs the Cell Ranger scRNA-Seq alignment workflow for St. Jude Cloud.</p>
<p>The workflow takes an input BAM file and splits it into FASTQ files for each read in the pair.
The read pairs are then passed through Cell Ranger to generate a BAM file and perform
quantification. Strandedness is inferred using ngsderive.
File validation is performed at several steps, including immediately preceeding output.</p>
<h2 id="licensing">LICENSING</h2>
<h4 id="mit-license">MIT License</h4>
<p>Copyright 2022-Present St. Jude Children's Research Hospital</p>
<p>Permission is hereby granted, free of charge, to any person obtaining a copy of this
software and associated documentation files (the "Software"), to deal in the Software
without restriction, including without limitation the rights to use, copy, modify, merge,
publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons
to whom the Software is furnished to do so, subject to the following conditions:</p>
<p>The above copyright notice and this permission notice shall be included in all copies or
substantial portions of the Software.</p>
<p>THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.</p>
<h2 id="scrnaseq_standard">scrnaseq_standard</h2>
<h3 id="inputs">Inputs</h3>
<h4 id="required">Required</h4>
<ul>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file to quality check</li>
<li><code>gtf</code> (File, <strong>required</strong>): Gzipped GTF feature file</li>
<li><code>transcriptome_tar_gz</code> (File, <strong>required</strong>): Database of reference files for Cell Ranger. Can be downloaded from 10x Genomics.</li>
<li><code>compute_checksum._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>count._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>strandedness._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>subsample._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>validate_bam._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>validate_input_bam._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>cell_ranger_bam_to_fastqs.bamtofastq._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>cell_ranger_bam_to_fastqs.fqlint._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>cell_ranger_bam_to_fastqs.quickcheck._runtime</code> (Any, <strong>required</strong>)</li>
</ul>
<h4 id="optional">Optional</h4>
<ul>
<li><code>validate_bam.reference_fasta</code> (File?)</li>
<li><code>validate_input_bam.reference_fasta</code> (File?)</li>
</ul>
<h4 id="defaults">Defaults</h4>
<ul>
<li><code>prefix</code> (String, default=basename(bam,".bam")): Prefix for output files</li>
<li><code>subsample_n_reads</code> (Int, default=-1): Only process a random sampling of <code>n</code> reads. &lt;=<code>0</code> for processing entire input BAM.</li>
<li><code>use_all_cores</code> (Boolean, default=false): Use all cores for multi-core steps?</li>
<li><code>validate_input</code> (Boolean, default=true): Ensure input BAM is well-formed before beginning harmonization?</li>
<li><code>cell_ranger_bam_to_fastqs.cellranger11</code> (Boolean, default=false)</li>
<li><code>cell_ranger_bam_to_fastqs.gemcode</code> (Boolean, default=false)</li>
<li><code>cell_ranger_bam_to_fastqs.longranger20</code> (Boolean, default=false)</li>
<li><code>compute_checksum.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>count.max_retries</code> (Int, default=1)</li>
<li><code>count.memory_gb</code> (Int, default=16)</li>
<li><code>count.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>count.ncpu</code> (Int, default=1)</li>
<li><code>strandedness.min_mapq</code> (Int, default=30)</li>
<li><code>strandedness.min_reads_per_gene</code> (Int, default=10)</li>
<li><code>strandedness.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>strandedness.num_genes</code> (Int, default=1000)</li>
<li><code>strandedness.outfile_name</code> (String, default=basename(bam,".bam") + ".strandedness.tsv")</li>
<li><code>strandedness.split_by_rg</code> (Boolean, default=false)</li>
<li><code>subsample.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>subsample.ncpu</code> (Int, default=2)</li>
<li><code>subsample.prefix</code> (String, default=basename(bam,".bam")): Prefix for output files</li>
<li><code>validate_bam.ignore_list</code> (Array[String], default=[])</li>
<li><code>validate_bam.index_validation_stringency_less_exhaustive</code> (Boolean, default=false)</li>
<li><code>validate_bam.max_errors</code> (Int, default=2147483647)</li>
<li><code>validate_bam.memory_gb</code> (Int, default=16)</li>
<li><code>validate_bam.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>validate_bam.outfile_name</code> (String, default=basename(bam,".bam") + ".ValidateSamFile.txt")</li>
<li><code>validate_bam.succeed_on_errors</code> (Boolean, default=false)</li>
<li><code>validate_bam.succeed_on_warnings</code> (Boolean, default=true)</li>
<li><code>validate_bam.summary_mode</code> (Boolean, default=false)</li>
<li><code>validate_input_bam.ignore_list</code> (Array[String], default=[])</li>
<li><code>validate_input_bam.index_validation_stringency_less_exhaustive</code> (Boolean, default=false)</li>
<li><code>validate_input_bam.max_errors</code> (Int, default=2147483647)</li>
<li><code>validate_input_bam.memory_gb</code> (Int, default=16)</li>
<li><code>validate_input_bam.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>validate_input_bam.outfile_name</code> (String, default=basename(bam,".bam") + ".ValidateSamFile.txt")</li>
<li><code>validate_input_bam.succeed_on_errors</code> (Boolean, default=false)</li>
<li><code>validate_input_bam.succeed_on_warnings</code> (Boolean, default=true)</li>
<li><code>validate_input_bam.summary_mode</code> (Boolean, default=false)</li>
<li><code>cell_ranger_bam_to_fastqs.bamtofastq.max_retries</code> (Int, default=1)</li>
<li><code>cell_ranger_bam_to_fastqs.bamtofastq.memory_gb</code> (Int, default=40)</li>
<li><code>cell_ranger_bam_to_fastqs.bamtofastq.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>cell_ranger_bam_to_fastqs.bamtofastq.ncpu</code> (Int, default=1)</li>
<li><code>cell_ranger_bam_to_fastqs.fqlint.disable_validator_codes</code> (Array[String], default=[])</li>
<li><code>cell_ranger_bam_to_fastqs.fqlint.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>cell_ranger_bam_to_fastqs.fqlint.modify_memory_gb</code> (Int, default=0)</li>
<li><code>cell_ranger_bam_to_fastqs.fqlint.paired_read_validation_level</code> (String, default="high")</li>
<li><code>cell_ranger_bam_to_fastqs.fqlint.panic</code> (Boolean, default=true)</li>
<li><code>cell_ranger_bam_to_fastqs.fqlint.single_read_validation_level</code> (String, default="high")</li>
<li><code>cell_ranger_bam_to_fastqs.quickcheck.modify_disk_size_gb</code> (Int, default=0)</li>
</ul>
<h3 id="outputs">Outputs</h3>
<ul>
<li><code>harmonized_bam</code> (File)</li>
<li><code>bam_checksum</code> (File)</li>
<li><code>bam_index</code> (File)</li>
<li><code>qc</code> (File)</li>
<li><code>barcodes</code> (File)</li>
<li><code>features</code> (File)</li>
<li><code>matrix</code> (File)</li>
<li><code>filtered_gene_h5</code> (File)</li>
<li><code>raw_gene_h5</code> (File)</li>
<li><code>raw_barcodes</code> (File)</li>
<li><code>raw_features</code> (File)</li>
<li><code>raw_matrix</code> (File)</li>
<li><code>mol_info_h5</code> (File)</li>
<li><code>web_summary</code> (File)</li>
<li><code>inferred_strandedness</code> (File)</li>
</ul>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../samtools-merge/" class="btn btn-neutral float-left" title="Samtools merge"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../star-db-build/" class="btn btn-neutral float-right" title="Star db build">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
      <p>Â© 2020-Present St. Jude Children's Research Hospital</p>
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/stjudecloud/workflows" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../samtools-merge/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../star-db-build/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "../..";</script>
    <script src="../../js/theme_extra.js"></script>
    <script src="../../js/theme.js"></script>
      <script src="../../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
