<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="author" content="St. Jude Cloud" />
      <link rel="shortcut icon" href="../../img/favicon.ico" />
    <title>Picard - St. Jude Cloud Bioinformatics Workflows</title>
    <link rel="stylesheet" href="../../css/theme.css" />
    <link rel="stylesheet" href="../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Picard";
        var mkdocs_page_input_path = "tasks/picard.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../.." class="icon icon-home"> St. Jude Cloud Bioinformatics Workflows
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../build_for_dnanexus/">Building WDL workflows for DNAnexus</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Tasks</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../bwa/">Bwa</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../cellranger/">Cellranger</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../deeptools/">Deeptools</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../estimate/">Estimate</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../fastqc/">Fastqc</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../fq/">Fq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../htseq/">Htseq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../kraken2/">Kraken2</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../md5sum/">Md5sum</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../mosdepth/">Mosdepth</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../multiqc/">Multiqc</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ngsderive/">Ngsderive</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="./">Picard</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#inputs">Inputs</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#required">Required</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#defaults">Defaults</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#outputs">Outputs</a>
    </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../qualimap/">Qualimap</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../samtools/">Samtools</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../star/">Star</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../util/">Utilities</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Workflows</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/10x-bam-to-fastqs/">Cell Ranger BAM to FASTQs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/ESTIMATE/">ESTIMATE</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/bam-to-fastqs/">Bam to fastqs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/bwa-db-build/">Bwa db build</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/dnaseq-standard/">Dnaseq standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/make-qc-reference/">Make qc reference</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/markdups-post/">MarkDuplicates Post</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/quality-check-standard/">Quality check standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/samtools-merge/">Samtools merge</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/scrnaseq-standard/">Scrnaseq standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/star-db-build/">Star db build</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../..">St. Jude Cloud Bioinformatics Workflows</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../.." class="icon icon-home" aria-label="Docs"></a></li>
          <li class="breadcrumb-item">Tasks</li>
      <li class="breadcrumb-item active">Picard</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/stjudecloud/workflows/edit/master/docs/tasks/picard.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <p><a href="https://broadinstitute.github.io/picard/">Homepage</a></p>
<h2 id="mark_duplicates">mark_duplicates</h2>
<dl>
<dt>description</dt>
<dd>Marks duplicate reads in the input BAM file using Picard</dd>
<dt>external_help</dt>
<dd>https://gatk.broadinstitute.org/hc/en-us/articles/360037052812-MarkDuplicates-Picard-</dd>
<dt>outputs</dt>
<dd>{'duplicate_marked_bam': 'The input BAM with computationally determined duplicates marked.', 'duplicate_marked_bam_index': 'The <code>.bai</code> BAM index file associated with <code>duplicate_marked_bam</code>', 'duplicate_marked_bam_md5': 'The md5sum of <code>duplicate_marked_bam</code>', 'mark_duplicates_metrics': {'description': 'The METRICS_FILE result of <code>picard MarkDuplicates</code>', 'external_help': 'http://broadinstitute.github.io/picard/picard-metric-definitions.html#DuplicationMetrics'}}</dd>
</dl>
<h3 id="inputs">Inputs</h3>
<h4 id="required">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file in which to mark duplicates</li>
</ul>
<h4 id="defaults">Defaults</h4>
<ul>
<li><code>create_bam</code> (Boolean, default=true): Enable BAM creation (true)? Or only output MarkDuplicates metrics (false)?</li>
<li><code>max_retries</code> (Int, default=1): Number of times to retry in case of failure</li>
<li><code>memory_gb</code> (Int, default=50): RAM to allocate for task, specified in GB</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>prefix</code> (String, default=basename(bam,".bam") + ".MarkDuplicates"): Prefix for the MarkDuplicates result files. The extensions <code>.bam</code>, <code>.bam.bai</code>, <code>.bam.md5</code>, and <code>.metrics.txt</code> will be added.</li>
</ul>
<h3 id="outputs">Outputs</h3>
<ul>
<li><code>duplicate_marked_bam</code> (File?)</li>
<li><code>duplicate_marked_bam_index</code> (File?)</li>
<li><code>duplicate_marked_bam_md5</code> (File?)</li>
<li><code>mark_duplicates_metrics</code> (File)</li>
</ul>
<h2 id="validate_bam">validate_bam</h2>
<dl>
<dt>description</dt>
<dd>Validates the input BAM file for correct formatting using Picard</dd>
<dt>external_help</dt>
<dd>https://gatk.broadinstitute.org/hc/en-us/articles/360057440611-ValidateSamFile-Picard-</dd>
<dt>outputs</dt>
<dd>{'validate_report': 'Validation report produced by <code>picard ValidateSamFile</code>. Validation warnings and errors are logged.', 'validated_bam': 'The unmodified input BAM after it has been succesfully validated'}</dd>
</dl>
<h3 id="inputs_1">Inputs</h3>
<h4 id="required_1">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file to validate</li>
</ul>
<h4 id="optional">Optional</h4>
<ul>
<li><code>reference_fasta</code> (File?): Reference genome in FASTA format. Presence of the reference FASTA allows for <code>NM</code> tag validation.</li>
</ul>
<h4 id="defaults_1">Defaults</h4>
<ul>
<li><code>ignore_list</code> (Array[String], default=["MISSING_PLATFORM_VALUE", "INVALID_PLATFORM_VALUE", "INVALID_MAPPING_QUALITY"]); <strong>description</strong>: List of Picard errors and warnings to ignore. Possible values can be found on the GATK website (see <code>external_help</code>).; <strong>external_help</strong>: https://gatk.broadinstitute.org/hc/en-us/articles/360035891231-Errors-in-SAM-or-BAM-files-can-be-diagnosed-with-ValidateSamFile</li>
<li><code>index_validation_stringency_less_exhaustive</code> (Boolean, default=false): Set <code>INDEX_VALIDATION_STRINGENCY=LESS_EXHAUSTIVE</code>?</li>
<li><code>max_errors</code> (Int, default=2147483647): Set the value of MAX_OUTPUT for <code>picard ValidateSamFile</code>. The Picard default is 100, a lower number can enable fast fail behavior</li>
<li><code>max_retries</code> (Int, default=1): Number of times to retry in case of failure</li>
<li><code>memory_gb</code> (Int, default=16): RAM to allocate for task, specified in GB</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>outfile_name</code> (String, default=basename(bam,".bam") + ".ValidateSamFile.txt"): Name for the ValidateSamFile report file</li>
<li><code>succeed_on_errors</code> (Boolean, default=false): Succeed the task even if errors <em>and/or</em> warnings are detected</li>
<li><code>succeed_on_warnings</code> (Boolean, default=true): Succeed the task if warnings are detected and there are no errors. Overridden by <code>succeed_on_errors</code></li>
<li><code>summary_mode</code> (Boolean, default=false): Enable SUMMARY mode?</li>
</ul>
<h3 id="outputs_1">Outputs</h3>
<ul>
<li><code>validate_report</code> (File)</li>
<li><code>validated_bam</code> (File)</li>
</ul>
<h2 id="sort">sort</h2>
<dl>
<dt>description</dt>
<dd>Sorts the input BAM file</dd>
<dt>external_help</dt>
<dd>https://gatk.broadinstitute.org/hc/en-us/articles/360036510732-SortSam-Picard-</dd>
<dt>outputs</dt>
<dd>{'sorted_bam': 'The input BAM after it has been sorted according to <code>sort_order</code>', 'sorted_bam_index': 'The <code>.bai</code> BAM index file associated with <code>sorted_bam</code>', 'sorted_bam_md5': 'The md5sum of <code>sorted_bam</code>'}</dd>
</dl>
<h3 id="inputs_2">Inputs</h3>
<h4 id="required_2">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file to sort</li>
</ul>
<h4 id="defaults_2">Defaults</h4>
<ul>
<li><code>max_retries</code> (Int, default=1): Number of times to retry in case of failure</li>
<li><code>memory_gb</code> (Int, default=25): RAM to allocate for task, specified in GB</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>prefix</code> (String, default=basename(bam,".bam") + ".sorted"): Prefix for the sorted BAM file and accessory files. The extensions <code>.bam</code>, <code>.bam.bai</code>, and <code>.bam.md5</code> will be added.</li>
<li><code>sort_order</code> (String, default="coordinate"); <strong>description</strong>: Order by which to sort the input BAM; <strong>choices</strong>: ['queryname', 'coordinate', 'duplicate']</li>
</ul>
<h3 id="outputs_2">Outputs</h3>
<ul>
<li><code>sorted_bam</code> (File)</li>
<li><code>sorted_bam_index</code> (File)</li>
<li><code>sorted_bam_md5</code> (File)</li>
</ul>
<h2 id="merge_sam_files">merge_sam_files</h2>
<dl>
<dt>description</dt>
<dd>Merges the input BAM files into a single BAM file. All input BAMs are assumed to be sorted according to <code>sort_order</code>.</dd>
<dt>external_help</dt>
<dd>https://gatk.broadinstitute.org/hc/en-us/articles/360057440751-MergeSamFiles-Picard-</dd>
<dt>outputs</dt>
<dd>{'merged_bam': 'The BAM resulting from merging all the input BAMs', 'merged_bam_index': 'The <code>.bai</code> BAM index file associated with <code>merged_bam</code>', 'merged_bam_md5': 'The md5sum of <code>merged_bam</code>'}</dd>
</dl>
<h3 id="inputs_3">Inputs</h3>
<h4 id="required_3">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bams</code> (Array[File], <strong>required</strong>): Input BAMs to merge. All BAMs are assumed to be sorted according to <code>sort_order</code>.</li>
<li><code>prefix</code> (String, <strong>required</strong>): Prefix for the merged BAM file and accessory files. The extensions <code>.bam</code>, <code>.bam.bai</code>, and <code>.bam.md5</code> will be added.</li>
</ul>
<h4 id="defaults_3">Defaults</h4>
<ul>
<li><code>max_retries</code> (Int, default=1): Number of times to retry in case of failure</li>
<li><code>memory_gb</code> (Int, default=40): RAM to allocate for task, specified in GB</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>sort_order</code> (String, default="coordinate"); <strong>description</strong>: Sort order for the output merged BAM. It is assumed all input BAMs share this order.; <strong>choices</strong>: ['unsorted', 'queryname', 'coordinate', 'duplicate', 'unknown']</li>
<li><code>threading</code> (Boolean, default=true): Option to create a background thread to encode, compress and write to disk the output file. The threaded version uses about 20% more CPU and decreases runtime by ~20% when writing out a compressed BAM file.</li>
</ul>
<h3 id="outputs_3">Outputs</h3>
<ul>
<li><code>merged_bam</code> (File)</li>
<li><code>merged_bam_index</code> (File)</li>
<li><code>merged_bam_md5</code> (File)</li>
</ul>
<h2 id="clean_sam">clean_sam</h2>
<dl>
<dt>description</dt>
<dd>Cleans the input BAM file. Cleans soft-clipping beyond end-of-reference, sets MAPQ=0 for unmapped reads.</dd>
<dt>external_help</dt>
<dd>https://gatk.broadinstitute.org/hc/en-us/articles/360036885571-CleanSam-Picard-</dd>
<dt>outputs</dt>
<dd>{'cleaned_bam': 'A cleaned version of the input BAM', 'cleaned_bam_index': 'The <code>.bai</code> BAM index file associated with <code>cleaned_bam</code>', 'cleaned_bam_md5': 'The md5sum of <code>cleaned_bam</code>'}</dd>
</dl>
<h3 id="inputs_4">Inputs</h3>
<h4 id="required_4">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file to clean</li>
</ul>
<h4 id="defaults_4">Defaults</h4>
<ul>
<li><code>max_retries</code> (Int, default=1): Number of times to retry in case of failure</li>
<li><code>memory_gb</code> (Int, default=25): RAM to allocate for task, specified in GB</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>prefix</code> (String, default=basename(bam,".bam") + ".cleaned"): Prefix for the cleaned BAM file and accessory files. The extensions <code>.bam</code>, <code>.bam.bai</code>, and <code>.bam.md5</code> will be added.</li>
</ul>
<h3 id="outputs_4">Outputs</h3>
<ul>
<li><code>cleaned_bam</code> (File)</li>
<li><code>cleaned_bam_index</code> (File)</li>
<li><code>cleaned_bam_md5</code> (File)</li>
</ul>
<h2 id="collect_wgs_metrics">collect_wgs_metrics</h2>
<dl>
<dt>description</dt>
<dd>Runs <code>picard CollectWgsMetrics</code> to collect metrics about the fractions of reads that pass base- and mapping-quality filters as well as coverage (read-depth) levels</dd>
<dt>external_help</dt>
<dd>https://gatk.broadinstitute.org/hc/en-us/articles/360037226132-CollectWgsMetrics-Picard-</dd>
<dt>outputs</dt>
<dd>{'wgs_metrics': {'description': 'Output report of <code>picard CollectWgsMetrics</code>', 'external_help': 'https://broadinstitute.github.io/picard/picard-metric-definitions.html#CollectWgsMetrics.WgsMetrics'}}</dd>
</dl>
<h3 id="inputs_5">Inputs</h3>
<h4 id="required_5">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file for which to calculate WGS metrics</li>
<li><code>reference_fasta</code> (File, <strong>required</strong>): Gzipped reference genome in FASTA format</li>
</ul>
<h4 id="defaults_5">Defaults</h4>
<ul>
<li><code>max_retries</code> (Int, default=1): Number of times to retry in case of failure</li>
<li><code>memory_gb</code> (Int, default=12): RAM to allocate for task, specified in GB</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>outfile_name</code> (String, default=basename(bam,".bam") + ".CollectWgsMetrics.txt"): Name for the metrics result file</li>
</ul>
<h3 id="outputs_5">Outputs</h3>
<ul>
<li><code>wgs_metrics</code> (File)</li>
</ul>
<h2 id="collect_alignment_summary_metrics">collect_alignment_summary_metrics</h2>
<dl>
<dt>description</dt>
<dd>Runs <code>picard CollectAlignmentSummaryMetrics</code> to calculate metrics detailing the quality of the read alignments as well as the proportion of the reads that passed machine signal-to-noise threshold quality filters</dd>
<dt>external_help</dt>
<dd>https://gatk.broadinstitute.org/hc/en-us/articles/360040507751-CollectAlignmentSummaryMetrics-Picard-</dd>
<dt>outputs</dt>
<dd>{'alignment_metrics': {'description': 'The text file output of <code>CollectAlignmentSummaryMetrics</code>', 'external_help': 'http://broadinstitute.github.io/picard/picard-metric-definitions.html#AlignmentSummaryMetrics'}, 'alignment_metrics_pdf': 'The PDF file output of <code>CollectAlignmentSummaryMetrics</code>'}</dd>
</dl>
<h3 id="inputs_6">Inputs</h3>
<h4 id="required_6">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file for which to calculate alignment metrics</li>
</ul>
<h4 id="defaults_6">Defaults</h4>
<ul>
<li><code>max_retries</code> (Int, default=1): Number of times to retry in case of failure</li>
<li><code>memory_gb</code> (Int, default=8): RAM to allocate for task, specified in GB</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>prefix</code> (String, default=basename(bam,".bam") + ".CollectAlignmentSummaryMetrics"): Prefix for the output report files. The extensions <code>.txt</code> and <code>.pdf</code> will be added.</li>
</ul>
<h3 id="outputs_6">Outputs</h3>
<ul>
<li><code>alignment_metrics</code> (File)</li>
<li><code>alignment_metrics_pdf</code> (File)</li>
</ul>
<h2 id="collect_gc_bias_metrics">collect_gc_bias_metrics</h2>
<dl>
<dt>description</dt>
<dd>Runs <code>picard CollectGcBiasMetrics</code> to collect information about the relative proportions of guanine (G) and cytosine (C) nucleotides</dd>
<dt>external_help</dt>
<dd>https://gatk.broadinstitute.org/hc/en-us/articles/360037593931-CollectGcBiasMetrics-Picard-</dd>
<dt>outputs</dt>
<dd>{'gc_bias_metrics': {'description': 'The full text file output of <code>CollectGcBiasMetrics</code>', 'external_help': 'http://broadinstitute.github.io/picard/picard-metric-definitions.html#GcBiasDetailMetrics'}, 'gc_bias_metrics_summary': {'description': 'The summary text file output of <code>CollectGcBiasMetrics</code>', 'external_help': 'http://broadinstitute.github.io/picard/picard-metric-definitions.html#GcBiasSummaryMetrics'}, 'gc_bias_metrics_pdf': 'The PDF file output of <code>CollectGcBiasMetrics</code>'}</dd>
</dl>
<h3 id="inputs_7">Inputs</h3>
<h4 id="required_7">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file for which to calculate GC bias metrics</li>
<li><code>reference_fasta</code> (File, <strong>required</strong>): Reference sequences in FASTA format</li>
</ul>
<h4 id="defaults_7">Defaults</h4>
<ul>
<li><code>max_retries</code> (Int, default=1): Number of times to retry in case of failure</li>
<li><code>memory_gb</code> (Int, default=8): RAM to allocate for task, specified in GB</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>prefix</code> (String, default=basename(bam,".bam") + ".CollectGcBiasMetrics"): Prefix for the output report files. The extensions <code>.txt</code>, <code>.summary.txt</code>, and <code>.pdf</code> will be added.</li>
</ul>
<h3 id="outputs_7">Outputs</h3>
<ul>
<li><code>gc_bias_metrics</code> (File)</li>
<li><code>gc_bias_metrics_summary</code> (File)</li>
<li><code>gc_bias_metrics_pdf</code> (File)</li>
</ul>
<h2 id="collect_insert_size_metrics">collect_insert_size_metrics</h2>
<dl>
<dt>description</dt>
<dd>Runs <code>picard CollectInsertSizeMetrics</code> to collect metrics for validating library construction including the insert size distribution and read orientation of paired-end libraries</dd>
<dt>external_help</dt>
<dd>https://gatk.broadinstitute.org/hc/en-us/articles/360037055772-CollectInsertSizeMetrics-Picard-</dd>
<dt>outputs</dt>
<dd>{'insert_size_metrics': {'description': 'The text file output of <code>CollectInsertSizeMetrics</code>', 'external_help': 'http://broadinstitute.github.io/picard/picard-metric-definitions.html#InsertSizeMetrics'}, 'insert_size_metrics_pdf': 'The PDF file output of <code>CollectInsertSizeMetrics</code>'}</dd>
</dl>
<h3 id="inputs_8">Inputs</h3>
<h4 id="required_8">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file for which to calculate insert size metrics</li>
</ul>
<h4 id="defaults_8">Defaults</h4>
<ul>
<li><code>max_retries</code> (Int, default=1): Number of times to retry in case of failure</li>
<li><code>memory_gb</code> (Int, default=8): RAM to allocate for task, specified in GB</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>prefix</code> (String, default=basename(bam,".bam") + ".CollectInsertSizeMetrics"): Prefix for the output report files. The extensions <code>.txt</code> and <code>.pdf</code> will be added.</li>
</ul>
<h3 id="outputs_8">Outputs</h3>
<ul>
<li><code>insert_size_metrics</code> (File)</li>
<li><code>insert_size_metrics_pdf</code> (File)</li>
</ul>
<h2 id="quality_score_distribution">quality_score_distribution</h2>
<dl>
<dt>description</dt>
<dd>Runs <code>picard QualityScoreDistribution</code> to calculate the range of quality scores and creates an accompanying chart</dd>
<dt>external_help</dt>
<dd>https://gatk.broadinstitute.org/hc/en-us/articles/360037057312-QualityScoreDistribution-Picard-</dd>
<dt>outputs</dt>
<dd>{'quality_score_distribution_txt': 'The text file output of <code>QualityScoreDistribution</code>', 'quality_score_distribution_pdf': 'The PDF file output of <code>QualityScoreDistribution</code>'}</dd>
</dl>
<h3 id="inputs_9">Inputs</h3>
<h4 id="required_9">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file for which to calculate quality score distribution</li>
</ul>
<h4 id="defaults_9">Defaults</h4>
<ul>
<li><code>max_retries</code> (Int, default=1): Number of times to retry in case of failure</li>
<li><code>memory_gb</code> (Int, default=8): RAM to allocate for task, specified in GB</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>prefix</code> (String, default=basename(bam,".bam") + ".QualityScoreDistribution"): Prefix for the output report files. The extensions <code>.txt</code> and <code>.pdf</code> will be added.</li>
</ul>
<h3 id="outputs_9">Outputs</h3>
<ul>
<li><code>quality_score_distribution_txt</code> (File)</li>
<li><code>quality_score_distribution_pdf</code> (File)</li>
</ul>
<h2 id="bam_to_fastq">bam_to_fastq</h2>
<dl>
<dt>description</dt>
<dd><strong>[Deprecated]</strong> This WDL task converts the input BAM file into FASTQ format files. This task has been deprecated in favor of <code>samtools.collate_to_fastq</code> which is more performant and doesn't error on 'illegal mate states'.</dd>
<dt>deprecated</dt>
<dd>true</dd>
</dl>
<h3 id="inputs_10">Inputs</h3>
<h4 id="required_10">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file to convert to FASTQ</li>
</ul>
<h4 id="defaults_10">Defaults</h4>
<ul>
<li><code>max_retries</code> (Int, default=1): Number of times to retry in case of failure</li>
<li><code>memory_gb</code> (Int, default=56): RAM to allocate for task, specified in GB</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>paired</code> (Boolean, default=true): Is the data paired-end (true) or single-end (false)?</li>
<li><code>prefix</code> (String, default=basename(bam,".bam"))</li>
</ul>
<h3 id="outputs_10">Outputs</h3>
<ul>
<li><code>read_one_fastq_gz</code> (File)</li>
<li><code>read_two_fastq_gz</code> (File?)</li>
</ul>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../ngsderive/" class="btn btn-neutral float-left" title="Ngsderive"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../qualimap/" class="btn btn-neutral float-right" title="Qualimap">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
      <p>Â© 2020-Present St. Jude Children's Research Hospital</p>
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/stjudecloud/workflows" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../ngsderive/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../qualimap/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "../..";</script>
    <script src="../../js/theme_extra.js"></script>
    <script src="../../js/theme.js"></script>
      <script src="../../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
