<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="author" content="St. Jude Cloud" />
      <link rel="shortcut icon" href="../../img/favicon.ico" />
    <title>RNA-Seq Standard from FastQ - St. Jude Cloud Bioinformatics Workflows</title>
    <link rel="stylesheet" href="../../css/theme.css" />
    <link rel="stylesheet" href="../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "RNA-Seq Standard from FastQ";
        var mkdocs_page_input_path = "workflows/rnaseq-standard-fastq.md";
        var mkdocs_page_url = null;
      </script>
    
    <script src="../../js/jquery-3.6.0.min.js" defer></script>
    <!--[if lt IE 9]>
      <script src="../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
      <script>hljs.initHighlightingOnLoad();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../.." class="icon icon-home"> St. Jude Cloud Bioinformatics Workflows
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../build_for_dnanexus/">Building WDL workflows for DNAnexus</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Tasks</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/bwa/">BWA</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/cellranger/">Cellranger</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/deeptools/">Deeptools</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/estimate/">ESTIMATE</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/fastq_screen/">FastQ Screen</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/fastqc/">FastQC</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/fq/">FQ</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/htseq/">HTSeq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/md5sum/">Md5sum</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/multiqc/">MultiQC</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/ngsderive/">ngsderive</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/picard/">Picard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/qualimap/">QualiMap</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/samtools/">SAMtools</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/sequencerr/">sequencErr</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/star/">STAR</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/util/">Utilities</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Workflows</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../10x-bam-to-fastqs/">Cell Ranger Bam to FastQs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ESTIMATE/">ESTIMATE</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../bam-to-fastqs/">Bam to FastQs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../bwa-db-build/">BWA DB build</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../calc-gene-lengths/">Calculate Gene Lengths</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../chipseq-bwa-db-build/">BWA DB build</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../make-qc-reference/">Make QC Reference</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../quality-check-standard/">Quality Check Standard</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="./">RNA-Seq Standard from FastQ</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#licensing">LICENSING</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#mit-license">MIT License</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#rnaseq_standard_fastq">rnaseq_standard_fastq</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#inputs">Inputs</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#required">Required</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#optional">Optional</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#defaults">Defaults</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#outputs">Outputs</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#parse_input">parse_input</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#inputs_1">Inputs</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#required_1">Required</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#optional_1">Optional</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#outputs_1">Outputs</a>
    </li>
        </ul>
    </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../rnaseq-standard/">RNA-Seq Standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../rnaseq-star-db-build/">RNASeq STAR DB build</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../scrnaseq-standard/">Scrnaseq standard</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../..">St. Jude Cloud Bioinformatics Workflows</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../.." class="icon icon-home" alt="Docs"></a> &raquo;</li>
          <li>Workflows &raquo;</li>
      <li>RNA-Seq Standard from FastQ</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/stjudecloud/workflows/edit/master/docs/workflows/rnaseq-standard-fastq.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="rna-seq-standard-from-fastq">RNA-Seq Standard from FastQ</h1>
<p>This WDL workflow runs the STAR RNA-seq alignment workflow for St. Jude Cloud from fastq input.
The workflow takes paired input fastq files. 
The read pairs are then passed through STAR alignment to generate a BAM file. The BAM is run
through several QC steps. Quantification is done using htseq-count. </p>
<h2 id="licensing">LICENSING</h2>
<h4 id="mit-license">MIT License</h4>
<p>Copyright 2020-Present St. Jude Children's Research Hospital</p>
<p>Permission is hereby granted, free of charge, to any person obtaining a copy of this
software and associated documentation files (the "Software"), to deal in the Software
without restriction, including without limitation the rights to use, copy, modify, merge,
publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons
to whom the Software is furnished to do so, subject to the following conditions:</p>
<p>The above copyright notice and this permission notice shall be included in all copies or
substantial portions of the Software.</p>
<p>THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.</p>
<h2 id="rnaseq_standard_fastq">rnaseq_standard_fastq</h2>
<h3 id="inputs">Inputs</h3>
<h4 id="required">Required</h4>
<ul>
<li><code>gtf</code> (File, <strong>required</strong>): GTF feature file</li>
<li><code>output_prefix</code> (String, <strong>required</strong>): Prefix for output files</li>
<li><code>read_groups</code> (String, <strong>required</strong>): A space-delimited read group record for each read group. Exactly one fastq filename must match each read group ID from <code>read_one_fastqs</code> and <code>read_two_fastqs</code>. Read group fields (Required fields: ID, LB, PL, PU, &amp; SM.) should be space delimited. Read groups should be comma separated, with a space on each side (e.g. ' , '). The ID field must come first for each read group and must match the basename of a fastq file (up to the first period). Expected form: <code>ID:rg1 PU:flowcell1.lane1 SM:sample1 PL:illumina LB:sample1_lib1 , ID:rg2 PU:flowcell1.lane2 SM:sample1 PL:illumina LB:sample1_lib1</code></li>
<li><code>read_one_fastqs</code> (Array[File], <strong>required</strong>): Input Fastq format file(s) with 1st read in pair to align</li>
<li><code>read_two_fastqs</code> (Array[File], <strong>required</strong>): Input Fastq format file(s) with 2nd read in pair to align</li>
<li><code>stardb_tar_gz</code> (File, <strong>required</strong>): Database of reference files for the STAR aligner. Can be generated by <code>rnaseq-star-db-build.wdl</code></li>
</ul>
<h4 id="optional">Optional</h4>
<ul>
<li><code>contaminant_stardb_tar_gz</code> (File?): If using XenoCP to clean contaminant reads, provide a STAR reference for the contaminant genome</li>
<li><code>alignment.disk_size_gb</code> (Int?)</li>
<li><code>picard_sort.disk_size_gb</code> (Int?)</li>
<li><code>xenocp.bwa_aln_align.disk_size_gb</code> (Int?)</li>
<li><code>xenocp.bwa_mem_align.disk_size_gb</code> (Int?)</li>
<li><code>xenocp.cleanse.disk_size_gb</code> (Int?)</li>
<li><code>xenocp.combine_files.disk_size_gb</code> (Int?)</li>
<li><code>xenocp.combine_tie_bam.disk_size_gb</code> (Int?)</li>
<li><code>xenocp.create_contam_list.disk_size_gb</code> (Int?)</li>
<li><code>xenocp.extract_by_chrom.disk_size_gb</code> (Int?)</li>
<li><code>xenocp.final_bam.disk_size_gb</code> (Int?)</li>
<li><code>xenocp.get_chroms.disk_size_gb</code> (Int?)</li>
<li><code>xenocp.mapped_fastq.disk_size_gb</code> (Int?)</li>
<li><code>xenocp.mismatch.disk_size_gb</code> (Int?)</li>
<li><code>xenocp.qc.disk_size_gb</code> (Int?)</li>
<li><code>xenocp.sort.disk_size_gb</code> (Int?)</li>
<li><code>xenocp.star_align.disk_size_gb</code> (Int?)</li>
<li><code>xenocp.star_align.read_groups</code> (String?): A space-delimited read group record for each read group. Exactly one fastq filename must match each read group ID from <code>read_one_fastqs</code> and <code>read_two_fastqs</code>. Read group fields (Required fields: ID, LB, PL, PU, &amp; SM.) should be space delimited. Read groups should be comma separated, with a space on each side (e.g. ' , '). The ID field must come first for each read group and must match the basename of a fastq file (up to the first period). Expected form: <code>ID:rg1 PU:flowcell1.lane1 SM:sample1 PL:illumina LB:sample1_lib1 , ID:rg2 PU:flowcell1.lane2 SM:sample1 PL:illumina LB:sample1_lib1</code></li>
<li><code>xenocp.star_align.read_two_fastqs</code> (Array[File]?): Input Fastq format file(s) with 2nd read in pair to align</li>
<li><code>xenocp.tie_sort.disk_size_gb</code> (Int?)</li>
<li><code>xenocp.unmapped.disk_size_gb</code> (Int?)</li>
</ul>
<h4 id="defaults">Defaults</h4>
<ul>
<li><code>cleanse_xenograft</code> (Boolean, default=false): For xenograft samples, enable XenoCP cleansing of mouse contamination</li>
<li><code>detect_nproc</code> (Boolean, default=false): Use all available cores for multi-core steps</li>
<li><code>max_retries</code> (Int, default=1): Number of times to retry failed steps</li>
<li><code>strandedness</code> (String, default=""): empty, 'Stranded-Reverse', 'Stranded-Forward', or 'Unstranded'. If missing, will be inferred</li>
<li><code>validate_input</code> (Boolean, default=true)</li>
<li><code>alignment.memory_gb</code> (Int, default=50)</li>
<li><code>alignment.ncpu</code> (Int, default=1)</li>
<li><code>deeptools_bamCoverage.detect_nproc</code> (Boolean, default=false): Use all available cores for multi-core steps</li>
<li><code>deeptools_bamCoverage.memory_gb</code> (Int, default=5)</li>
<li><code>deeptools_bamCoverage.ncpu</code> (Int, default=1)</li>
<li><code>deeptools_bamCoverage.prefix</code> (String, default=basename(bam,".bam"))</li>
<li><code>fqlint.memory_gb</code> (Int, default=8)</li>
<li><code>htseq_count.added_memory_gb</code> (Int, default=20)</li>
<li><code>htseq_count.outfile</code> (String, default=basename(bam,".bam") + ".feature-counts.txt")</li>
<li><code>ngsderive_strandedness.memory_gb</code> (Int, default=5)</li>
<li><code>picard_sort.memory_gb</code> (Int, default=25)</li>
<li><code>picard_sort.output_filename</code> (String, default=basename(bam,".bam") + ".sorted.bam")</li>
<li><code>picard_sort.sort_order</code> (String, default="coordinate")</li>
<li><code>samtools_index.memory_gb</code> (Int, default=15)</li>
<li><code>samtools_index.ncpu</code> (Int, default=1)</li>
<li><code>samtools_index.outfile</code> (String, default=basename(bam) + ".bai")</li>
<li><code>validate_bam.ignore_list</code> (Array[String], default=["MISSING_PLATFORM_VALUE", "INVALID_PLATFORM_VALUE", "INVALID_MAPPING_QUALITY"])</li>
<li><code>validate_bam.index_validation_stringency_less_exhaustive</code> (Boolean, default=false)</li>
<li><code>validate_bam.max_errors</code> (Int, default=2147483647)</li>
<li><code>validate_bam.memory_gb</code> (Int, default=8)</li>
<li><code>validate_bam.output_filename</code> (String, default=basename(bam,".bam") + ".ValidateSamFile.txt")</li>
<li><code>validate_bam.succeed_on_errors</code> (Boolean, default=false)</li>
<li><code>validate_bam.succeed_on_warnings</code> (Boolean, default=true)</li>
<li><code>validate_bam.summary_mode</code> (Boolean, default=false)</li>
<li><code>xenocp.n_threads</code> (Int, default=1)</li>
<li><code>xenocp.validation_stringency</code> (String, default="SILENT")</li>
<li><code>xenocp.bwa_aln_align.detect_nproc</code> (Boolean, default=false): Use all available cores for multi-core steps</li>
<li><code>xenocp.bwa_aln_align.max_retries</code> (Int, default=1): Number of times to retry failed steps</li>
<li><code>xenocp.bwa_aln_align.memory_gb</code> (Int, default=5)</li>
<li><code>xenocp.bwa_aln_align.output_bam</code> (String, default=basename(fastq,".fq.gz") + ".bam")</li>
<li><code>xenocp.bwa_aln_align.read_group</code> (String, default="")</li>
<li><code>xenocp.bwa_mem_align.detect_nproc</code> (Boolean, default=false): Use all available cores for multi-core steps</li>
<li><code>xenocp.bwa_mem_align.max_retries</code> (Int, default=1): Number of times to retry failed steps</li>
<li><code>xenocp.bwa_mem_align.output_bam</code> (String, default=basename(fastq,".fq.gz") + ".bam")</li>
<li><code>xenocp.bwa_mem_align.read_group</code> (String, default="")</li>
<li><code>xenocp.cleanse.max_retries</code> (Int, default=1): Number of times to retry failed steps</li>
<li><code>xenocp.cleanse.memory_gb</code> (Int, default=7)</li>
<li><code>xenocp.cleanse.output_bam</code> (String, default=basename(input_bam,".bam") + "-xenocp.bam")</li>
<li><code>xenocp.cleanse.sort_order</code> (String, default="coordinate")</li>
<li><code>xenocp.combine_files.max_retries</code> (Int, default=1): Number of times to retry failed steps</li>
<li><code>xenocp.combine_files.memory_gb</code> (Int, default=1)</li>
<li><code>xenocp.combine_tie_bam.max_retries</code> (Int, default=1): Number of times to retry failed steps</li>
<li><code>xenocp.combine_tie_bam.memory_gb</code> (Int, default=1)</li>
<li><code>xenocp.create_contam_list.max_retries</code> (Int, default=1): Number of times to retry failed steps</li>
<li><code>xenocp.create_contam_list.memory_gb</code> (Int, default=2)</li>
<li><code>xenocp.create_contam_list.output_contam_list</code> (String, default=basename(input_bam,".bam") + ".contam.txt")</li>
<li><code>xenocp.create_contam_list.tie_bam</code> (String, default=basename(input_bam,".bam") + ".tie.bam")</li>
<li><code>xenocp.extract_by_chrom.max_retries</code> (Int, default=1): Number of times to retry failed steps</li>
<li><code>xenocp.extract_by_chrom.memory_gb</code> (Int, default=8)</li>
<li><code>xenocp.final_bam.max_retries</code> (Int, default=1): Number of times to retry failed steps</li>
<li><code>xenocp.final_bam.memory_gb</code> (Int, default=1)</li>
<li><code>xenocp.get_chroms.detect_nproc</code> (Boolean, default=false): Use all available cores for multi-core steps</li>
<li><code>xenocp.get_chroms.max_retries</code> (Int, default=1): Number of times to retry failed steps</li>
<li><code>xenocp.get_chroms.memory_gb</code> (Int, default=1)</li>
<li><code>xenocp.mapped_fastq.max_retries</code> (Int, default=1): Number of times to retry failed steps</li>
<li><code>xenocp.mapped_fastq.memory_gb</code> (Int, default=1)</li>
<li><code>xenocp.mapped_fastq.output_fastq</code> (String, default=basename(input_bam,".bam") + ".fastq")</li>
<li><code>xenocp.mismatch.max_retries</code> (Int, default=1): Number of times to retry failed steps</li>
<li><code>xenocp.mismatch.memory_gb</code> (Int, default=2)</li>
<li><code>xenocp.qc.max_retries</code> (Int, default=1): Number of times to retry failed steps</li>
<li><code>xenocp.qc.memory_gb</code> (Int, default=1)</li>
<li><code>xenocp.qc.ncpu</code> (Int, default=1)</li>
<li><code>xenocp.qc.status</code> (String, default="indexed")</li>
<li><code>xenocp.sort.max_retries</code> (Int, default=1): Number of times to retry failed steps</li>
<li><code>xenocp.sort.memory_gb</code> (Int, default=25)</li>
<li><code>xenocp.sort.output_filename</code> (String, default=basename(bam,".bam") + ".sorted.bam")</li>
<li><code>xenocp.star_align.detect_nproc</code> (Boolean, default=false): Use all available cores for multi-core steps</li>
<li><code>xenocp.star_align.max_retries</code> (Int, default=1): Number of times to retry failed steps</li>
<li><code>xenocp.star_align.memory_gb</code> (Int, default=50)</li>
<li><code>xenocp.tie_sort.max_retries</code> (Int, default=1): Number of times to retry failed steps</li>
<li><code>xenocp.tie_sort.memory_gb</code> (Int, default=25)</li>
<li><code>xenocp.tie_sort.output_filename</code> (String, default=basename(bam,".bam") + ".sorted.bam")</li>
<li><code>xenocp.unmapped.detect_nproc</code> (Boolean, default=false): Use all available cores for multi-core steps</li>
<li><code>xenocp.unmapped.max_retries</code> (Int, default=1): Number of times to retry failed steps</li>
<li><code>xenocp.unmapped.memory_gb</code> (Int, default=1)</li>
</ul>
<h3 id="outputs">Outputs</h3>
<ul>
<li><code>bam</code> (File)</li>
<li><code>bam_index</code> (File)</li>
<li><code>star_log</code> (File)</li>
<li><code>gene_counts</code> (File)</li>
<li><code>inferred_strandedness</code> (File)</li>
<li><code>bigwig</code> (File)</li>
</ul>
<h2 id="parse_input">parse_input</h2>
<h3 id="inputs_1">Inputs</h3>
<h4 id="required_1">Required</h4>
<ul>
<li><code>cleanse_xenograft</code> (Boolean, <strong>required</strong>)</li>
<li><code>input_strand</code> (String, <strong>required</strong>)</li>
</ul>
<h4 id="optional_1">Optional</h4>
<ul>
<li><code>contaminant_stardb_tar_gz</code> (File?)</li>
</ul>
<h3 id="outputs_1">Outputs</h3>
<ul>
<li><code>input_check</code> (String)</li>
</ul>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../quality-check-standard/" class="btn btn-neutral float-left" title="Quality Check Standard"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../rnaseq-standard/" class="btn btn-neutral float-right" title="RNA-Seq Standard">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
      <p>© 2020-Present St. Jude Children's Research Hospital</p>
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/stjudecloud/workflows" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../quality-check-standard/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../rnaseq-standard/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script>var base_url = '../..';</script>
    <script src="../../js/theme_extra.js" defer></script>
    <script src="../../js/theme.js" defer></script>
      <script src="../../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
