<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="author" content="St. Jude Cloud" />
      <link rel="shortcut icon" href="../../img/favicon.ico" />
    <title>Make qc reference - St. Jude Cloud Bioinformatics Workflows</title>
    <link rel="stylesheet" href="../../css/theme.css" />
    <link rel="stylesheet" href="../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Make qc reference";
        var mkdocs_page_input_path = "workflows/make-qc-reference.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../.." class="icon icon-home"> St. Jude Cloud Bioinformatics Workflows
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../build_for_dnanexus/">Building WDL workflows for DNAnexus</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Tasks</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/arriba/">Arriba</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/bwa/">Bwa</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/cellranger/">Cellranger</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/deeptools/">Deeptools</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/estimate/">Estimate</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/fastqc/">Fastqc</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/fq/">Fq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/gatk4/">Gatk4</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/htseq/">Htseq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/kraken2/">Kraken2</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/librarian/">librarian</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/md5sum/">Md5sum</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/mosdepth/">Mosdepth</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/multiqc/">Multiqc</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/ngsderive/">Ngsderive</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/picard/">Picard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/qualimap/">Qualimap</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/read_group/">Read group</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/sambamba/">Sambamba</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/samtools/">Samtools</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/star/">Star</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/util/">Utilities</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Workflows</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../10x-bam-to-fastqs/">10x bam to fastqs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ESTIMATE/">ESTIMATE</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../bam-to-fastqs/">Bam to fastqs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../bwa-db-build/">Bwa db build</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../dnaseq-core/">Dnaseq core</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../dnaseq-standard-fastq/">Dnaseq standard fastq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../dnaseq-standard/">Dnaseq standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../flag_filter/">FlagFilter</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../gatk-reference/">Gatk reference</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="#">Make qc reference</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#inputs">Inputs</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#required">Required</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#optional">Optional</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#defaults">Defaults</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#outputs">Outputs</a>
    </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../markdups-post/">MarkDuplicates Post</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../quality-check-standard/">Quality check standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../rnaseq-variant-calling/">Rnaseq variant calling</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../samtools-merge/">Samtools merge</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../scrnaseq-standard/">Scrnaseq standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../star-db-build/">Star db build</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../..">St. Jude Cloud Bioinformatics Workflows</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../.." class="icon icon-home" aria-label="Docs"></a></li>
          <li class="breadcrumb-item">Workflows</li>
      <li class="breadcrumb-item active">Make qc reference</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/stjudecloud/workflows/edit/master/docs/workflows/make-qc-reference.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h2 id="make_qc_reference">make_qc_reference</h2>
<dl>
<dt>description</dt>
<dd>Downloads and creates all reference files needed to run the <code>quality_check</code> workflow</dd>
<dt>outputs</dt>
<dd>{'reference_fa': 'FASTA format reference file', 'gtf': 'GTF feature file', 'kraken_db': 'A complete Kraken2 database', 'coverage_beds': 'BED file for each feature type to use for coverage calculation'}</dd>
<dt>allowNestedInputs</dt>
<dd>true</dd>
</dl>
<h3 id="inputs">Inputs</h3>
<h4 id="required">Required</h4>
<ul>
<li><code>gtf_name</code> (String, <strong>required</strong>): Name of output GTF file</li>
<li><code>gtf_url</code> (String, <strong>required</strong>): URL to retrieve the reference GTF file from</li>
<li><code>reference_fa_name</code> (String, <strong>required</strong>): Name of output reference FASTA file</li>
<li><code>reference_fa_url</code> (String, <strong>required</strong>): URL to retrieve the reference FASTA file from</li>
<li><code>create_library_from_fastas._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>download_library._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>download_taxonomy._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>fastas_download._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>gtf_download._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>kraken_build_db._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>make_coverage_regions_bed._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>reference_download._runtime</code> (Any, <strong>required</strong>)</li>
</ul>
<h4 id="optional">Optional</h4>
<ul>
<li><code>fastas_download.md5sum</code> (String?)</li>
<li><code>gtf_download.md5sum</code> (String?)</li>
<li><code>reference_download.md5sum</code> (String?)</li>
</ul>
<h4 id="defaults">Defaults</h4>
<ul>
<li><code>coverage_feature_types</code> (Array[String], default=["exon", "CDS", "UTR"]): <code>choices</code> below are the possible values from a GENCODE GTF file. If you are using a different GTF source, you may need to adjust this parameter.; <strong>Choices</strong>: ['gene', 'transcript', 'exon', 'CDS', 'UTR', 'start_codon', 'stop_codon', 'Selenocysteine']; <strong>description</strong>: List of feature types to use for coverage calculation</li>
<li><code>gtf_disk_size_gb</code> (Int, default=10): Disk size (in GB) to allocate for downloading the GTF file</li>
<li><code>kraken_fasta_urls</code> (Array[String], default=[]): URLs for any additional FASTA files in NCBI format to download and include in the Kraken2 database. This allows the addition of individual genomes (or other sequences) of interest.</li>
<li><code>kraken_fastas</code> (Array[File], default=[]): Array of gzipped FASTA files. Each sequence's ID must contain either an NCBI accession number or an explicit assignment of the taxonomy ID using <code>kraken:taxid</code></li>
<li><code>kraken_fastas_disk_size_gb</code> (Int, default=10): Disk size (in GB) to allocate for downloading the FASTA files</li>
<li><code>kraken_libraries</code> (Array[String], default=["archaea", "bacteria", "plasmid", "viral", "human", "fungi", "protozoa", "UniVec_Core"]); <strong>description</strong>: List of kraken libraries to download; <strong>choices</strong>: ['archaea', 'bacteria', 'plasmid', 'viral', 'human', 'fungi', 'plant', 'protozoa', 'nt', 'UniVec', 'UniVec_Core']</li>
<li><code>protein</code> (Boolean, default=false): Construct a protein database?</li>
<li><code>reference_fa_disk_size_gb</code> (Int, default=10): Disk size (in GB) to allocate for downloading the reference FASTA file</li>
<li><code>create_library_from_fastas.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>download_library.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>kraken_build_db.db_name</code> (String, default="kraken2_db")</li>
<li><code>kraken_build_db.kmer_len</code> (Int, default=if protein then 15 else 35)</li>
<li><code>kraken_build_db.max_db_size_gb</code> (Int, default=-1)</li>
<li><code>kraken_build_db.minimizer_len</code> (Int, default=if protein then 12 else 31)</li>
<li><code>kraken_build_db.minimizer_spaces</code> (Int, default=if protein then 0 else 7)</li>
<li><code>kraken_build_db.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>kraken_build_db.modify_memory_gb</code> (Int, default=0)</li>
<li><code>kraken_build_db.ncpu</code> (Int, default=4)</li>
<li><code>kraken_build_db.use_all_cores</code> (Boolean, default=false)</li>
<li><code>make_coverage_regions_bed.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>make_coverage_regions_bed.outfile_name</code> (String, default=basename(gtf,"gtf.gz") + feature_type + ".bed")</li>
</ul>
<h3 id="outputs">Outputs</h3>
<ul>
<li><code>reference_fa</code> (File)</li>
<li><code>gtf</code> (File)</li>
<li><code>kraken_db</code> (File)</li>
<li><code>coverage_beds</code> (Array[File])</li>
</ul>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../gatk-reference/" class="btn btn-neutral float-left" title="Gatk reference"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../markdups-post/" class="btn btn-neutral float-right" title="MarkDuplicates Post">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
      <p>© 2020-Present St. Jude Children's Research Hospital</p>
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/stjudecloud/workflows" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../gatk-reference/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../markdups-post/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "../..";</script>
    <script src="../../js/theme_extra.js"></script>
    <script src="../../js/theme.js"></script>
      <script src="../../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
