<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="author" content="St. Jude Cloud" />
      <link rel="shortcut icon" href="../../img/favicon.ico" />
    <title>SAMtools - St. Jude Cloud Bioinformatics Workflows</title>
    <link rel="stylesheet" href="../../css/theme.css" />
    <link rel="stylesheet" href="../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "SAMtools";
        var mkdocs_page_input_path = "tasks/samtools.md";
        var mkdocs_page_url = null;
      </script>
    
    <script src="../../js/jquery-3.6.0.min.js" defer></script>
    <!--[if lt IE 9]>
      <script src="../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
      <script>hljs.initHighlightingOnLoad();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../.." class="icon icon-home"> St. Jude Cloud Bioinformatics Workflows
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../build_for_dnanexus/">Building WDL workflows for DNAnexus</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Tasks</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../bwa/">BWA</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../cellranger/">Cellranger</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../deeptools/">Deeptools</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../estimate/">ESTIMATE</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../fastqc/">FastQC</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../fq/">FQ</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../htseq/">HTSeq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../kraken/">Kraken2</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../md5sum/">Md5sum</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../mosdepth/">Mosdepth</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../multiqc/">MultiQC</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ngsderive/">ngsderive</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../picard/">Picard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../qualimap/">QualiMap</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="./">SAMtools</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#quickcheck">quickcheck</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#inputs">Inputs</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#required">Required</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#defaults">Defaults</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#outputs">Outputs</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#split">split</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#inputs_1">Inputs</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#required_1">Required</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#optional">Optional</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#defaults_1">Defaults</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#outputs_1">Outputs</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#flagstat">flagstat</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#inputs_2">Inputs</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#required_2">Required</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#defaults_2">Defaults</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#outputs_2">Outputs</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#index">index</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#inputs_3">Inputs</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#required_3">Required</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#defaults_3">Defaults</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#outputs_3">Outputs</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#subsample">subsample</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#inputs_4">Inputs</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#required_4">Required</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#defaults_4">Defaults</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#outputs_4">Outputs</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#merge">merge</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#inputs_5">Inputs</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#required_5">Required</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#optional_1">Optional</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#defaults_5">Defaults</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#outputs_5">Outputs</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#addreplacerg">addreplacerg</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#inputs_6">Inputs</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#required_6">Required</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#defaults_6">Defaults</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#outputs_6">Outputs</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#collate">collate</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#inputs_7">Inputs</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#required_7">Required</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#defaults_7">Defaults</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#outputs_7">Outputs</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#bam_to_fastq">bam_to_fastq</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#inputs_8">Inputs</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#required_8">Required</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#defaults_8">Defaults</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#outputs_8">Outputs</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#collate_to_fastq">collate_to_fastq</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#inputs_9">Inputs</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#required_9">Required</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#defaults_9">Defaults</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#outputs_9">Outputs</a>
    </li>
        </ul>
    </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../star/">STAR</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../util/">Utilities</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Workflows</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/10x-bam-to-fastqs/">Cell Ranger Bam to FastQs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/ESTIMATE/">ESTIMATE</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/bam-to-fastqs/">Bam to FastQs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/bwa-db-build/">BWA DB build</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/calc-gene-lengths/">Calculate Gene Lengths</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/make-qc-reference/">Make QC Reference</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/quality-check-standard/">Quality Check Standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/scrnaseq-standard/">Scrnaseq standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/star-db-build/">STAR DB build</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../..">St. Jude Cloud Bioinformatics Workflows</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../.." class="icon icon-home" aria-label="Docs"></a> &raquo;</li>
          <li>Tasks &raquo;</li>
      <li>SAMtools</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/stjudecloud/workflows/edit/master/docs/tasks/samtools.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="samtools">SAMtools</h1>
<p>This WDL file wraps the <a href="http://samtools.sourceforge.net/">SAMtools package</a>.
SAMtools provides utlities for manipulating SAM format sequence alignments.</p>
<h2 id="quickcheck">quickcheck</h2>
<dl>
<dt>author</dt>
<dd>Andrew Thrasher, Andrew Frantz</dd>
<dt>email</dt>
<dd>andrew.thrasher@stjude.org, andrew.frantz@stjude.org</dd>
<dt>description</dt>
<dd>This WDL task runs Samtools quickcheck on the input BAM file. This checks that the BAM file appears to be intact, e.g. header exists, at least one sequence is present, and the end-of-file marker exists.</dd>
</dl>
<h3 id="inputs">Inputs</h3>
<h4 id="required">Required</h4>
<ul>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file to quickcheck</li>
</ul>
<h4 id="defaults">Defaults</h4>
<ul>
<li><code>max_retries</code> (Int, default=1)</li>
</ul>
<h3 id="outputs">Outputs</h3>
<ul>
<li><code>checked_bam</code> (File)</li>
</ul>
<h2 id="split">split</h2>
<dl>
<dt>author</dt>
<dd>Andrew Thrasher, Andrew Frantz</dd>
<dt>email</dt>
<dd>andrew.thrasher@stjude.org, andrew.frantz@stjude.org</dd>
<dt>description</dt>
<dd>This WDL task runs Samtools split on the input BAM file. This splits the BAM by read group into one or more output files. It optionally errors if there are reads present that do not belong to a read group.</dd>
</dl>
<h3 id="inputs_1">Inputs</h3>
<h4 id="required_1">Required</h4>
<ul>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file to split</li>
</ul>
<h4 id="optional">Optional</h4>
<ul>
<li><code>disk_size_gb</code> (Int?)</li>
</ul>
<h4 id="defaults_1">Defaults</h4>
<ul>
<li><code>max_retries</code> (Int, default=1)</li>
<li><code>ncpu</code> (Int, default=1)</li>
<li><code>prefix</code> (String, default=basename(bam,".bam"))</li>
<li><code>reject_unaccounted</code> (Boolean, default=true): If true, error if there are reads present that do not have read group information.</li>
<li><code>use_all_cores</code> (Boolean, default=false)</li>
</ul>
<h3 id="outputs_1">Outputs</h3>
<ul>
<li><code>split_bams</code> (Array[File])</li>
</ul>
<h2 id="flagstat">flagstat</h2>
<dl>
<dt>author</dt>
<dd>Andrew Thrasher, Andrew Frantz</dd>
<dt>email</dt>
<dd>andrew.thrasher@stjude.org, andrew.frantz@stjude.org</dd>
<dt>description</dt>
<dd>This WDL tool generates a <code>samtools flagstat</code> report for the input BAM file.</dd>
</dl>
<h3 id="inputs_2">Inputs</h3>
<h4 id="required_2">Required</h4>
<ul>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file to generate flagstat for</li>
</ul>
<h4 id="defaults_2">Defaults</h4>
<ul>
<li><code>max_retries</code> (Int, default=1)</li>
<li><code>memory_gb</code> (Int, default=5)</li>
<li><code>outfile_name</code> (String, default=basename(bam,".bam") + ".flagstat.txt")</li>
</ul>
<h3 id="outputs_2">Outputs</h3>
<ul>
<li><code>flagstat_report</code> (File)</li>
</ul>
<h2 id="index">index</h2>
<dl>
<dt>author</dt>
<dd>Andrew Thrasher, Andrew Frantz</dd>
<dt>email</dt>
<dd>andrew.thrasher@stjude.org, andrew.frantz@stjude.org</dd>
<dt>description</dt>
<dd>This WDL task runs Samtools flagstat on the input BAM file. Produces statistics about the alignments based on the bit flags set in the BAM.</dd>
</dl>
<h3 id="inputs_3">Inputs</h3>
<h4 id="required_3">Required</h4>
<ul>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file to index</li>
</ul>
<h4 id="defaults_3">Defaults</h4>
<ul>
<li><code>max_retries</code> (Int, default=1)</li>
<li><code>memory_gb</code> (Int, default=15)</li>
<li><code>ncpu</code> (Int, default=1)</li>
<li><code>outfile_name</code> (String, default=basename(bam) + ".bai")</li>
<li><code>use_all_cores</code> (Boolean, default=false)</li>
</ul>
<h3 id="outputs_3">Outputs</h3>
<ul>
<li><code>bam_index</code> (File)</li>
</ul>
<h2 id="subsample">subsample</h2>
<h3 id="inputs_4">Inputs</h3>
<h4 id="required_4">Required</h4>
<ul>
<li><code>bam</code> (File, <strong>required</strong>)</li>
</ul>
<h4 id="defaults_4">Defaults</h4>
<ul>
<li><code>desired_reads</code> (Int, default=500000)</li>
<li><code>max_retries</code> (Int, default=1)</li>
<li><code>ncpu</code> (Int, default=1)</li>
<li><code>outfile_name</code> (String, default=basename(bam,".bam") + ".subsampled.bam")</li>
<li><code>use_all_cores</code> (Boolean, default=false)</li>
</ul>
<h3 id="outputs_4">Outputs</h3>
<ul>
<li><code>sampled_bam</code> (File)</li>
</ul>
<h2 id="merge">merge</h2>
<h3 id="inputs_5">Inputs</h3>
<h4 id="required_5">Required</h4>
<ul>
<li><code>bams</code> (Array[File], <strong>required</strong>)</li>
</ul>
<h4 id="optional_1">Optional</h4>
<ul>
<li><code>new_header</code> (File?)</li>
</ul>
<h4 id="defaults_5">Defaults</h4>
<ul>
<li><code>attach_rg</code> (Boolean, default=true)</li>
<li><code>max_retries</code> (Int, default=1)</li>
<li><code>ncpu</code> (Int, default=1)</li>
<li><code>outfile_name</code> (String, default=basename(bams[0],".bam") + ".merged.bam")</li>
<li><code>use_all_cores</code> (Boolean, default=false)</li>
</ul>
<h3 id="outputs_5">Outputs</h3>
<ul>
<li><code>merged_bam</code> (File)</li>
</ul>
<h2 id="addreplacerg">addreplacerg</h2>
<dl>
<dt>author</dt>
<dd>Andrew Thrasher</dd>
<dt>email</dt>
<dd>andrew.thrasher@stjude.org</dd>
<dt>description</dt>
<dd>This WDL task runs Samtools addreplacerg on the input BAM file. This adds an existing read group record to reads in the BAM lacking read group tags.</dd>
</dl>
<h3 id="inputs_6">Inputs</h3>
<h4 id="required_6">Required</h4>
<ul>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file to add read group information</li>
<li><code>read_group_id</code> (String, <strong>required</strong>): Existing read group ID in BAM to add to reads</li>
</ul>
<h4 id="defaults_6">Defaults</h4>
<ul>
<li><code>max_retries</code> (Int, default=1)</li>
<li><code>ncpu</code> (Int, default=1)</li>
<li><code>outfile_name</code> (String, default=basename(bam,".bam") + ".read_group.bam")</li>
<li><code>use_all_cores</code> (Boolean, default=false)</li>
</ul>
<h3 id="outputs_6">Outputs</h3>
<ul>
<li><code>tagged_bam</code> (File)</li>
</ul>
<h2 id="collate">collate</h2>
<dl>
<dt>description</dt>
<dd>This WDL task runs <code>samtools collate</code> on the input BAM file. Shuffles and groups reads together by their names.</dd>
<dt>outputs</dt>
<dd>{'collated_bam': 'A collated BAM (reads sharing a name next to each other, no other guarantee of sort order)'}</dd>
</dl>
<h3 id="inputs_7">Inputs</h3>
<h4 id="required_7">Required</h4>
<ul>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file to collate</li>
</ul>
<h4 id="defaults_7">Defaults</h4>
<ul>
<li><code>f</code> (Boolean, default=true): Fast mode (primary alignments only)</li>
<li><code>max_retries</code> (Int, default=1): Number of times to retry in case of failure</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>modify_memory_gb</code> (Int, default=0): Add to or subtract from dynamic memory allocation. Default memory is determined by the size of the inputs. Specified in GB.</li>
<li><code>ncpu</code> (Int, default=1): Number of cores to allocate for task</li>
<li><code>prefix</code> (String, default=basename(bam,".bam")): Prefix for the collated BAM file. The extension <code>.collated.bam</code> will be added.</li>
<li><code>use_all_cores</code> (Boolean, default=false): Use all cores? Recommended for cloud environments. Not recommended for cluster environments.</li>
</ul>
<h3 id="outputs_7">Outputs</h3>
<ul>
<li><code>collated_bam</code> (File)</li>
</ul>
<h2 id="bam_to_fastq">bam_to_fastq</h2>
<dl>
<dt>description</dt>
<dd>This WDL task runs <code>samtools fastq</code> on the input BAM file. Splits the BAM into FastQ files. Assumes either a name sorted or collated BAM. For splitting a position sorted BAM see <code>collate_to_fastq</code>.</dd>
<dt>outputs</dt>
<dd>{'read_one_fastq_gz': 'Gzipped FastQ file with 1st reads in pair', 'read_two_fastq_gz': 'Gzipped FastQ file with 2nd reads in pair', 'singleton_reads_fastq_gz': 'A gzipped FastQ containing singleton reads', 'interleaved_reads_fastq_gz': 'An interleaved gzipped paired-end FastQ'}</dd>
</dl>
<h3 id="inputs_8">Inputs</h3>
<h4 id="required_8">Required</h4>
<ul>
<li><code>bam</code> (File, <strong>required</strong>): Input name sorted or collated BAM format file to convert into FastQ(s)</li>
</ul>
<h4 id="defaults_8">Defaults</h4>
<ul>
<li><code>F</code> (String, default="0x900"): Do not output alignments with any bits set in INT present in the FLAG field. INT can be specified in hex by beginning with <code>0x</code> (i.e. /^0x[0-9A-F]+/) or in octal by beginning with <code>0</code> (i.e. /^0[0-7]+/). This defaults to 0x900 representing filtering of secondary and supplementary alignments.</li>
<li><code>G</code> (String, default="0"): Only EXCLUDE reads with all of the bits set in INT present in the FLAG field. INT can be specified in hex by beginning with <code>0x</code> (i.e. /^0x[0-9A-F]+/) or in octal by beginning with <code>0</code> (i.e. /^0[0-7]+/).</li>
<li><code>f</code> (String, default="0"): Only output alignments with all bits set in INT present in the FLAG field. INT can be specified in hex by beginning with <code>0x</code> (i.e. /^0x[0-9A-F]+/) or in octal by beginning with <code>0</code> (i.e. /^0[0-7]+/).</li>
<li><code>interleaved</code> (Boolean, default=false): Create an interleaved FastQ file from paired-end data?</li>
<li><code>max_retries</code> (Int, default=1): Number of times to retry in case of failure</li>
<li><code>memory_gb</code> (Int, default=4): RAM to allocate for task, specified in GB</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>ncpu</code> (Int, default=1): Number of cores to allocate for task</li>
<li><code>output_singletons</code> (Boolean, default=false): Output singleton reads as their own FastQ?</li>
<li><code>paired_end</code> (Boolean, default=true): Is the data paired-end?</li>
<li><code>prefix</code> (String, default=basename(bam,".bam")): Prefix for output FastQ(s). Extensions <code>[,_R1,_R2,.singleton].fastq.gz</code> will be added depending on other options.</li>
<li><code>use_all_cores</code> (Boolean, default=false): Use all available cores? Recommended for cloud environments. Not recommended for cluster environments.</li>
</ul>
<h3 id="outputs_8">Outputs</h3>
<ul>
<li><code>read_one_fastq_gz</code> (File?)</li>
<li><code>read_two_fastq_gz</code> (File?)</li>
<li><code>singleton_reads_fastq_gz</code> (File?)</li>
<li><code>interleaved_reads_fastq_gz</code> (File?)</li>
</ul>
<h2 id="collate_to_fastq">collate_to_fastq</h2>
<dl>
<dt>description</dt>
<dd>This WDL task runs <code>samtools collate</code> on the input BAM file then converts it into FastQ(s) using <code>samtools fastq</code>.</dd>
<dt>outputs</dt>
<dd>{'collated_bam': 'A collated BAM (reads sharing a name next to each other, no other guarantee of sort order)', 'read_one_fastq_gz': 'Gzipped FastQ file with 1st reads in pair', 'read_two_fastq_gz': 'Gzipped FastQ file with 2nd reads in pair', 'singleton_reads_fastq_gz': 'Gzipped FastQ containing singleton reads', 'interleaved_reads_fastq_gz': 'Interleaved gzipped paired-end FastQ'}</dd>
</dl>
<h3 id="inputs_9">Inputs</h3>
<h4 id="required_9">Required</h4>
<ul>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file to collate and convert to FastQ(s)</li>
</ul>
<h4 id="defaults_9">Defaults</h4>
<ul>
<li><code>F</code> (String, default="0x900"): Do not output alignments with any bits set in INT present in the FLAG field. INT can be specified in hex by beginning with <code>0x</code> (i.e. /^0x[0-9A-F]+/) or in octal by beginning with <code>0</code> (i.e. /^0[0-7]+/). This defaults to 0x900 representing filtering of secondary and supplementary alignments.</li>
<li><code>G</code> (String, default="0"): Only EXCLUDE reads with all of the bits set in INT present in the FLAG field. INT can be specified in hex by beginning with <code>0x</code> (i.e. /^0x[0-9A-F]+/) or in octal by beginning with <code>0</code> (i.e. /^0[0-7]+/).</li>
<li><code>f</code> (String, default="0"): Only output alignments with all bits set in INT present in the FLAG field. INT can be specified in hex by beginning with <code>0x</code> (i.e. /^0x[0-9A-F]+/) or in octal by beginning with <code>0</code> (i.e. /^0[0-7]+/).</li>
<li><code>fast_mode</code> (Boolean, default=true): Fast mode for <code>samtools collate</code> (primary alignments only)</li>
<li><code>interleaved</code> (Boolean, default=false): Create an interleaved FastQ file from paired-end data?</li>
<li><code>max_retries</code> (Int, default=1): Number of times to retry in case of failure</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>modify_memory_gb</code> (Int, default=0): Add to or subtract from dynamic memory allocation. Default memory is determined by the size of the inputs. Specified in GB.</li>
<li><code>ncpu</code> (Int, default=1): Number of cores to allocate for task</li>
<li><code>output_singletons</code> (Boolean, default=false): Output singleton reads as their own FastQ?</li>
<li><code>paired_end</code> (Boolean, default=true): Is the data paired-end?</li>
<li><code>prefix</code> (String, default=basename(bam,".bam")): Prefix for the collated BAM and FastQ files. The extensions <code>.collated.bam</code> and <code>[,_R1,_R2,.singleton].fastq.gz</code> will be added.</li>
<li><code>store_collated_bam</code> (Boolean, default=false): Save the collated BAM (true) or delete it after FastQ split (false)?</li>
<li><code>use_all_cores</code> (Boolean, default=false): Use all cores? Recommended for cloud environments. Not recommended for cluster environments.</li>
</ul>
<h3 id="outputs_9">Outputs</h3>
<ul>
<li><code>collated_bam</code> (File?)</li>
<li><code>read_one_fastq_gz</code> (File?)</li>
<li><code>read_two_fastq_gz</code> (File?)</li>
<li><code>singleton_reads_fastq_gz</code> (File?)</li>
<li><code>interleaved_reads_fastq_gz</code> (File?)</li>
</ul>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../qualimap/" class="btn btn-neutral float-left" title="QualiMap"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../star/" class="btn btn-neutral float-right" title="STAR">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
      <p>© 2020-Present St. Jude Children's Research Hospital</p>
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/stjudecloud/workflows" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../qualimap/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../star/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script>var base_url = '../..';</script>
    <script src="../../js/theme_extra.js" defer></script>
    <script src="../../js/theme.js" defer></script>
      <script src="../../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
