<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="author" content="St. Jude Cloud" />
      <link rel="shortcut icon" href="../../img/favicon.ico" />
    <title>Samtools - St. Jude Cloud Bioinformatics Workflows</title>
    <link rel="stylesheet" href="../../css/theme.css" />
    <link rel="stylesheet" href="../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Samtools";
        var mkdocs_page_input_path = "tasks/samtools.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../.." class="icon icon-home"> St. Jude Cloud Bioinformatics Workflows
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../build_for_dnanexus/">Building WDL workflows for DNAnexus</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Tasks</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../bwa/">Bwa</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../cellranger/">Cellranger</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../deeptools/">Deeptools</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../estimate/">Estimate</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../fastqc/">Fastqc</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../fq/">Fq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../htseq/">Htseq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../kraken2/">Kraken2</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../md5sum/">Md5sum</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../mosdepth/">Mosdepth</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../multiqc/">Multiqc</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ngsderive/">Ngsderive</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../picard/">Picard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../qualimap/">Qualimap</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="./">Samtools</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#inputs">Inputs</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#required">Required</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#defaults">Defaults</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#outputs">Outputs</a>
    </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../star/">Star</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../util/">Utilities</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Workflows</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/10x-bam-to-fastqs/">Cell Ranger BAM to FASTQs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/ESTIMATE/">ESTIMATE</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/bam-to-fastqs/">Bam to fastqs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/bwa-db-build/">Bwa db build</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/dnaseq-standard/">Dnaseq standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/make-qc-reference/">Make qc reference</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/markdups-post/">MarkDuplicates Post</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/quality-check-standard/">Quality check standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/samtools-merge/">Samtools merge</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/scrnaseq-standard/">Scrnaseq standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/star-db-build/">Star db build</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../..">St. Jude Cloud Bioinformatics Workflows</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../.." class="icon icon-home" aria-label="Docs"></a></li>
          <li class="breadcrumb-item">Tasks</li>
      <li class="breadcrumb-item active">Samtools</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/stjudecloud/workflows/edit/master/docs/tasks/samtools.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <p><a href="http://samtools.sourceforge.net/">Homepage</a></p>
<h2 id="quickcheck">quickcheck</h2>
<dl>
<dt>description</dt>
<dd>Runs Samtools quickcheck on the input BAM file. This checks that the BAM file appears to be intact, e.g. header exists, at least one sequence is present, and the end-of-file marker exists.</dd>
<dt>outputs</dt>
<dd>{'checked_bam': 'The unmodfied input BAM after it has been successfully quickchecked'}</dd>
</dl>
<h3 id="inputs">Inputs</h3>
<h4 id="required">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file to quickcheck</li>
</ul>
<h4 id="defaults">Defaults</h4>
<ul>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
</ul>
<h3 id="outputs">Outputs</h3>
<ul>
<li><code>checked_bam</code> (File)</li>
</ul>
<h2 id="split">split</h2>
<dl>
<dt>description</dt>
<dd>Runs Samtools split on the input BAM file. This splits the BAM by read group into one or more output files. It optionally errors if there are reads present that do not belong to a read group.</dd>
</dl>
<h3 id="inputs_1">Inputs</h3>
<h4 id="required_1">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file to split</li>
</ul>
<h4 id="defaults_1">Defaults</h4>
<ul>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>ncpu</code> (Int, default=2); <strong>description</strong>: Number of cores to allocate for task; <strong>common</strong>: true</li>
<li><code>prefix</code> (String, default=basename(bam,".bam")): Prefix for the split BAM files. The extensions will contain read group IDs, and will end in <code>.bam</code>.</li>
<li><code>reject_unaccounted</code> (Boolean, default=true); <strong>description</strong>: If true, error if there are reads present that do not have read group information.; <strong>common</strong>: true</li>
<li><code>use_all_cores</code> (Boolean, default=false); <strong>description</strong>: Use all cores? Recommended for cloud environments.; <strong>common</strong>: true</li>
</ul>
<h3 id="outputs_1">Outputs</h3>
<ul>
<li><code>split_bams</code> (Array[File])</li>
</ul>
<h2 id="flagstat">flagstat</h2>
<dl>
<dt>description</dt>
<dd>Produces a <code>samtools flagstat</code> report containing statistics about the alignments based on the bit flags set in the BAM</dd>
<dt>outputs</dt>
<dd>{'flagstat_report': '<code>samtools flagstat</code> STDOUT redirected to a file'}</dd>
</dl>
<h3 id="inputs_2">Inputs</h3>
<h4 id="required_2">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file to generate flagstat for</li>
</ul>
<h4 id="defaults_2">Defaults</h4>
<ul>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>ncpu</code> (Int, default=2); <strong>description</strong>: Number of cores to allocate for task; <strong>common</strong>: true</li>
<li><code>outfile_name</code> (String, default=basename(bam,".bam") + ".flagstat.txt"): Name for the flagstat report file</li>
<li><code>use_all_cores</code> (Boolean, default=false); <strong>description</strong>: Use all cores? Recommended for cloud environments.; <strong>common</strong>: true</li>
</ul>
<h3 id="outputs_2">Outputs</h3>
<ul>
<li><code>flagstat_report</code> (File)</li>
</ul>
<h2 id="index">index</h2>
<dl>
<dt>description</dt>
<dd>Creates a <code>.bai</code> BAM index for the input BAM</dd>
<dt>outputs</dt>
<dd>{'bam_index': "A <code>.bai</code> BAM index associated with the input BAM. Filename will be <code>basename(bam) + '.bai'</code>."}</dd>
</dl>
<h3 id="inputs_3">Inputs</h3>
<h4 id="required_3">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file to index</li>
</ul>
<h4 id="defaults_3">Defaults</h4>
<ul>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>ncpu</code> (Int, default=2); <strong>description</strong>: Number of cores to allocate for task; <strong>common</strong>: true</li>
<li><code>use_all_cores</code> (Boolean, default=false); <strong>description</strong>: Use all cores? Recommended for cloud environments.; <strong>common</strong>: true</li>
</ul>
<h3 id="outputs_3">Outputs</h3>
<ul>
<li><code>bam_index</code> (File)</li>
</ul>
<h2 id="subsample">subsample</h2>
<dl>
<dt>description</dt>
<dd>Randomly subsamples the input BAM. Sampling is probabalistic and will be approximate to <code>desired_reads</code>. Read count will not be exact. A <code>sampled_bam</code> will not be produced if the input BAM read count is less than or equal to <code>desired_reads</code>.</dd>
<dt>outputs</dt>
<dd>{'orig_read_count': 'A TSV report containing the original read count before subsampling', 'sampled_bam': 'The subsampled input BAM.'}</dd>
</dl>
<h3 id="inputs_4">Inputs</h3>
<h4 id="required_4">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file to subsample</li>
<li><code>desired_reads</code> (Int, <strong>required</strong>): How many reads should be in the ouput BAM? Output BAM read count will be approximate to this value.</li>
</ul>
<h4 id="defaults_4">Defaults</h4>
<ul>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>ncpu</code> (Int, default=2); <strong>description</strong>: Number of cores to allocate for task; <strong>common</strong>: true</li>
<li><code>prefix</code> (String, default=basename(bam,".bam")): Prefix for the BAM file. The extension <code>.subsampled.bam</code> will be added.</li>
<li><code>use_all_cores</code> (Boolean, default=false); <strong>description</strong>: Use all cores? Recommended for cloud environments.; <strong>common</strong>: true</li>
</ul>
<h3 id="outputs_4">Outputs</h3>
<ul>
<li><code>orig_read_count</code> (File)</li>
<li><code>sampled_bam</code> (File?)</li>
</ul>
<h2 id="merge">merge</h2>
<dl>
<dt>description</dt>
<dd>Merges multiple sorted BAMs into a single BAM</dd>
<dt>outputs</dt>
<dd>{'merged_bam': 'The BAM resulting from merging all the input BAMs'}</dd>
</dl>
<h3 id="inputs_5">Inputs</h3>
<h4 id="required_5">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bams</code> (Array[File], <strong>required</strong>): An array of BAMs to merge into one combined BAM</li>
<li><code>prefix</code> (String, <strong>required</strong>): Prefix for the BAM file. The extension <code>.bam</code> will be added.</li>
</ul>
<h4 id="optional">Optional</h4>
<ul>
<li><code>new_header</code> (File?): Use the lines of FILE as <code>@</code> headers to be copied to the merged BAM, replacing any header lines that would otherwise be copied from the first BAM file in the list. (File may actually be in SAM format, though any alignment records it may contain are ignored.)</li>
</ul>
<h4 id="defaults_5">Defaults</h4>
<ul>
<li><code>attach_rg</code> (Boolean, default=true); <strong>description</strong>: Attach an RG tag to each alignment. The tag value is inferred from file names.; <strong>common</strong>: true</li>
<li><code>combine_pg</code> (Boolean, default=true); <strong>description</strong>: Similarly to <code>combine_rg</code>: for each @PG ID in the set of files to merge, use the @PG line of the first file we find that ID in rather than adding a suffix to differentiate similar IDs.; <strong>common</strong>: true</li>
<li><code>combine_rg</code> (Boolean, default=true); <strong>description</strong>: When several input files contain @RG headers with the same ID, emit only one of them (namely, the header line from the first file we find that ID in) to the merged output file. Combining these similar headers is usually the right thing to do when the files being merged originated from the same file. Without <code>-c</code>, all @RG headers appear in the output file, with random suffixes added to their IDs where necessary to differentiate them.; <strong>common</strong>: true</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>name_sorted</code> (Boolean, default=false); <strong>description</strong>: Are <em>all</em> input BAMs <code>queryname</code> sorted (true)? Or are <em>all</em> input BAMs <code>coordinate</code> sorted (false)?; <strong>common</strong>: true</li>
<li><code>ncpu</code> (Int, default=2); <strong>description</strong>: Number of cores to allocate for task; <strong>common</strong>: true</li>
<li><code>region</code> (String, default=""): Merge files in the specified region (Format: <code>chr:start-end</code>)</li>
<li><code>use_all_cores</code> (Boolean, default=false); <strong>description</strong>: Use all cores? Recommended for cloud environments.; <strong>common</strong>: true</li>
</ul>
<h3 id="outputs_5">Outputs</h3>
<ul>
<li><code>merged_bam</code> (File)</li>
</ul>
<h2 id="addreplacerg">addreplacerg</h2>
<dl>
<dt>description</dt>
<dd>Adds or replaces read group tags</dd>
<dt>outputs</dt>
<dd>{'tagged_bam': 'The transformed input BAM after read group modifications have been applied'}</dd>
</dl>
<h3 id="inputs_6">Inputs</h3>
<h4 id="required_6">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file to add read group information</li>
</ul>
<h4 id="optional_1">Optional</h4>
<ul>
<li><code>read_group_id</code> (String?): Allows you to specify the read group ID of an existing @RG line and applies it to the reads specified by the <code>orphan_only</code> option</li>
</ul>
<h4 id="defaults_6">Defaults</h4>
<ul>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>ncpu</code> (Int, default=2); <strong>description</strong>: Number of cores to allocate for task; <strong>common</strong>: true</li>
<li><code>orphan_only</code> (Boolean, default=true); <strong>description</strong>: Only add RG tags to orphans (true)? Or <em>also</em> overwrite all existing RG tags (including any in the header) (false)?; <strong>common</strong>: true</li>
<li><code>overwrite_header_record</code> (Boolean, default=false); <strong>description</strong>: Overwrite an existing @RG line, if a new one with the same ID value is provided?; <strong>common</strong>: true</li>
<li><code>prefix</code> (String, default=basename(bam,".bam") + ".addreplacerg"): Prefix for the BAM file. The extension <code>.bam</code> will be added.</li>
<li><code>read_group_line</code> (Array[String], default=[]); <strong>description</strong>: Allows you to specify a read group line to append to (or replace in) the header and applies it to the reads specified by the <code>orphan_only</code> option. Each String in the Array should correspond to one field of the read group line. Tab literals will be inserted between each entry in the final BAM. Only <strong>one</strong> read group line can be supplied per invocation of this task.; <strong>common</strong>: true</li>
<li><code>use_all_cores</code> (Boolean, default=false); <strong>description</strong>: Use all cores? Recommended for cloud environments.; <strong>common</strong>: true</li>
</ul>
<h3 id="outputs_6">Outputs</h3>
<ul>
<li><code>tagged_bam</code> (File)</li>
</ul>
<h2 id="collate">collate</h2>
<dl>
<dt>description</dt>
<dd>Runs <code>samtools collate</code> on the input BAM file. Shuffles and groups reads together by their names.</dd>
<dt>outputs</dt>
<dd>{'collated_bam': 'A collated BAM (reads sharing a name next to each other, no other guarantee of sort order)'}</dd>
</dl>
<h3 id="inputs_7">Inputs</h3>
<h4 id="required_7">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file to collate</li>
</ul>
<h4 id="defaults_7">Defaults</h4>
<ul>
<li><code>fast_mode</code> (Boolean, default=true); <strong>description</strong>: Use fast mode (output primary alignments only)?; <strong>common</strong>: true</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>modify_memory_gb</code> (Int, default=0): Add to or subtract from dynamic memory allocation. Default memory is determined by the size of the inputs. Specified in GB.</li>
<li><code>ncpu</code> (Int, default=2); <strong>description</strong>: Number of cores to allocate for task; <strong>common</strong>: true</li>
<li><code>prefix</code> (String, default=basename(bam,".bam") + ".collated"): Prefix for the collated BAM file. The extension <code>.bam</code> will be added.</li>
<li><code>use_all_cores</code> (Boolean, default=false); <strong>description</strong>: Use all cores? Recommended for cloud environments.; <strong>common</strong>: true</li>
</ul>
<h3 id="outputs_7">Outputs</h3>
<ul>
<li><code>collated_bam</code> (File)</li>
</ul>
<h2 id="bam_to_fastq">bam_to_fastq</h2>
<dl>
<dt>description</dt>
<dd>Runs <code>samtools fastq</code> on the input BAM file. Converts the BAM into FASTQ files. If <code>paired_end = false</code>, then <em>all</em> reads in the BAM will be output to a single FASTQ file. Use filtering arguments to remove any unwanted reads. Assumes either a name sorted or collated BAM. For splitting a position sorted BAM see <code>collate_to_fastq</code>.</dd>
<dt>outputs</dt>
<dd>{'read_one_fastq_gz': 'Gzipped FASTQ file with 1st reads in pair', 'read_two_fastq_gz': 'Gzipped FASTQ file with 2nd reads in pair', 'singleton_reads_fastq_gz': 'A gzipped FASTQ containing singleton reads', 'interleaved_reads_fastq_gz': 'An interleaved gzipped paired-end FASTQ', 'single_end_reads_fastq_gz': 'A gzipped FASTQ containing all reads'}</dd>
</dl>
<h3 id="inputs_8">Inputs</h3>
<h4 id="required_8">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam</code> (File, <strong>required</strong>): Input name sorted or collated BAM format file to convert into FASTQ(s)</li>
</ul>
<h4 id="defaults_8">Defaults</h4>
<ul>
<li><code>F</code> (String, default="0x900"): Do not output alignments with any bits set in INT present in the FLAG field. INT can be specified in hex by beginning with <code>0x</code> (i.e. /^0x[0-9A-F]+/) or in octal by beginning with <code>0</code> (i.e. /^0[0-7]+/). This defaults to 0x900 representing filtering of secondary and supplementary alignments.</li>
<li><code>G</code> (String, default="0"): Only EXCLUDE reads with all of the bits set in INT present in the FLAG field. INT can be specified in hex by beginning with <code>0x</code> (i.e. /^0x[0-9A-F]+/) or in octal by beginning with <code>0</code> (i.e. /^0[0-7]+/).</li>
<li><code>append_read_number</code> (Boolean, default=true); <strong>description</strong>: Append /1 and /2 suffixes to read names; <strong>common</strong>: true</li>
<li><code>f</code> (String, default="0"): Only output alignments with all bits set in INT present in the FLAG field. INT can be specified in hex by beginning with <code>0x</code> (i.e. /^0x[0-9A-F]+/) or in octal by beginning with <code>0</code> (i.e. /^0[0-7]+/).</li>
<li><code>fail_on_unexpected_reads</code> (Boolean, default=false): The definition of 'unexpected' depends on whether the values of <code>paired_end</code> and <code>output_singletons</code> are true or false. In any case, reads that have neither or both <code>first</code> and <code>last</code> bits set are considered unexpected. If <code>paired_end</code> is <code>true</code> and <code>output_singletons</code> is <code>false</code>, singleton reads are considered unexpected. A singleton read is a read with either the <code>first</code> or the <code>last</code> bit set and that possesses a <em>unique</em> QNAME; i.e. it is a read without a pair when all reads are expected to be paired. But if <code>output_singletons</code> is <code>true</code>, these singleton reads will be output as their own FASTQ instead of causing the task to fail. If <code>fail_on_unexpected_reads</code> is <code>false</code>, then all the above cases will be ignored. Any 'unexpected' reads will be silently discarded. If <code>paired_end</code> is <code>false</code>, no reads are considered unexpected, and <em>every</em> read will be present in the resulting FASTQ regardless of bit settings.; <strong>description</strong>: Should the task fail if reads with an unexpected <code>first</code>/<code>last</code> bit setting are discovered?; <strong>common</strong>: true</li>
<li><code>interleaved</code> (Boolean, default=false); <strong>description</strong>: Create an interleaved FASTQ file from Paired-End data? Ignored if <code>paired_end = false</code>.; <strong>common</strong>: true</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>ncpu</code> (Int, default=2); <strong>description</strong>: Number of cores to allocate for task; <strong>common</strong>: true</li>
<li><code>output_singletons</code> (Boolean, default=false): Output singleton reads as their own FASTQ?</li>
<li><code>paired_end</code> (Boolean, default=true); <strong>description</strong>: Is the data Paired-End? If <code>paired_end = false</code>, then <em>all</em> reads in the BAM will be output to a single FASTQ file. Use filtering arguments to remove any unwanted reads.; <strong>common</strong>: true</li>
<li><code>prefix</code> (String, default=basename(bam,".bam")): Prefix for output FASTQ(s). Extensions <code>[,.R1,.R2,.singleton].fastq.gz</code> will be added depending on other options.</li>
<li><code>use_all_cores</code> (Boolean, default=false); <strong>description</strong>: Use all cores? Recommended for cloud environments.; <strong>common</strong>: true</li>
</ul>
<h3 id="outputs_8">Outputs</h3>
<ul>
<li><code>read_one_fastq_gz</code> (File?)</li>
<li><code>read_two_fastq_gz</code> (File?)</li>
<li><code>singleton_reads_fastq_gz</code> (File?)</li>
<li><code>interleaved_reads_fastq_gz</code> (File?)</li>
<li><code>single_end_reads_fastq_gz</code> (File?)</li>
</ul>
<h2 id="collate_to_fastq">collate_to_fastq</h2>
<dl>
<dt>description</dt>
<dd>Runs <code>samtools collate</code> on the input BAM file then converts it into FASTQ(s) using <code>samtools fastq</code></dd>
<dt>outputs</dt>
<dd>{'collated_bam': 'A collated BAM (reads sharing a name next to each other, no other guarantee of sort order)', 'read_one_fastq_gz': 'Gzipped FASTQ file with 1st reads in pair', 'read_two_fastq_gz': 'Gzipped FASTQ file with 2nd reads in pair', 'singleton_reads_fastq_gz': 'Gzipped FASTQ containing singleton reads', 'interleaved_reads_fastq_gz': 'Interleaved gzipped paired-end FASTQ', 'single_end_reads_fastq_gz': 'A gzipped FASTQ containing all reads'}</dd>
</dl>
<h3 id="inputs_9">Inputs</h3>
<h4 id="required_9">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file to collate and convert to FASTQ(s)</li>
</ul>
<h4 id="defaults_9">Defaults</h4>
<ul>
<li><code>F</code> (String, default="0x900"): Do not output alignments with any bits set in INT present in the FLAG field. INT can be specified in hex by beginning with <code>0x</code> (i.e. /^0x[0-9A-F]+/) or in octal by beginning with <code>0</code> (i.e. /^0[0-7]+/). This defaults to 0x900 representing filtering of secondary and supplementary alignments.</li>
<li><code>G</code> (String, default="0"): Only EXCLUDE reads with all of the bits set in INT present in the FLAG field. INT can be specified in hex by beginning with <code>0x</code> (i.e. /^0x[0-9A-F]+/) or in octal by beginning with <code>0</code> (i.e. /^0[0-7]+/).</li>
<li><code>append_read_number</code> (Boolean, default=true); <strong>description</strong>: Append /1 and /2 suffixes to read names; <strong>common</strong>: true</li>
<li><code>f</code> (String, default="0"): Only output alignments with all bits set in INT present in the FLAG field. INT can be specified in hex by beginning with <code>0x</code> (i.e. /^0x[0-9A-F]+/) or in octal by beginning with <code>0</code> (i.e. /^0[0-7]+/).</li>
<li><code>fail_on_unexpected_reads</code> (Boolean, default=false): The definition of 'unexpected' depends on whether the values of <code>paired_end</code> and <code>output_singletons</code> are true or false. In any case, reads that have neither or both <code>first</code> and <code>last</code> bits set are considered unexpected. If <code>paired_end</code> is <code>true</code> and <code>output_singletons</code> is <code>false</code>, singleton reads are considered unexpected. A singleton read is a read with either the <code>first</code> or the <code>last</code> bit set and that possesses a <em>unique</em> QNAME; i.e. it is a read without a pair when all reads are expected to be paired. But if <code>output_singletons</code> is <code>true</code>, these singleton reads will be output as their own FASTQ instead of causing the task to fail. If <code>fail_on_unexpected_reads</code> is <code>false</code>, then all the above cases will be ignored. Any 'unexpected' reads will be silently discarded. If <code>paired_end</code> is <code>false</code>, no reads are considered unexpected, and <em>every</em> read will be present in the resulting FASTQ regardless of bit settings.; <strong>description</strong>: Should the task fail if reads with an unexpected <code>first</code>/<code>last</code> bit setting are discovered?; <strong>common</strong>: true</li>
<li><code>fast_mode</code> (Boolean, default=true); <strong>description</strong>: Fast mode for <code>samtools collate</code> (primary alignments only); <strong>common</strong>: true</li>
<li><code>interleaved</code> (Boolean, default=false); <strong>description</strong>: Create an interleaved FASTQ file from Paired-End data? Ignored if <code>paired_end = false</code>.; <strong>common</strong>: true</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>modify_memory_gb</code> (Int, default=0): Add to or subtract from dynamic memory allocation. Default memory is determined by the size of the inputs. Specified in GB.</li>
<li><code>ncpu</code> (Int, default=2); <strong>description</strong>: Number of cores to allocate for task; <strong>common</strong>: true</li>
<li><code>output_singletons</code> (Boolean, default=false): Output singleton reads as their own FASTQ?</li>
<li><code>paired_end</code> (Boolean, default=true); <strong>description</strong>: Is the data Paired-End? If <code>paired_end = false</code>, then <em>all</em> reads in the BAM will be output to a single FASTQ file. Use filtering arguments to remove any unwanted reads.; <strong>common</strong>: true</li>
<li><code>prefix</code> (String, default=basename(bam,".bam")): Prefix for the collated BAM and FASTQ files. The extensions <code>.collated.bam</code> and <code>[,.R1,.R2,.singleton].fastq.gz</code> will be added.</li>
<li><code>store_collated_bam</code> (Boolean, default=false); <strong>description</strong>: Save the collated BAM to disk and output it (true)?; <strong>common</strong>: true</li>
<li><code>use_all_cores</code> (Boolean, default=false); <strong>description</strong>: Use all cores? Recommended for cloud environments.; <strong>common</strong>: true</li>
</ul>
<h3 id="outputs_9">Outputs</h3>
<ul>
<li><code>collated_bam</code> (File?)</li>
<li><code>read_one_fastq_gz</code> (File?)</li>
<li><code>read_two_fastq_gz</code> (File?)</li>
<li><code>singleton_reads_fastq_gz</code> (File?)</li>
<li><code>interleaved_reads_fastq_gz</code> (File?)</li>
<li><code>single_end_reads_fastq_gz</code> (File?)</li>
</ul>
<h2 id="fixmate">fixmate</h2>
<dl>
<dt>description</dt>
<dd>Runs <code>samtools fixmate</code> on the input BAM file. This fills in mate coordinates and insert size fields.</dd>
</dl>
<h3 id="inputs_10">Inputs</h3>
<h4 id="required_10">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file to add mate information. Must be name-sorted or name-collated.</li>
</ul>
<h4 id="defaults_10">Defaults</h4>
<ul>
<li><code>add_cigar</code> (Boolean, default=true): Add template cigar ct tag</li>
<li><code>add_mate_score</code> (Boolean, default=true): Add mate score tags. These are used by markdup to select the best reads to keep.</li>
<li><code>disable_proper_pair_check</code> (Boolean, default=false): Disable proper pair check [ensure one forward and one reverse read in each pair]</li>
<li><code>extension</code> (String, default=".bam"); <strong>description</strong>: File format extension to use for output file.; <strong>choices</strong>: ['.bam', '.cram']; <strong>common</strong>: true</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>ncpu</code> (Int, default=2); <strong>description</strong>: Number of cores to allocate for task; <strong>common</strong>: true</li>
<li><code>prefix</code> (String, default=basename(bam,".bam") + ".fixmate"): Prefix for the output file. The extension specified with the <code>extension</code> parameter will be added.</li>
<li><code>remove_unaligned_and_secondary</code> (Boolean, default=false): Remove unmapped and secondary reads</li>
<li><code>use_all_cores</code> (Boolean, default=false); <strong>description</strong>: Use all cores? Recommended for cloud environments.; <strong>common</strong>: true</li>
</ul>
<h3 id="outputs_10">Outputs</h3>
<ul>
<li><code>fixmate_bam</code> (File)</li>
</ul>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../qualimap/" class="btn btn-neutral float-left" title="Qualimap"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../star/" class="btn btn-neutral float-right" title="Star">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
      <p>© 2020-Present St. Jude Children's Research Hospital</p>
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/stjudecloud/workflows" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../qualimap/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../star/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "../..";</script>
    <script src="../../js/theme_extra.js"></script>
    <script src="../../js/theme.js"></script>
      <script src="../../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
