<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="author" content="St. Jude Cloud" />
      <link rel="shortcut icon" href="../../img/favicon.ico" />
    <title>Htseq - St. Jude Cloud Bioinformatics Workflows</title>
    <link rel="stylesheet" href="../../css/theme.css" />
    <link rel="stylesheet" href="../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Htseq";
        var mkdocs_page_input_path = "tasks/htseq.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../.." class="icon icon-home"> St. Jude Cloud Bioinformatics Workflows
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../build_for_dnanexus/">Building WDL workflows for DNAnexus</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Tasks</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../bwa/">Bwa</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../cellranger/">Cellranger</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../deeptools/">Deeptools</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../estimate/">Estimate</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../fastqc/">Fastqc</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../fq/">Fq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../gatk4/">Gatk4</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="#">Htseq</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#inputs">Inputs</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#required">Required</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#defaults">Defaults</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#outputs">Outputs</a>
    </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../kraken2/">Kraken2</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../librarian/">librarian</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../md5sum/">Md5sum</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../mosdepth/">Mosdepth</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../multiqc/">Multiqc</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ngsderive/">Ngsderive</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../picard/">Picard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../qualimap/">Qualimap</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../read_group/">Read group</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../sambamba/">Sambamba</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../samtools/">Samtools</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../star/">Star</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../util/">Utilities</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Workflows</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/10x-bam-to-fastqs/">Cell Ranger BAM to FASTQs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/ESTIMATE/">ESTIMATE</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/bam-to-fastqs/">Bam to fastqs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/bwa-db-build/">Bwa db build</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/dnaseq-core/">Dnaseq core</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/dnaseq-standard-fastq/">Dnaseq standard fastq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/dnaseq-standard/">Dnaseq standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/flag_filter/">FlagFilter</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/gatk-reference/">Gatk reference</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/make-qc-reference/">Make qc reference</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/markdups-post/">MarkDuplicates Post</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/quality-check-standard/">Quality check standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/rnaseq-variant-calling/">Rnaseq variant calling</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/samtools-merge/">Samtools merge</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/scrnaseq-standard/">Scrnaseq standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/star-db-build/">Star db build</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../..">St. Jude Cloud Bioinformatics Workflows</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../.." class="icon icon-home" aria-label="Docs"></a></li>
          <li class="breadcrumb-item">Tasks</li>
      <li class="breadcrumb-item active">Htseq</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/stjudecloud/workflows/edit/master/docs/tasks/htseq.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <p><a href="https://github.com/htseq/htseq">Homepage</a></p>
<h2 id="count">count</h2>
<dl>
<dt>description</dt>
<dd>Performs read counting for a set of features in the input BAM file</dd>
<dt>outputs</dt>
<dd>{'feature_counts': 'A two column headerless TSV file. First column is feature names and second column is counts.'}</dd>
</dl>
<h3 id="inputs">Inputs</h3>
<h4 id="required">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file to generate coverage for</li>
<li><code>gtf</code> (File, <strong>required</strong>): Input genomic features in gzipped GTF format to count reads for</li>
<li><code>strandedness</code> (String, <strong>required</strong>); <strong>description</strong>: Strandedness protocol of the RNA-Seq experiment; <strong>external_help</strong>: https://htseq.readthedocs.io/en/latest/htseqcount.html#cmdoption-htseq-count-s; <strong>choices</strong>: ['yes', 'reverse', 'no']</li>
</ul>
<h4 id="defaults">Defaults</h4>
<ul>
<li><code>feature_type</code> (String, default="exon"); <strong>description</strong>: Feature type (3rd column in GTF file) to be used, all features of other type are ignored; <strong>common</strong>: true</li>
<li><code>idattr</code> (String, default="gene_name"); <strong>description</strong>: GFF attribute to be used as feature ID; <strong>common</strong>: true</li>
<li><code>include_custom_header</code> (Boolean, default=false); <strong>description</strong>: Include a custom header for the output file? This is not an official feature of HTSeq. If true, the first line of the output file will be <code>feature   ~{prefix}</code>. This may break downstream tools that expect the typical headerless HTSeq output format.; <strong>common</strong>: true</li>
<li><code>minaqual</code> (Int, default=10); <strong>description</strong>: Skip all reads with alignment quality lower than the given minimum value; <strong>common</strong>: true</li>
<li><code>mode</code> (String, default="union"); <strong>description</strong>: Mode to handle reads overlapping more than one feature. <code>union</code> is recommended for most use-cases.; <strong>external_help</strong>: https://htseq.readthedocs.io/en/latest/htseqcount.html#htseq-count-counting-reads-within-features; <strong>choices</strong>: ['union', 'intersection-strict', 'intersection-nonempty']</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>modify_memory_gb</code> (Int, default=0): Add to or subtract from dynamic memory allocation. Default memory is determined by the size of the inputs. Specified in GB.</li>
<li><code>nonunique</code> (Boolean, default=false); <strong>description</strong>: Score reads that align to or are assigned to more than one feature?; <strong>common</strong>: true</li>
<li><code>pos_sorted</code> (Boolean, default=true); <strong>description</strong>: Is the BAM position sorted (true) or name sorted (false)?; <strong>common</strong>: true</li>
<li><code>prefix</code> (String, default=basename(bam,".bam")): Prefix for the feature counts file. The extension <code>.feature-counts.txt</code> will be added.</li>
<li><code>secondary_alignments</code> (Boolean, default=false); <strong>description</strong>: Score secondary alignments (SAM flag 0x100)?; <strong>common</strong>: true</li>
<li><code>supplementary_alignments</code> (Boolean, default=false); <strong>description</strong>: Score supplementary/chimeric alignments (SAM flag 0x800)?; <strong>common</strong>: true</li>
</ul>
<h3 id="outputs">Outputs</h3>
<ul>
<li><code>feature_counts</code> (File)</li>
</ul>
<h2 id="calc_tpm">calc_tpm</h2>
<dl>
<dt>description</dt>
<dd>Given a gene counts file and a gene lengths file, calculate Transcripts Per Million (TPM)</dd>
<dt>outputs</dt>
<dd>{'tpm_file': 'Transcripts Per Million (TPM) file. A two column headered TSV file.'}</dd>
</dl>
<h3 id="inputs_1">Inputs</h3>
<h4 id="required_1">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>counts</code> (File, <strong>required</strong>): A two column headerless TSV file with gene names in the first column and counts (as integers) in the second column. Entries starting with '__' will be discarded. Can be generated with the <code>count</code> task.</li>
<li><code>gene_lengths</code> (File, <strong>required</strong>): A two column headered TSV file with gene names (matching those in the <code>counts</code> file) in the first column and feature lengths (as integers) in the second column. Can be generated with the <code>calc_gene_lengths</code> task in <code>util.wdl</code>.</li>
</ul>
<h4 id="defaults_1">Defaults</h4>
<ul>
<li><code>prefix</code> (String, default=basename(counts,".feature-counts.txt")): Prefix for the TPM file. The extension <code>.TPM.txt</code> will be added.</li>
</ul>
<h3 id="outputs_1">Outputs</h3>
<ul>
<li><code>tpm_file</code> (File)</li>
</ul>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../gatk4/" class="btn btn-neutral float-left" title="Gatk4"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../kraken2/" class="btn btn-neutral float-right" title="Kraken2">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
      <p>© 2020-Present St. Jude Children's Research Hospital</p>
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/stjudecloud/workflows" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../gatk4/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../kraken2/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "../..";</script>
    <script src="../../js/theme_extra.js"></script>
    <script src="../../js/theme.js"></script>
      <script src="../../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
