<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="author" content="St. Jude Cloud" />
      <link rel="shortcut icon" href="../../img/favicon.ico" />
    <title>Arriba - St. Jude Cloud Bioinformatics Workflows</title>
    <link rel="stylesheet" href="../../css/theme.css" />
    <link rel="stylesheet" href="../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Arriba";
        var mkdocs_page_input_path = "tasks/arriba.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../.." class="icon icon-home"> St. Jude Cloud Bioinformatics Workflows
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../build_for_dnanexus/">Building WDL workflows for DNAnexus</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Tasks</span></p>
              <ul class="current">
                  <li class="toctree-l1 current"><a class="reference internal current" href="#">Arriba</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#inputs">Inputs</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#required">Required</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#optional">Optional</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#defaults">Defaults</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#outputs">Outputs</a>
    </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../bwa/">Bwa</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../deeptools/">Deeptools</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../estimate/">Estimate</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../fastqc/">Fastqc</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../fq/">Fq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../gatk4/">Gatk4</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../htseq/">Htseq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../kraken2/">Kraken2</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../librarian/">librarian</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../md5sum/">Md5sum</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../mosdepth/">Mosdepth</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../multiqc/">Multiqc</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ngsderive/">Ngsderive</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../picard/">Picard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../qualimap/">Qualimap</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../read_group/">Read group</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../sambamba/">Sambamba</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../samtools/">Samtools</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../star/">Star</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../util/">Utilities</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Workflows</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/ESTIMATE/">ESTIMATE</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/bam-to-fastqs/">Bam to fastqs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/bwa-db-build/">Bwa db build</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/dnaseq-core/">Dnaseq core</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/dnaseq-standard-fastq/">Dnaseq standard fastq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/dnaseq-standard/">Dnaseq standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/flag_filter/">FlagFilter</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/gatk-reference/">Gatk reference</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/make-qc-reference/">Make qc reference</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/markdups-post/">MarkDuplicates Post</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/quality-check-standard/">Quality check standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/rnaseq-variant-calling/">Rnaseq variant calling</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/samtools-merge/">Samtools merge</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/star-db-build/">Star db build</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../..">St. Jude Cloud Bioinformatics Workflows</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../.." class="icon icon-home" aria-label="Docs"></a></li>
          <li class="breadcrumb-item">Tasks</li>
      <li class="breadcrumb-item active">Arriba</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/stjudecloud/workflows/edit/master/docs/tasks/arriba.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <p><a href="https://arriba.readthedocs.io/en/latest/">Homepage</a></p>
<h2 id="arriba">arriba</h2>
<dl>
<dt>description</dt>
<dd>Run Arriba structural variant caller on a RNA-Seq BAM file.</dd>
<dt>help</dt>
<dd>Typical input is a STAR-aligned BAM. Arriba also supports DRAGEN-aligned BAMs and any spec compliant BAM. That is discordant mates must have <code>BAM_FPROPER_PAIR (0x2)</code>, split reads must have <code>BAM_FSUPPLEMENTARY (0x800)</code>, and the anchor read must have a <code>SA</code> tag. Arriba also uses the <code>HI</code> tag to group supplementary alignments.</dd>
<dt>outputs</dt>
<dd>{'fusions': 'Output file of fusions in TSV format', 'discarded_fusions': 'Output file of discarded fusions in TSV format'}</dd>
</dl>
<h3 id="inputs">Inputs</h3>
<h4 id="required">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file from which to call fusions</li>
<li><code>gtf</code> (File, <strong>required</strong>): GTF features file. Gzipped or uncompressed.</li>
<li><code>reference_fasta_gz</code> (File, <strong>required</strong>): Gzipped reference genome in FASTA format</li>
</ul>
<h4 id="optional">Optional</h4>
<ul>
<li><code>annotate_fusions</code> (File?); <strong>description</strong>: Optional input file in tab delimited format of fusions to annotate with tags; <strong>external_help</strong>: https://arriba.readthedocs.io/en/v2.4.0/input-files/#tags</li>
<li><code>chimeric_sam</code> (File?): Optional input file of chimeric reads in SAM format, from older versions of STAR</li>
<li><code>exclude_list</code> (File?); <strong>description</strong>: Optional input file of regions to exclude from analysis in tab delimited format; <strong>external_help</strong>: https://arriba.readthedocs.io/en/v2.4.0/input-files/#blacklist</li>
<li><code>known_fusions</code> (File?); <strong>description</strong>: Optional input file of known fusions in tab delimited format; <strong>external_help</strong>: https://arriba.readthedocs.io/en/v2.4.0/input-files/#known-fusions</li>
<li><code>protein_domains</code> (File?); <strong>description</strong>: Optional input file of protein domains coordinates in GFF3 format; <strong>external_help</strong>: https://arriba.readthedocs.io/en/v2.4.0/input-files/#protein-domains</li>
<li><code>wgs_svs</code> (File?); <strong>description</strong>: Optional input file of structural variants found by WGS in tab delimited or VCF format; <strong>external_help</strong>: https://arriba.readthedocs.io/en/v2.4.0/input-files/#structural-variant-calls-from-wgs</li>
</ul>
<h4 id="defaults">Defaults</h4>
<ul>
<li><code>coverage_fraction</code> (Float, default=0.05): Minimum fraction of viral contig transcription.</li>
<li><code>disable_filters</code> (Array[String], default=[]); <strong>description</strong>: Array of filters to disable.; <strong>choices</strong>: ['top_expressed_viral_contigs', 'viral_contigs', 'low_coverage_viral_contigs', 'uninteresting_contigs', 'no_genomic_support', 'short_anchor', 'select_best', 'many_spliced', 'long_gap', 'merge_adjacent', 'hairpin', 'small_insert_size', 'same_gene', 'genomic_support', 'read_through', 'no_coverage', 'mismatches', 'homopolymer', 'low_entropy', 'multimappers', 'inconsistently_clipped', 'duplicates', 'homologs', 'blacklist', 'mismappers', 'spliced', 'relative_support', 'min_support', 'known_fusions', 'end_to_end', 'non_coding_neighbors', 'isoforms', 'intronic', 'in_vitro', 'intragenic_exonic', 'internal_tandem_duplication']</li>
<li><code>exonic_fraction</code> (Float, default=0.33): Minimum fraction of exonic sequence between breakpoints.</li>
<li><code>feature_name</code> (String, default="gene_name=gene_name|gene_id,gene_id=gene_id,transcript_id=transcript_id,feature_exon=exon,feature_CDS=CDS"): The Arriba default it designed to handle RefSeq, GENCODE, or ENSEMBL format annotations. <code>feature_name</code> expects a string of space/comma separated options. The required fields are <code>gene_name</code>, <code>gene_id</code>, <code>transcript_id</code>, <code>feature_exon</code>, and <code>feature_CDS</code>. The fields should space separated. The values should be provided with <code>field=value</code>. Mutliple values can be provided and separated by a pipe (<code>|</code>), e.g. <code>=value1|value2</code>. A complete example is <code>gene_name=gene_name|gene_id gene_id=gene_id transcript_id=transcript_id feature_exon=exon feature_CDS=CDS</code>.; <strong>description</strong>: List of feature names to use in GTF.; <strong>external_help</strong>: https://arriba.readthedocs.io/en/v2.4.0/command-line-options/; <strong>common</strong>: false</li>
<li><code>fill_gaps</code> (Boolean, default=false): Fill gaps in assembled transcripts with reference bases. Expands the fusion sequence to the complete sequence of the fusion gene.</li>
<li><code>fragment_length</code> (Int, default=200): For single-end data, this is the fragment length. With paired-end reads, this is ignored and determined automatically.</li>
<li><code>homopolymer_length</code> (Int, default=6): Maximum homopolymer length adjacent to breakpoints.</li>
<li><code>interesting_contigs</code> (Array[String], default=["1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21", "22", "X", "Y", "AC_<em>", "NC_</em>"]): Array of contigs to consider for analysis. Contigs can be specified with or without the prefix <code>chr</code>.</li>
<li><code>many_spliced_events</code> (Int, default=4): Recover fusions with at least this many spliced breakpoints.</li>
<li><code>mark_duplicates</code> (Boolean, default=true): Arriba performs marking of duplicates internally based on identical mapping coordinates. When this switch is set, internal marking of duplicates is disabled and Arriba assumes that duplicates have been marked by a preceding program. In this case, Arriba only discards alignments flagged with the BAM_FDUP flag. This makes sense when duplicates cannot be reliably identified solely based on their mapping coordinates, e.g. when unique molecular identifiers (UMIs) are used or when independently generated libraries are merged in a single BAM file and the read group must be interrogated to distinguish duplicates from reads that map to the same coordinates by chance. In addition, when this switch is set, duplicate reads are not considered for the calculation of the coverage at fusion breakpoints (columns coverage1 and coverage2 in the output file).; <strong>description</strong>: Mark duplicates in the input BAM file with Arriba.</li>
<li><code>max_e_value</code> (Float, default=0.3): Maximum E-value for read support.</li>
<li><code>max_genomic_breakpoint_distance</code> (Int, default=1000000): With 'wgs_svs', threshold for relating genomic and transcriptomic events.</li>
<li><code>max_homolog_identity</code> (Float, default=0.3): Maximum fraction of homologous sequence for genes.</li>
<li><code>max_itd_length</code> (Int, default=100): Maximum length of internal tandem duplications.</li>
<li><code>max_kmer_content</code> (Float, default=0.6): Maximum fraction of repetitive 3-mer content in the fusion region.</li>
<li><code>max_mismappers</code> (Float, default=0.8): Maximum fraction of mismapped reads in the fusion region.</li>
<li><code>max_mismatch_pvalue</code> (Float, default=0.01): Maximum p-value for mismatches in the fusion region.</li>
<li><code>max_reads</code> (Int, default=300): Subsample fusions with more than this number of reads.</li>
<li><code>min_anchor_length</code> (Int, default=23): Minimum anchor length for split reads.</li>
<li><code>min_itd_allele_fraction</code> (Float, default=0.07): Minimum supporting read fraction for internal tandem duplications.</li>
<li><code>min_itd_supporting_reads</code> (Int, default=10): Minimum number of supporting reads for internal tandem duplications.</li>
<li><code>min_supporting_reads</code> (Int, default=2): Minimum number of supporting reads for a fusion.</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>modify_memory_gb</code> (Int, default=0): Add to or subtract from dynamic memory allocation. Default memory is determined by the size of the inputs. Specified in GB.</li>
<li><code>prefix</code> (String, default=basename(bam,".bam") + ".fusions"): Prefix for the fusion result files. The extensions <code>.tsv</code> and <code>.discarded.tsv</code> will be added.</li>
<li><code>quantile</code> (Float, default=0.998): Genes with expression above the given quantile are eligible for filtering.</li>
<li><code>read_through_distance</code> (Int, default=10000): Minimum distance between breakpoints for read-through events.</li>
<li><code>report_additional_columns</code> (Boolean, default=false): Report additional columns ['fusion_transcript', 'peptide_sequence', 'read_identifiers'] in the discarded fusions file.</li>
<li><code>strandedness</code> (String, default="auto"); <strong>description</strong>: Strandedness of the input data.; <strong>external_help</strong>: https://arriba.readthedocs.io/en/v2.4.0/command-line-options/; <strong>choices</strong>: ['auto', 'yes', 'no', 'reverse']</li>
<li><code>top_n</code> (Int, default=5): Only report the top N most highly expressed viral integration sites.</li>
<li><code>viral_contigs</code> (Array[String], default=["AC_<em>", "NC_</em>"]): Array of contigs to consider for viral integration site analysis.</li>
</ul>
<h3 id="outputs">Outputs</h3>
<ul>
<li><code>fusions</code> (File)</li>
<li><code>discarded_fusions</code> (File)</li>
</ul>
<h2 id="arriba_tsv_to_vcf">arriba_tsv_to_vcf</h2>
<dl>
<dt>description</dt>
<dd>Convert Arriba TSV format fusions to VCF format.</dd>
<dt>outputs</dt>
<dd>{'fusions_vcf': 'Output file of fusions in VCF format'}</dd>
</dl>
<h3 id="inputs_1">Inputs</h3>
<h4 id="required_1">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>fusions</code> (File, <strong>required</strong>): Input fusions in TSV format to convert to VCF</li>
<li><code>reference_fasta</code> (File, <strong>required</strong>): Reference genome in FASTA format. Either gzipped or uncompressed.</li>
</ul>
<h4 id="defaults_1">Defaults</h4>
<ul>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>prefix</code> (String, default=basename(fusions,".tsv")): Output file name for fusions in VCF format. The extension <code>.vcf</code> will be appended.</li>
</ul>
<h3 id="outputs_1">Outputs</h3>
<ul>
<li><code>fusions_vcf</code> (File)</li>
</ul>
<h2 id="arriba_extract_fusion_supporting_alignments">arriba_extract_fusion_supporting_alignments</h2>
<dl>
<dt>description</dt>
<dd>Extract alignments that support fusions.</dd>
<dt>outputs</dt>
<dd>{'fusion_bams': 'Array of BAM files corresponding with fusions in the input file', 'fusion_bam_indexes': "Array of BAM indexes corresponding with the BAMs in the 'fusion_bams'"}</dd>
</dl>
<h3 id="inputs_2">Inputs</h3>
<h4 id="required_2">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file from which fusions were called</li>
<li><code>bam_index</code> (File, <strong>required</strong>): BAM index file corresponding to the input BAM</li>
<li><code>fusions</code> (File, <strong>required</strong>): Input fusions in TSV format for which to extract supporting alignments</li>
</ul>
<h4 id="defaults_2">Defaults</h4>
<ul>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>prefix</code> (String, default=basename(fusions,".tsv")): Output file name prefix for the extracted BAM files. The extension <code>.bam</code> will be appended.</li>
</ul>
<h3 id="outputs_2">Outputs</h3>
<ul>
<li><code>fusion_bams</code> (Array[File])</li>
<li><code>fusion_bam_indexes</code> (Array[File])</li>
</ul>
<h2 id="arriba_annotate_exon_numbers">arriba_annotate_exon_numbers</h2>
<dl>
<dt>description</dt>
<dd>Annotate fusions with exon numbers.</dd>
<dt>outputs</dt>
<dd>{'fusion_tsv': 'TSV file with fusions annotated with exon numbers'}</dd>
</dl>
<h3 id="inputs_3">Inputs</h3>
<h4 id="required_3">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>fusions</code> (File, <strong>required</strong>): Input fusions in TSV format for which to annotate gene exon numbers</li>
<li><code>gtf</code> (File, <strong>required</strong>): GTF features file. Gzipped or uncompressed.</li>
</ul>
<h4 id="defaults_3">Defaults</h4>
<ul>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>prefix</code> (String, default=basename(fusions,".tsv")): Output file name for annotated fusions in TSV format. The extension <code>.annotated.tsv</code> will be appended.</li>
</ul>
<h3 id="outputs_3">Outputs</h3>
<ul>
<li><code>fusion_tsv</code> (File)</li>
</ul>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../../build_for_dnanexus/" class="btn btn-neutral float-left" title="Building WDL workflows for DNAnexus"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../bwa/" class="btn btn-neutral float-right" title="Bwa">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
      <p>Â© 2020-Present St. Jude Children's Research Hospital</p>
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/stjudecloud/workflows" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../../build_for_dnanexus/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../bwa/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "../..";</script>
    <script src="../../js/theme_extra.js"></script>
    <script src="../../js/theme.js"></script>
      <script src="../../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
