version 1.1

import "../../data_structures/read_group.wdl"
import "../../tools/fq.wdl"
import "./rnaseq-core.wdl" as rnaseq_core_wf
import "./rnaseq-standard.wdl" as rnaseq_standard

workflow rnaseq_standard_fastq {
    meta {
        name: "RNA-Seq Standard (FASTQ)"
        description: "Runs the STAR RNA-Seq alignment workflow for St. Jude Cloud from FASTQ input"
        category: "Harmonization"
        outputs: {
            bam: "Harmonized RNA-Seq BAM",
            bam_index: "BAI index file associated with `bam`",
            bam_checksum: "STDOUT of the `md5sum` command run on the harmonized BAM that has been redirected to a file",
            star_log: "Summary mapping statistics after mapping job is complete",
            bigwig: "BigWig format coverage file generated from `bam`",
            feature_counts: "A two column headerless TSV file. First column is feature names and second column is counts.",
            inferred_strandedness: "TSV file containing the `ngsderive strandedness` report",
            inferred_strandedness_string: "Derived strandedness from `ngsderive strandedness`",
            fastp_reports: "An array of `fastp` reports (in HTML format) corresponding to each read group",
            fastp_jsons: "An array of `fastp` reports (in JSON format) corresponding to each read group",
        }
        allowNestedInputs: true
    }

    parameter_meta {
        gtf: "Gzipped GTF feature file"
        star_db: "Database of reference files for the STAR aligner. Can be generated by `star-db-build.wdl`"
        read_one_fastqs_gz: "Input gzipped FASTQ format file(s) with 1st read in pair to align"
        read_two_fastqs_gz: "Input gzipped FASTQ format file(s) with 2nd read in pair to align"
        read_groups: {
            description: "An Array of structs defining read groups to include in the harmonized BAM",
            help: "Must correspond to input FASTQs. Each read group ID must be contained in the basename of a FASTQ file or pair of FASTQ files if Paired-End. This requirement means the length of `read_groups` must equal the length of `read_one_fastqs_gz` and the length of `read_two_fastqs_gz` if non-zero. Only the `ID` field is required, and it must be unique for each read group defined.",
            warning: "See `data_structures/read_group.wdl` for more information and help formatting your input JSON.",
        }
        contaminant_db: "A compressed reference database corresponding to the aligner chosen with `xenocp_aligner` for the contaminant genome"
        prefix: {
            description: "Prefix for output files",
            help: "See `../../README.md` for more information on the default prefix evaluation.",
            group: "Common",
        }
        xenocp_aligner: {
            description: "Aligner to use to map reads to the host genome for detecting contamination",
            choices: [
                "bwa aln",
                "bwa mem",
                "star",
            ],
        }
        strandedness: {
            description: "Strandedness protocol of the RNA-Seq experiment. If unspecified, strandedness will be inferred by `ngsderive`.",
            choices: [
                "",
                "Stranded-Reverse",
                "Stranded-Forward",
                "Unstranded",
            ],
        }
        enable_read_trimming: "Enable read trimming with `fastp`?"
        mark_duplicates: "Add SAM flag to computationally determined duplicate reads?"
        cleanse_xenograft: "Use XenoCP to unmap reads from contaminant genome?"
        validate_input: "Ensure input FASTQs are well-formed before beginning harmonization?"
        use_all_cores: "Use all cores for multi-core steps?"
        subsample_n_reads: "Only process a random sampling of `n` reads. Any `n`<=`0` for processing entire input."
    }

    input {
        File gtf
        File star_db
        Array[File] read_one_fastqs_gz
        Array[File] read_two_fastqs_gz
        Array[ReadGroup] read_groups
        File? contaminant_db
        String prefix = sub(
            basename(read_one_fastqs_gz[0]),
            "(([_.][rR](?:ead)?[12])((?:[_.-][^_.-]*?)*?))?\\.(fastq|fq)(\\.gz)?$",
            ""  # Once replacing with capturing groups is supported, replace with group 3
        )
        String xenocp_aligner = "star"
        String strandedness = ""
        Boolean enable_read_trimming = false
        Boolean mark_duplicates = false
        Boolean cleanse_xenograft = false
        Boolean validate_input = true
        Boolean use_all_cores = false
        Int subsample_n_reads = -1
    }

    call rnaseq_standard.parse_input { input:
        strand = strandedness,
        cleanse_xenograft,
        contaminant_db = defined(contaminant_db),
    }

    scatter (rg in read_groups) {
        call read_group.read_group_to_string after parse_input { input:
            read_group = rg,
        }
    }

    if (validate_input){
        scatter (reads in zip(read_one_fastqs_gz, read_two_fastqs_gz)) {
            call fq.fqlint after parse_input { input:
                read_one_fastq = reads.left,
                read_two_fastq = reads.right,
            }
        }
    }

    if (subsample_n_reads > 0) {
        Int reads_per_pair = ceil(subsample_n_reads / length(read_one_fastqs_gz))
        scatter (reads in zip(read_one_fastqs_gz, read_two_fastqs_gz)) {
            call fq.subsample after fqlint { input:
                read_one_fastq = reads.left,
                read_two_fastq = reads.right,
                record_count = reads_per_pair,
            }
        }
    }
    Array[File] selected_read_one_fastqs = select_first([
        subsample.subsampled_read1,
        read_one_fastqs_gz,
    ])
    Array[File] selected_read_two_fastqs = select_all(
        select_first([
            subsample.subsampled_read2,
            read_two_fastqs_gz,
        ])
    )

    call rnaseq_core_wf.rnaseq_core after fqlint { input:
        read_one_fastqs_gz = selected_read_one_fastqs,
        read_two_fastqs_gz = selected_read_two_fastqs,
        read_groups = read_group_to_string.validated_read_group,
        prefix,
        gtf,
        star_db,
        enable_read_trimming,
        mark_duplicates,
        contaminant_db,
        cleanse_xenograft,
        xenocp_aligner,
        strandedness,
        use_all_cores,
    }

    output {
        File bam = rnaseq_core.bam
        File bam_index = rnaseq_core.bam_index
        File bam_checksum = rnaseq_core.bam_checksum
        File star_log = rnaseq_core.star_log
        File bigwig = rnaseq_core.bigwig
        File feature_counts = rnaseq_core.feature_counts
        File inferred_strandedness = rnaseq_core.inferred_strandedness
        String inferred_strandedness_string = rnaseq_core.inferred_strandedness_string
        Array[File] fastp_reports = rnaseq_core.fastp_reports
        Array[File] fastp_jsons = rnaseq_core.fastp_jsons
    }
}
