<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="author" content="St. Jude Cloud" />
      <link rel="shortcut icon" href="../../img/favicon.ico" />
    <title>Gatk4 - St. Jude Cloud Bioinformatics Workflows</title>
    <link rel="stylesheet" href="../../css/theme.css" />
    <link rel="stylesheet" href="../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Gatk4";
        var mkdocs_page_input_path = "tasks/gatk4.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../.." class="icon icon-home"> St. Jude Cloud Bioinformatics Workflows
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../build_for_dnanexus/">Building WDL workflows for DNAnexus</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Tasks</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../arriba/">Arriba</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../bwa/">Bwa</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../cellranger/">Cellranger</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../deeptools/">Deeptools</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../estimate/">Estimate</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../fastqc/">Fastqc</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../fq/">Fq</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="#">Gatk4</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#inputs">Inputs</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#required">Required</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#defaults">Defaults</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#outputs">Outputs</a>
    </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../htseq/">Htseq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../kraken2/">Kraken2</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../librarian/">librarian</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../md5sum/">Md5sum</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../mosdepth/">Mosdepth</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../multiqc/">Multiqc</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ngsderive/">Ngsderive</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../picard/">Picard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../qualimap/">Qualimap</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../read_group/">Read group</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../sambamba/">Sambamba</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../samtools/">Samtools</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../star/">Star</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../util/">Utilities</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Workflows</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/10x-bam-to-fastqs/">Cell Ranger BAM to FASTQs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/ESTIMATE/">ESTIMATE</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/bam-to-fastqs/">Bam to fastqs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/bwa-db-build/">Bwa db build</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/dnaseq-core/">Dnaseq core</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/dnaseq-standard-fastq/">Dnaseq standard fastq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/dnaseq-standard/">Dnaseq standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/flag_filter/">FlagFilter</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/gatk-reference/">Gatk reference</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/make-qc-reference/">Make qc reference</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/markdups-post/">MarkDuplicates Post</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/quality-check-standard/">Quality check standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/rnaseq-variant-calling/">Rnaseq variant calling</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/samtools-merge/">Samtools merge</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/scrnaseq-standard/">Scrnaseq standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/star-db-build/">Star db build</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../..">St. Jude Cloud Bioinformatics Workflows</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../.." class="icon icon-home" aria-label="Docs"></a></li>
          <li class="breadcrumb-item">Tasks</li>
      <li class="breadcrumb-item active">Gatk4</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/stjudecloud/workflows/edit/master/docs/tasks/gatk4.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <p><a href="https://software.broadinstitute.org/gatk">Homepage</a></p>
<h2 id="split_n_cigar_reads">split_n_cigar_reads</h2>
<dl>
<dt>description</dt>
<dd>Splits reads that contain Ns in their CIGAR strings into multiple reads.</dd>
<dt>external_help</dt>
<dd>https://gatk.broadinstitute.org/hc/en-us/articles/360036858811-SplitNCigarReads</dd>
<dt>outputs</dt>
<dd>{'split_n_reads_bam': 'BAM file with reads split at N CIGAR elements and updated CIGAR strings.', 'split_n_reads_bam_index': 'Index file for the split BAM', 'split_n_reads_bam_md5': 'MD5 checksum for the split BAM'}</dd>
</dl>
<h3 id="inputs">Inputs</h3>
<h4 id="required">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file to with unsplit reads containing Ns in their CIGAR strings.</li>
<li><code>bam_index</code> (File, <strong>required</strong>): BAM index file corresponding to the input BAM</li>
<li><code>dict</code> (File, <strong>required</strong>): Dictionary file for FASTA format genome</li>
<li><code>fasta</code> (File, <strong>required</strong>): Reference genome in FASTA format. Must be uncompressed.</li>
<li><code>fasta_index</code> (File, <strong>required</strong>): Index for FASTA format genome</li>
<li><code>interval_list</code> (File, <strong>required</strong>): Interval list indicating regions in which to split reads</li>
</ul>
<h4 id="defaults">Defaults</h4>
<ul>
<li><code>memory_gb</code> (Int, default=25): RAM to allocate for task, specified in GB</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>ncpu</code> (Int, default=8): Number of cores to allocate for task</li>
<li><code>prefix</code> (String, default=basename(bam,".bam") + ".split"): Prefix for the BAM file. The extension <code>.bam</code> will be added.</li>
</ul>
<h3 id="outputs">Outputs</h3>
<ul>
<li><code>split_n_reads_bam</code> (File)</li>
<li><code>split_n_reads_bam_index</code> (File)</li>
<li><code>split_n_reads_bam_md5</code> (File)</li>
</ul>
<h2 id="base_recalibrator">base_recalibrator</h2>
<dl>
<dt>description</dt>
<dd>Generates recalibration report for base quality score recalibration.</dd>
<dt>external_help</dt>
<dd>https://gatk.broadinstitute.org/hc/en-us/articles/360036897372-BaseRecalibratorSpark-BETA</dd>
<dt>outputs</dt>
<dd>{'recalibration_report': 'Recalibration report file'}</dd>
</dl>
<h3 id="inputs_1">Inputs</h3>
<h4 id="required_1">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file on which to recabilbrate base quality scores</li>
<li><code>bam_index</code> (File, <strong>required</strong>): BAM index file corresponding to the input BAM</li>
<li><code>dbSNP_vcf</code> (File, <strong>required</strong>): dbSNP VCF file</li>
<li><code>dbSNP_vcf_index</code> (File, <strong>required</strong>): dbSNP VCF index file</li>
<li><code>dict</code> (File, <strong>required</strong>): Dictionary file for FASTA format genome</li>
<li><code>fasta</code> (File, <strong>required</strong>): Reference genome in FASTA format</li>
<li><code>fasta_index</code> (File, <strong>required</strong>): Index for FASTA format genome</li>
<li><code>known_indels_sites_indices</code> (Array[File], <strong>required</strong>): List of VCF index files corresponding to the VCF files in <code>known_indels_sites_vcfs</code></li>
<li><code>known_indels_sites_vcfs</code> (Array[File], <strong>required</strong>): List of VCF files containing known indels</li>
</ul>
<h4 id="defaults_1">Defaults</h4>
<ul>
<li><code>memory_gb</code> (Int, default=25): RAM to allocate for task, specified in GB</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>ncpu</code> (Int, default=4): Number of cores to allocate for task</li>
<li><code>outfile_name</code> (String, default=basename(bam,".bam") + ".recal.txt"): Name for the output recalibration report.</li>
<li><code>use_original_quality_scores</code> (Boolean, default=false): Use original quality scores from the input BAM. Default is to use recalibrated quality scores.</li>
</ul>
<h3 id="outputs_1">Outputs</h3>
<ul>
<li><code>recalibration_report</code> (File)</li>
</ul>
<h2 id="apply_bqsr">apply_bqsr</h2>
<dl>
<dt>description</dt>
<dd>Applies base quality score recalibration to a BAM file.</dd>
<dt>external_help</dt>
<dd>https://gatk.broadinstitute.org/hc/en-us/articles/360040097972-ApplyBQSRSpark-BETA</dd>
<dt>outputs</dt>
<dd>{'recalibrated_bam': 'Recalibrated BAM file', 'recalibrated_bam_index': 'Index file for the recalibrated BAM'}</dd>
</dl>
<h3 id="inputs_2">Inputs</h3>
<h4 id="required_2">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file on which to apply base quality score recalibration</li>
<li><code>bam_index</code> (File, <strong>required</strong>): BAM index file corresponding to the input BAM</li>
<li><code>recalibration_report</code> (File, <strong>required</strong>): Recalibration report file</li>
</ul>
<h4 id="defaults_2">Defaults</h4>
<ul>
<li><code>memory_gb</code> (Int, default=25): RAM to allocate for task, specified in GB</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>ncpu</code> (Int, default=4): Number of cores to allocate for task</li>
<li><code>prefix</code> (String, default=basename(bam,".bam")): Prefix for the output recalibrated BAM. The extension <code>.bqsr.bam</code> will be added.</li>
<li><code>use_original_quality_scores</code> (Boolean, default=false): Use original quality scores from the input BAM. Default is to use recalibrated quality scores.</li>
</ul>
<h3 id="outputs_2">Outputs</h3>
<ul>
<li><code>recalibrated_bam</code> (File)</li>
<li><code>recalibrated_bam_index</code> (File)</li>
</ul>
<h2 id="haplotype_caller">haplotype_caller</h2>
<dl>
<dt>description</dt>
<dd>Calls germline SNPs and indels via local re-assembly of haplotypes.</dd>
<dt>external_help</dt>
<dd>https://gatk.broadinstitute.org/hc/en-us/articles/360037225632-HaplotypeCaller</dd>
<dt>outputs</dt>
<dd>{'vcf': 'VCF file containing called variants', 'vcf_index': 'Index file for the VCF'}</dd>
</dl>
<h3 id="inputs_3">Inputs</h3>
<h4 id="required_3">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file on which to call variants</li>
<li><code>bam_index</code> (File, <strong>required</strong>): BAM index file corresponding to the input BAM</li>
<li><code>dbSNP_vcf</code> (File, <strong>required</strong>): dbSNP VCF file</li>
<li><code>dbSNP_vcf_index</code> (File, <strong>required</strong>): dbSNP VCF index file</li>
<li><code>dict</code> (File, <strong>required</strong>): Dictionary file for FASTA format genome</li>
<li><code>fasta</code> (File, <strong>required</strong>): Reference genome in FASTA format</li>
<li><code>fasta_index</code> (File, <strong>required</strong>): Index for FASTA format genome</li>
<li><code>interval_list</code> (File, <strong>required</strong>); <strong>description</strong>: Interval list indicating regions in which to call variants; <strong>external_help</strong>: https://gatk.broadinstitute.org/hc/en-us/articles/360035531852-Intervals-and-interval-lists</li>
</ul>
<h4 id="defaults_3">Defaults</h4>
<ul>
<li><code>memory_gb</code> (Int, default=25): RAM to allocate for task, specified in GB</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic memory allocation. Default memory is determined by the size of the inputs. Specified in GB.</li>
<li><code>ncpu</code> (Int, default=4): Number of cores to allocate for task</li>
<li><code>prefix</code> (String, default=basename(bam,".bam")): Prefix for the output VCF. The extension <code>.vcf.gz</code> will be added.</li>
<li><code>stand_call_conf</code> (Int, default=20); <strong>description</strong>: Minimum confidence threshold for calling variants; <strong>external_help</strong>: https://gatk.broadinstitute.org/hc/en-us/articles/360037225632-HaplotypeCaller#--standard-min-confidence-threshold-for-calling</li>
<li><code>use_soft_clipped_bases</code> (Boolean, default=false): Use soft clipped bases in variant calling. Default is to ignore soft clipped bases.</li>
</ul>
<h3 id="outputs_3">Outputs</h3>
<ul>
<li><code>vcf</code> (File)</li>
<li><code>vcf_index</code> (File)</li>
</ul>
<h2 id="variant_filtration">variant_filtration</h2>
<dl>
<dt>description</dt>
<dd>Filters variants based on specified criteria.</dd>
<dt>external_help</dt>
<dd>https://gatk.broadinstitute.org/hc/en-us/articles/360037434691-VariantFiltration</dd>
<dt>outputs</dt>
<dd>{'vcf_filtered': 'Filtered VCF file', 'vcf_filtered_index': 'Index file for the filtered VCF'}</dd>
</dl>
<h3 id="inputs_4">Inputs</h3>
<h4 id="required_4">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>dict</code> (File, <strong>required</strong>): Dictionary file for FASTA format genome</li>
<li><code>fasta</code> (File, <strong>required</strong>): Reference genome in FASTA format</li>
<li><code>fasta_index</code> (File, <strong>required</strong>): Index for FASTA format genome</li>
<li><code>vcf</code> (File, <strong>required</strong>): Input VCF format file to filter</li>
<li><code>vcf_index</code> (File, <strong>required</strong>): VCF index file corresponding to the input VCF</li>
</ul>
<h4 id="defaults_4">Defaults</h4>
<ul>
<li><code>cluster</code> (Int, default=3): Number of SNPs that must be present in a window to filter</li>
<li><code>filter_expressions</code> (Array[String], default=["FS &gt; 30.0", "QD &lt; 2.0"]); <strong>description</strong>: Expressions for the filters; <strong>external_help</strong>: https://gatk.broadinstitute.org/hc/en-us/articles/360037434691-VariantFiltration#--filter-expression</li>
<li><code>filter_names</code> (Array[String], default=["FS", "QD"]); <strong>description</strong>: Names of the filters to apply; <strong>external_help</strong>: https://gatk.broadinstitute.org/hc/en-us/articles/360037434691-VariantFiltration#--filter-name</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic memory allocation. Default memory is determined by the size of the inputs. Specified in GB.</li>
<li><code>ncpu</code> (Int, default=1): Number of cores to allocate for task</li>
<li><code>prefix</code> (String, default=basename(vcf,".vcf.gz")): Prefix for the output filtered VCF. The extension <code>.filtered.vcf.gz</code> will be added.</li>
<li><code>window</code> (Int, default=35): Size of the window (in bases) for filtering</li>
</ul>
<h3 id="outputs_4">Outputs</h3>
<ul>
<li><code>vcf_filtered</code> (File)</li>
<li><code>vcf_filtered_index</code> (File)</li>
</ul>
<h2 id="mark_duplicates_spark">mark_duplicates_spark</h2>
<dl>
<dt>description</dt>
<dd>Marks duplicate reads in the input BAM file using GATK's Spark implementation of Picard's MarkDuplicates.</dd>
<dt>external_help</dt>
<dd>https://gatk.broadinstitute.org/hc/en-us/articles/13832682540699-MarkDuplicatesSpark</dd>
<dt>outputs</dt>
<dd>{'duplicate_marked_bam': 'The input BAM with computationally determined duplicates marked.', 'duplicate_marked_bam_index': 'The <code>.bai</code> BAM index file associated with <code>duplicate_marked_bam</code>', 'mark_duplicates_metrics': {'description': 'The METRICS_FILE result of <code>picard MarkDuplicates</code>', 'external_help': 'http://broadinstitute.github.io/picard/picard-metric-definitions.html#DuplicationMetrics'}}</dd>
</dl>
<h3 id="inputs_5">Inputs</h3>
<h4 id="required_5">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file in which to mark duplicates</li>
</ul>
<h4 id="defaults_5">Defaults</h4>
<ul>
<li><code>create_bam</code> (Boolean, default=true); <strong>description</strong>: Enable BAM creation (true)? Or only output MarkDuplicates metrics (false)?; <strong>common</strong>: true</li>
<li><code>duplicate_scoring_strategy</code> (String, default="SUM_OF_BASE_QUALITIES"); <strong>description</strong>: Strategy for scoring duplicates.; <strong>choices</strong>: ['SUM_OF_BASE_QUALITIES', 'TOTAL_MAPPED_REFERENCE_LENGTH', 'RANDOM']</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>modify_memory_gb</code> (Int, default=0): Add to or subtract from the default memory allocation. Default memory allocation is determined by the size of the input BAM. Specified in GB.</li>
<li><code>ncpu</code> (Int, default=4): Number of cores to allocate for task</li>
<li><code>optical_distance</code> (Int, default=0): Maximum distance between read coordinates to consider them optical duplicates. If <code>0</code>, then optical duplicate marking is disabled. Suggested settings of 100 for unpatterned versions of the Illumina platform (e.g. HiSeq) or 2500 for patterned flowcell models (e.g. NovaSeq). Calculation of distance depends on coordinate data embedded in the read names, typically produced by the Illumina sequencing machines. Optical duplicate detection will not work on non-standard names without modifying <code>read_name_regex</code>.</li>
<li><code>prefix</code> (String, default=basename(bam,".bam") + ".MarkDuplicates"): Prefix for the MarkDuplicates result files. The extensions <code>.bam</code>, <code>.bam.bai</code>, and <code>.metrics.txt</code> will be added.</li>
<li><code>read_name_regex</code> (String, default="^[!-9;-?A-~:]+:([!-9;-?A-~]+):([0-9]+):([0-9]+)$"): Regular expression for extracting tile names, x coordinates, and y coordinates from read names. The default works for typical Illumina read names.</li>
<li><code>tagging_policy</code> (String, default="All"); <strong>description</strong>: Tagging policy for the output BAM.; <strong>choices</strong>: ['DontTag', 'OpticalOnly', 'All']</li>
<li><code>validation_stringency</code> (String, default="SILENT"); <strong>description</strong>: Validation stringency for parsing the input BAM.; <strong>choices</strong>: ['STRICT', 'LENIENT', 'SILENT']; <strong>tool_default</strong>: STRICT</li>
</ul>
<h3 id="outputs_5">Outputs</h3>
<ul>
<li><code>duplicate_marked_bam</code> (File?)</li>
<li><code>duplicate_marked_bam_index</code> (File?)</li>
<li><code>mark_duplicates_metrics</code> (File)</li>
</ul>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../fq/" class="btn btn-neutral float-left" title="Fq"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../htseq/" class="btn btn-neutral float-right" title="Htseq">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
      <p>© 2020-Present St. Jude Children's Research Hospital</p>
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/stjudecloud/workflows" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../fq/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../htseq/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "../..";</script>
    <script src="../../js/theme_extra.js"></script>
    <script src="../../js/theme.js"></script>
      <script src="../../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
