<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="author" content="St. Jude Cloud" />
      <link rel="shortcut icon" href="../../img/favicon.ico" />
    <title>Utilities - St. Jude Cloud Bioinformatics Workflows</title>
    <link rel="stylesheet" href="../../css/theme.css" />
    <link rel="stylesheet" href="../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Utilities";
        var mkdocs_page_input_path = "tasks/util.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../.." class="icon icon-home"> St. Jude Cloud Bioinformatics Workflows
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../build_for_dnanexus/">Building WDL workflows for DNAnexus</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Tasks</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../bwa/">Bwa</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../cellranger/">Cellranger</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../deeptools/">Deeptools</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../estimate/">Estimate</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../fastqc/">Fastqc</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../fq/">Fq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../htseq/">Htseq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../kraken2/">Kraken2</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../librarian/">librarian</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../md5sum/">Md5sum</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../mosdepth/">Mosdepth</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../multiqc/">Multiqc</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ngsderive/">Ngsderive</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../picard/">Picard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../qualimap/">Qualimap</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../samtools/">Samtools</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../star/">Star</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="./">Utilities</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#download">download</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#inputs">Inputs</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#required">Required</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#optional">Optional</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#outputs">Outputs</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#get_read_groups">get_read_groups</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#inputs_1">Inputs</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#required_1">Required</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#defaults">Defaults</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#outputs_1">Outputs</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#split_string">split_string</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#inputs_2">Inputs</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#required_2">Required</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#defaults_1">Defaults</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#outputs_2">Outputs</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#calc_gene_lengths">calc_gene_lengths</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#inputs_3">Inputs</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#required_3">Required</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#defaults_2">Defaults</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#outputs_3">Outputs</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#compression_integrity">compression_integrity</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#inputs_4">Inputs</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#required_4">Required</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#defaults_3">Defaults</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#outputs_4">Outputs</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#add_to_bam_header">add_to_bam_header</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#inputs_5">Inputs</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#required_5">Required</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#defaults_4">Defaults</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#outputs_5">Outputs</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#unpack_tarball">unpack_tarball</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#inputs_6">Inputs</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#required_6">Required</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#defaults_5">Defaults</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#outputs_6">Outputs</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#make_coverage_regions_beds">make_coverage_regions_beds</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#inputs_7">Inputs</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#required_7">Required</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#defaults_6">Defaults</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#outputs_7">Outputs</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#global_phred_scores">global_phred_scores</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#inputs_8">Inputs</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#required_8">Required</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#defaults_7">Defaults</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#outputs_8">Outputs</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#qc_summary">qc_summary</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#inputs_9">Inputs</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#required_9">Required</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#defaults_8">Defaults</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#outputs_9">Outputs</a>
    </li>
        </ul>
    </li>
    </ul>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Workflows</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/10x-bam-to-fastqs/">Cell Ranger BAM to FASTQs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/ESTIMATE/">ESTIMATE</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/bam-to-fastqs/">Bam to fastqs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/bwa-db-build/">Bwa db build</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/dnaseq-standard/">Dnaseq standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/flag_filter/">FlagFilter</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/make-qc-reference/">Make qc reference</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/markdups-post/">MarkDuplicates Post</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/quality-check-standard/">Quality check standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/samtools-merge/">Samtools merge</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/scrnaseq-standard/">Scrnaseq standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/star-db-build/">Star db build</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../..">St. Jude Cloud Bioinformatics Workflows</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../.." class="icon icon-home" aria-label="Docs"></a></li>
          <li class="breadcrumb-item">Tasks</li>
      <li class="breadcrumb-item active">Utilities</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/stjudecloud/workflows/edit/master/docs/tasks/util.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="utilities">Utilities</h1>
<h2 id="download">download</h2>
<dl>
<dt>description</dt>
<dd>Uses wget to download a file from a remote URL to the local filesystem</dd>
<dt>outputs</dt>
<dd>{'downloaded_file': 'File downloaded from provided URL'}</dd>
</dl>
<h3 id="inputs">Inputs</h3>
<h4 id="required">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>disk_size_gb</code> (Int, <strong>required</strong>): Disk space to allocate for task, specified in GB</li>
<li><code>outfile_name</code> (String, <strong>required</strong>): Name of the output file</li>
<li><code>url</code> (String, <strong>required</strong>): URL of the file to download</li>
</ul>
<h4 id="optional">Optional</h4>
<ul>
<li><code>md5sum</code> (String?): Optional md5sum to check against downloaded file. Recommended to use in order to catch corruption or an unintentional file swap.</li>
</ul>
<h3 id="outputs">Outputs</h3>
<ul>
<li><code>downloaded_file</code> (File)</li>
</ul>
<h2 id="get_read_groups">get_read_groups</h2>
<dl>
<dt>description</dt>
<dd>Gets read group information from a BAM file and writes it out to as a string</dd>
<dt>outputs</dt>
<dd>{'read_groups': 'An array of strings containing read group information. If <code>format_for_star = true</code>, all found read groups are contained in one string (<code>read_groups[0]</code>). If <code>format_for_star = false</code>, each found @RG line will be its own entry in output array <code>read_groups</code>.'}</dd>
</dl>
<h3 id="inputs_1">Inputs</h3>
<h4 id="required_1">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file to get read groups from</li>
</ul>
<h4 id="defaults">Defaults</h4>
<ul>
<li><code>format_for_star</code> (Boolean, default=true); <strong>description</strong>: Format read group information for the STAR aligner (true) or output @RG lines of the header without further processing (false)? STAR formatted results will be an array of length 1, where all found read groups are contained in one string (<code>read_groups[0]</code>). If no processing is selected, each found @RG line will be its own entry in output array <code>read_groups</code>.; <strong>common</strong>: true</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
</ul>
<h3 id="outputs_1">Outputs</h3>
<ul>
<li><code>read_groups</code> (Array[String])</li>
</ul>
<h2 id="split_string">split_string</h2>
<dl>
<dt>description</dt>
<dd>Split a string into an array of strings based on a delimiter</dd>
<dt>outputs</dt>
<dd>{'split_string': 'Split string as an array'}</dd>
</dl>
<h3 id="inputs_2">Inputs</h3>
<h4 id="required_2">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>input_string</code> (String, <strong>required</strong>): String to split on occurences of <code>delimiter</code></li>
</ul>
<h4 id="defaults_1">Defaults</h4>
<ul>
<li><code>delimiter</code> (String, default=" , "); <strong>description</strong>: Delimiter on which to split <code>input_string</code>; <strong>common</strong>: true</li>
</ul>
<h3 id="outputs_2">Outputs</h3>
<ul>
<li><code>split_strings</code> (Array[String])</li>
</ul>
<h2 id="calc_gene_lengths">calc_gene_lengths</h2>
<dl>
<dt>description</dt>
<dd>Calculate gene lengths from a GTF feature file using the "non-overlapping exonic length" algorithm</dd>
<dt>help</dt>
<dd>The "non-overlapping exonic length" algorithm can be implemented as the sum of each base covered by at least one exon; where each base is given a value of 1 regardless of how many exons overlap it.</dd>
<dt>outputs</dt>
<dd>{'gene_lengths': 'A two column headered TSV file with gene names in the first column and feature lengths (as integers) in the second column'}</dd>
</dl>
<h3 id="inputs_3">Inputs</h3>
<h4 id="required_3">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>gtf</code> (File, <strong>required</strong>): GTF feature file</li>
</ul>
<h4 id="defaults_2">Defaults</h4>
<ul>
<li><code>idattr</code> (String, default="gene_name"); <strong>description</strong>: GTF attribute to be used as feature ID. The value of this attribute will be used as the first column in the output file.; <strong>common</strong>: true</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>outfile_name</code> (String, default=basename(gtf,".gtf.gz") + ".genelengths.txt"): Name of the gene lengths file</li>
</ul>
<h3 id="outputs_3">Outputs</h3>
<ul>
<li><code>gene_lengths</code> (File)</li>
</ul>
<h2 id="compression_integrity">compression_integrity</h2>
<dl>
<dt>description</dt>
<dd>Checks the compression integrity of a bgzipped file</dd>
<dt>outputs</dt>
<dd>{'check': 'Dummy output to indicate success and to enable call-caching'}</dd>
</dl>
<h3 id="inputs_4">Inputs</h3>
<h4 id="required_4">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bgzipped_file</code> (File, <strong>required</strong>): Input bgzipped file to check integrity of</li>
</ul>
<h4 id="defaults_3">Defaults</h4>
<ul>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
</ul>
<h3 id="outputs_4">Outputs</h3>
<ul>
<li><code>check</code> (String)</li>
</ul>
<h2 id="add_to_bam_header">add_to_bam_header</h2>
<dl>
<dt>description</dt>
<dd>Adds another line of text to the bottom of a BAM header</dd>
<dt>outputs</dt>
<dd>{'reheadered_bam': 'The BAM after its header has been modified'}</dd>
</dl>
<h3 id="inputs_5">Inputs</h3>
<h4 id="required_5">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>additional_header</code> (String, <strong>required</strong>): A string to add as a new line in the BAM header. No format checking is done, so please ensure you do not invalidate your BAM with this task. Add only spec compliant entries to the header.</li>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file which will have its header added to</li>
</ul>
<h4 id="defaults_4">Defaults</h4>
<ul>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>prefix</code> (String, default=basename(bam,".bam") + ".reheader"): Prefix for the reheadered BAM. The extension <code>.bam</code> will be added.</li>
</ul>
<h3 id="outputs_5">Outputs</h3>
<ul>
<li><code>reheadered_bam</code> (File)</li>
</ul>
<h2 id="unpack_tarball">unpack_tarball</h2>
<dl>
<dt>description</dt>
<dd>Accepts a <code>.tar.gz</code> archive and converts it into a flat array of files. Any directory structure of the archive is ignored.</dd>
<dt>outputs</dt>
<dd>{'tarball_contents': 'An array of files found in the input tarball'}</dd>
</dl>
<h3 id="inputs_6">Inputs</h3>
<h4 id="required_6">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>tarball</code> (File, <strong>required</strong>): A <code>.tar.gz</code> archive to unpack into individual files</li>
</ul>
<h4 id="defaults_5">Defaults</h4>
<ul>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
</ul>
<h3 id="outputs_6">Outputs</h3>
<ul>
<li><code>tarball_contents</code> (Array[File])</li>
</ul>
<h2 id="make_coverage_regions_beds">make_coverage_regions_beds</h2>
<dl>
<dt>description</dt>
<dd>Takes in a GTF file, converts it to BED, then filters it down to two 3 column BED files: one of only 'exons', one of only 'CDS' regions</dd>
<dt>outputs</dt>
<dd>{'bed': 'Input GTF converted into BED format using the <code>gtf2bed</code> program', 'exon_bed': "3 column BED file corresponding to all 'exons' found in the input GTF", 'CDS_bed': "3 column BED file corresponding to all 'CDS' regions found in the input GTF"}</dd>
</dl>
<h3 id="inputs_7">Inputs</h3>
<h4 id="required_7">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>gtf</code> (File, <strong>required</strong>): GTF feature file from which to derive coverage regions BED files</li>
</ul>
<h4 id="defaults_6">Defaults</h4>
<ul>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
</ul>
<h3 id="outputs_7">Outputs</h3>
<ul>
<li><code>bed</code> (File)</li>
<li><code>exon_bed</code> (File)</li>
<li><code>CDS_bed</code> (File)</li>
</ul>
<h2 id="global_phred_scores">global_phred_scores</h2>
<dl>
<dt>description</dt>
<dd>Calculates statistics about PHRED scores of the input BAM</dd>
<dt>outputs</dt>
<dd>{'phred_scores': 'Headered TSV file containing PHRED score statistics'}</dd>
</dl>
<h3 id="inputs_8">Inputs</h3>
<h4 id="required_8">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file to calculate PHRED score statistics for</li>
</ul>
<h4 id="defaults_7">Defaults</h4>
<ul>
<li><code>fast_mode</code> (Boolean, default=true): Enable fast mode (true) or calculate statistics for <em><em>every</em></em> base in the BAM (false)?</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>prefix</code> (String, default=basename(bam,".bam")): Prefix for the output TSV file. The extension <code>.global_PHRED_scores.tsv</code> will be added.</li>
</ul>
<h3 id="outputs_8">Outputs</h3>
<ul>
<li><code>phred_scores</code> (File)</li>
</ul>
<h2 id="qc_summary">qc_summary</h2>
<dl>
<dt>description</dt>
<dd><strong>[OUT OF DATE]</strong> This WDL task pulls out keys metrics that can provide a high level overview of the sample, without needing to examine the entire MultiQC report. Currently, these key metrics come from Qualimap and ngsderive.</dd>
</dl>
<h3 id="inputs_9">Inputs</h3>
<h4 id="required_9">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>multiqc_tar_gz</code> (File, <strong>required</strong>)</li>
</ul>
<h4 id="defaults_8">Defaults</h4>
<ul>
<li><code>outfile_name</code> (String, default=basename(multiqc_tar_gz,".multiqc.tar.gz") + ".qc_summary.json")</li>
</ul>
<h3 id="outputs_9">Outputs</h3>
<ul>
<li><code>summary</code> (File)</li>
</ul>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../star/" class="btn btn-neutral float-left" title="Star"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../../workflows/10x-bam-to-fastqs/" class="btn btn-neutral float-right" title="Cell Ranger BAM to FASTQs">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
      <p>© 2020-Present St. Jude Children's Research Hospital</p>
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/stjudecloud/workflows" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../star/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../../workflows/10x-bam-to-fastqs/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "../..";</script>
    <script src="../../js/theme_extra.js"></script>
    <script src="../../js/theme.js"></script>
      <script src="../../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
