<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="author" content="St. Jude Cloud" />
      <link rel="shortcut icon" href="../../img/favicon.ico" />
    <title>Kraken2 - St. Jude Cloud Bioinformatics Workflows</title>
    <link rel="stylesheet" href="../../css/theme.css" />
    <link rel="stylesheet" href="../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Kraken2";
        var mkdocs_page_input_path = "tasks/kraken2.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../.." class="icon icon-home"> St. Jude Cloud Bioinformatics Workflows
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../build_for_dnanexus/">Building WDL workflows for DNAnexus</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Tasks</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../bwa/">Bwa</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../cellranger/">Cellranger</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../deeptools/">Deeptools</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../estimate/">Estimate</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../fastqc/">Fastqc</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../fq/">Fq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../gatk4/">Gatk4</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../htseq/">Htseq</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="#">Kraken2</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#inputs">Inputs</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#required">Required</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#defaults">Defaults</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#outputs">Outputs</a>
    </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../librarian/">librarian</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../md5sum/">Md5sum</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../mosdepth/">Mosdepth</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../multiqc/">Multiqc</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ngsderive/">Ngsderive</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../picard/">Picard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../qualimap/">Qualimap</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../sambamba/">Sambamba</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../samtools/">Samtools</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../star/">Star</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../util/">Utilities</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Workflows</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/10x-bam-to-fastqs/">Cell Ranger BAM to FASTQs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/ESTIMATE/">ESTIMATE</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/bam-to-fastqs/">Bam to fastqs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/bwa-db-build/">Bwa db build</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/dnaseq-standard/">Dnaseq standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/flag_filter/">FlagFilter</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/gatk-reference/">Gatk reference</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/make-qc-reference/">Make qc reference</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/markdups-post/">MarkDuplicates Post</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/quality-check-standard/">Quality check standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/rnaseq-variant-calling/">Rnaseq variant calling</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/samtools-merge/">Samtools merge</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/scrnaseq-standard/">Scrnaseq standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/star-db-build/">Star db build</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../..">St. Jude Cloud Bioinformatics Workflows</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../.." class="icon icon-home" aria-label="Docs"></a></li>
          <li class="breadcrumb-item">Tasks</li>
      <li class="breadcrumb-item active">Kraken2</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/stjudecloud/workflows/edit/master/docs/tasks/kraken2.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <p><a href="https://github.com/DerrickWood/kraken2">Homepage</a></p>
<h2 id="download_taxonomy">download_taxonomy</h2>
<dl>
<dt>description</dt>
<dd>Downloads the NCBI taxonomy which Kraken2 uses to create a tree and taxon map during the database build</dd>
<dt>outputs</dt>
<dd>{'taxonomy': 'The NCBI taxonomy, which is needed by the <code>build_db</code> task. This output is not human-readable or meant for anything other than building a Kraken2 database.'}</dd>
</dl>
<h3 id="inputs">Inputs</h3>
<h4 id="required">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
</ul>
<h4 id="defaults">Defaults</h4>
<ul>
<li><code>protein</code> (Boolean, default=false): Construct a protein database?</li>
</ul>
<h3 id="outputs">Outputs</h3>
<ul>
<li><code>taxonomy</code> (File)</li>
</ul>
<h2 id="download_library">download_library</h2>
<dl>
<dt>description</dt>
<dd>Downloads a predefined library of reference genomes from NCBI. Detailed organism list for libraries (except nt) available <a href="https://ftp.ncbi.nlm.nih.gov/genomes/refseq/">here</a></dd>
<dt>outputs</dt>
<dd>{'library': 'A library of reference genomes, which is needed by the <code>build_db</code> task. This output is not human-readable or meant for anything other than building a Kraken2 database.'}</dd>
</dl>
<h3 id="inputs_1">Inputs</h3>
<h4 id="required_1">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>library_name</code> (String, <strong>required</strong>); <strong>description</strong>: Library to download. Note that <code>protein</code> must equal <code>true</code> if downloading the <code>nr</code> library, and <code>protein</code> must equal <code>false</code> if downloading the <code>UniVec</code> or <code>UniVec_Core</code> library.; <strong>choices</strong>: ['archaea', 'bacteria', 'plasmid', 'viral', 'human', 'fungi', 'plant', 'protozoa', 'nt', 'nr', 'UniVec', 'UniVec_Core']</li>
</ul>
<h4 id="defaults_1">Defaults</h4>
<ul>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation, specified in GB. Default disk size is determined dynamically based on <code>library_name</code>. Note that the default sizes are adequate as of April 2023, but new genomes are constantly being added to the NCBI database. More disk space may be required depending on when in the future this task is run.</li>
<li><code>protein</code> (Boolean, default=false): Construct a protein database?</li>
</ul>
<h3 id="outputs_1">Outputs</h3>
<ul>
<li><code>library</code> (File)</li>
</ul>
<h2 id="create_library_from_fastas">create_library_from_fastas</h2>
<dl>
<dt>description</dt>
<dd>Adds custom entries from FASTA files to a Kraken2 DB</dd>
<dt>outputs</dt>
<dd>{'custom_library': 'Kraken2 compatible library, which is needed by the <code>build_db</code> task. This output is not human-readable or meant for anything other than building a Kraken2 database.'}</dd>
</dl>
<h3 id="inputs_2">Inputs</h3>
<h4 id="required_2">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>fastas_gz</code> (Array[File], <strong>required</strong>): Array of gzipped FASTA files. Each FASTA sequence ID must contain either an NCBI accession number or an explicit assignment of the taxonomy ID using <code>kraken:taxid</code></li>
</ul>
<h4 id="defaults_2">Defaults</h4>
<ul>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>protein</code> (Boolean, default=false): Construct a protein database?</li>
</ul>
<h3 id="outputs_2">Outputs</h3>
<ul>
<li><code>custom_library</code> (File)</li>
</ul>
<h2 id="build_db">build_db</h2>
<dl>
<dt>description</dt>
<dd>Builds a custom Kraken2 database</dd>
<dt>outputs</dt>
<dd>{'built_db': 'A complete Kraken2 database'}</dd>
</dl>
<h3 id="inputs_3">Inputs</h3>
<h4 id="required_3">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>tarballs</code> (Array[File], <strong>required</strong>): Tarballs containing the NCBI taxonomy (generated by the <code>download_taxonomy</code> task) and at least one library (generated by the <code>download_library</code> or <code>create_library_from_fastas</code> task). Tarballs must not have a root directory.</li>
</ul>
<h4 id="defaults_3">Defaults</h4>
<ul>
<li><code>db_name</code> (String, default="kraken2_db"); <strong>description</strong>: Name for output in compressed, archived format. The suffix <code>.tar.gz</code> will be added.; <strong>common</strong>: true</li>
<li><code>kmer_len</code> (Int, default=if protein then 15 else 35): K-mer length in bp that will be used to build the database</li>
<li><code>max_db_size_gb</code> (Int, default=-1): Maximum number of GBs for Kraken 2 hash table; if the Kraken 2 estimator determines more would normally be needed, the reference library will be downsampled to fit.</li>
<li><code>minimizer_len</code> (Int, default=if protein then 12 else 31): Minimizer length in bp that will be used to build the database</li>
<li><code>minimizer_spaces</code> (Int, default=if protein then 0 else 7): Number of characters in minimizer that are ignored in comparisons</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>modify_memory_gb</code> (Int, default=0): Add to or subtract from dynamic memory allocation. Default memory is determined by the size of the inputs. Specified in GB.</li>
<li><code>ncpu</code> (Int, default=4); <strong>description</strong>: Number of cores to allocate for task; <strong>common</strong>: true</li>
<li><code>protein</code> (Boolean, default=false): Construct a protein database?</li>
<li><code>use_all_cores</code> (Boolean, default=false); <strong>description</strong>: Use all cores? Recommended for cloud environments.; <strong>common</strong>: true</li>
</ul>
<h3 id="outputs_3">Outputs</h3>
<ul>
<li><code>built_db</code> (File)</li>
</ul>
<h2 id="kraken">kraken</h2>
<dl>
<dt>description</dt>
<dd>Runs Kraken2 on a pair of fastq files</dd>
<dt>outputs</dt>
<dd>{'report': {'description': 'A Kraken2 summary report', 'external_help': 'https://github.com/DerrickWood/kraken2/blob/master/docs/MANUAL.markdown#sample-report-output-format'}, 'sequences': {'description': 'Detailed Kraken2 output that has been gzipped', 'external_help': 'https://github.com/DerrickWood/kraken2/blob/master/docs/MANUAL.markdown#standard-kraken-output-format'}}</dd>
</dl>
<h3 id="inputs_4">Inputs</h3>
<h4 id="required_4">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>db</code> (File, <strong>required</strong>): Kraken2 database. Can be generated with <code>make-qc-reference.wdl</code>. Must be a tarball without a root directory.</li>
<li><code>read_one_fastq_gz</code> (File, <strong>required</strong>): Gzipped FASTQ file with 1st reads in pair</li>
<li><code>read_two_fastq_gz</code> (File, <strong>required</strong>): Gzipped FASTQ file with 2nd reads in pair</li>
</ul>
<h4 id="defaults_4">Defaults</h4>
<ul>
<li><code>min_base_quality</code> (Int, default=0): Minimum base quality used in classification</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>modify_memory_gb</code> (Int, default=0): Add to or subtract from dynamic memory allocation. Default memory is determined by the size of the inputs. Specified in GB.</li>
<li><code>ncpu</code> (Int, default=4); <strong>description</strong>: Number of cores to allocate for task; <strong>common</strong>: true</li>
<li><code>prefix</code> (String, default=sub(basename(read_one_fastq_gz),"([_\.][rR][12])?(\.subsampled)?\.(fastq|fq)(\.gz)?$","")): Prefix for the Kraken2 output files. The extensions <code>.kraken2.txt</code> and <code>.kraken2.sequences.txt.gz</code> will be added.</li>
<li><code>store_sequences</code> (Boolean, default=false); <strong>description</strong>: Store and output main Kraken2 output in addition to the summary report?; <strong>common</strong>: true</li>
<li><code>use_all_cores</code> (Boolean, default=false); <strong>description</strong>: Use all cores? Recommended for cloud environments.; <strong>common</strong>: true</li>
<li><code>use_names</code> (Boolean, default=true): Print scientific names instead of just taxids?</li>
</ul>
<h3 id="outputs_4">Outputs</h3>
<ul>
<li><code>report</code> (File)</li>
<li><code>sequences</code> (File?)</li>
</ul>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../htseq/" class="btn btn-neutral float-left" title="Htseq"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../librarian/" class="btn btn-neutral float-right" title="librarian">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
      <p>Â© 2020-Present St. Jude Children's Research Hospital</p>
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/stjudecloud/workflows" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../htseq/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../librarian/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "../..";</script>
    <script src="../../js/theme_extra.js"></script>
    <script src="../../js/theme.js"></script>
      <script src="../../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
