<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="author" content="St. Jude Cloud" />
      <link rel="shortcut icon" href="../../img/favicon.ico" />
    <title>Ngsderive - St. Jude Cloud Bioinformatics Workflows</title>
    <link rel="stylesheet" href="../../css/theme.css" />
    <link rel="stylesheet" href="../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Ngsderive";
        var mkdocs_page_input_path = "tasks/ngsderive.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../.." class="icon icon-home"> St. Jude Cloud Bioinformatics Workflows
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../build_for_dnanexus/">Building WDL workflows for DNAnexus</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Tasks</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../bwa/">Bwa</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../cellranger/">Cellranger</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../deeptools/">Deeptools</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../estimate/">Estimate</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../fastqc/">Fastqc</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../fq/">Fq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../htseq/">Htseq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../kraken2/">Kraken2</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../librarian/">librarian</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../md5sum/">Md5sum</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../mosdepth/">Mosdepth</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../multiqc/">Multiqc</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="./">Ngsderive</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#inputs">Inputs</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#required">Required</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#defaults">Defaults</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#outputs">Outputs</a>
    </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../picard/">Picard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../qualimap/">Qualimap</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../samtools/">Samtools</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../star/">Star</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../util/">Utilities</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Workflows</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/10x-bam-to-fastqs/">Cell Ranger BAM to FASTQs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/ESTIMATE/">ESTIMATE</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/bam-to-fastqs/">Bam to fastqs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/bwa-db-build/">Bwa db build</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/dnaseq-standard/">Dnaseq standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/flag_filter/">FlagFilter</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/make-qc-reference/">Make qc reference</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/markdups-post/">MarkDuplicates Post</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/quality-check-standard/">Quality check standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/samtools-merge/">Samtools merge</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/scrnaseq-standard/">Scrnaseq standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../workflows/star-db-build/">Star db build</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../..">St. Jude Cloud Bioinformatics Workflows</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../.." class="icon icon-home" aria-label="Docs"></a></li>
          <li class="breadcrumb-item">Tasks</li>
      <li class="breadcrumb-item active">Ngsderive</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/stjudecloud/workflows/edit/master/docs/tasks/ngsderive.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <p><a href="https://github.com/stjudecloud/ngsderive">Homepage</a></p>
<h2 id="strandedness">strandedness</h2>
<dl>
<dt>description</dt>
<dd>Derives the experimental strandedness protocol used to generate the input RNA-Seq BAM file. Reports evidence supporting final results.</dd>
<dt>outputs</dt>
<dd>{'strandedness_file': 'TSV file containing the <code>ngsderive strandedness</code> report', 'strandedness': 'The derived strandedness, in string format'}</dd>
</dl>
<h3 id="inputs">Inputs</h3>
<h4 id="required">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file to derive strandedness for</li>
<li><code>bam_index</code> (File, <strong>required</strong>): BAM index file corresponding to the input BAM</li>
<li><code>gene_model</code> (File, <strong>required</strong>): Gene model as a GFF/GTF file</li>
</ul>
<h4 id="defaults">Defaults</h4>
<ul>
<li><code>min_mapq</code> (Int, default=30); <strong>description</strong>: Minimum MAPQ to consider for supporting reads; <strong>common</strong>: true</li>
<li><code>min_reads_per_gene</code> (Int, default=10); <strong>description</strong>: Filter any genes that don't have at least <code>min_reads_per_gene</code> reads mapping to them; <strong>common</strong>: true</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>num_genes</code> (Int, default=1000); <strong>description</strong>: How many genes to sample; <strong>common</strong>: true</li>
<li><code>outfile_name</code> (String, default=basename(bam,".bam") + ".strandedness.tsv"): Name for the strandedness TSV file</li>
<li><code>split_by_rg</code> (Boolean, default=false); <strong>description</strong>: Contain one entry in the output TSV per read group, in addition to an <code>overall</code> entry; <strong>common</strong>: true</li>
</ul>
<h3 id="outputs">Outputs</h3>
<ul>
<li><code>strandedness_string</code> (String)</li>
<li><code>strandedness_file</code> (File)</li>
</ul>
<h2 id="instrument">instrument</h2>
<dl>
<dt>description</dt>
<dd>Derives the instrument used to sequence the input BAM file. Reports evidence supporting final results.</dd>
<dt>outputs</dt>
<dd>{'instrument_file': 'TSV file containing the <code>ngsderive isntrument</code> report for the input BAM file'}</dd>
</dl>
<h3 id="inputs_1">Inputs</h3>
<h4 id="required_1">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file to derive instrument for</li>
</ul>
<h4 id="defaults_1">Defaults</h4>
<ul>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>num_reads</code> (Int, default=10000); <strong>description</strong>: How many reads to analyze from the start of the file. Any n &lt; 1 to parse whole file.; <strong>common</strong>: true</li>
<li><code>outfile_name</code> (String, default=basename(bam,".bam") + ".instrument.tsv"): Name for the instrument TSV file</li>
</ul>
<h3 id="outputs_1">Outputs</h3>
<ul>
<li><code>instrument_file</code> (File)</li>
<li><code>instrument_string</code> (String)</li>
</ul>
<h2 id="read_length">read_length</h2>
<dl>
<dt>description</dt>
<dd>Derives the original experimental read length of the input BAM. Reports evidence supporting final results.</dd>
<dt>outputs</dt>
<dd>{'read_length_file': 'TSV file containing the <code>ngsderive readlen</code> report for the input BAM file'}</dd>
</dl>
<h3 id="inputs_2">Inputs</h3>
<h4 id="required_2">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file to derive read length for</li>
<li><code>bam_index</code> (File, <strong>required</strong>): BAM index file corresponding to the input BAM</li>
</ul>
<h4 id="defaults_2">Defaults</h4>
<ul>
<li><code>majority_vote_cutoff</code> (Float, default=0.7); <strong>description</strong>: To call a majority readlen, the maximum read length must have at least <code>majority-vote-cutoff</code>% reads in support; <strong>common</strong>: true</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>num_reads</code> (Int, default=-1); <strong>description</strong>: How many reads to analyze from the start of the file. Any n &lt; 1 to parse whole file.; <strong>common</strong>: true</li>
<li><code>outfile_name</code> (String, default=basename(bam,".bam") + ".readlength.tsv"): Name for the readlen TSV file</li>
</ul>
<h3 id="outputs_2">Outputs</h3>
<ul>
<li><code>read_length_file</code> (File)</li>
</ul>
<h2 id="encoding">encoding</h2>
<dl>
<dt>description</dt>
<dd>Derives the encoding of the input NGS file(s). Reports evidence supporting final results.</dd>
<dt>outputs</dt>
<dd>{'encoding_file': 'TSV file containing the <code>ngsderive encoding</code> report for all input files', 'inferred_encoding': 'The most permissive encoding found among the input files, in string format'}</dd>
</dl>
<h3 id="inputs_3">Inputs</h3>
<h4 id="required_3">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>ngs_files</code> (Array[File], <strong>required</strong>): An array of FASTQs and/or BAMs for which to derive encoding</li>
<li><code>outfile_name</code> (String, <strong>required</strong>): Name for the encoding TSV file</li>
</ul>
<h4 id="defaults_3">Defaults</h4>
<ul>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>num_reads</code> (Int, default=1000000); <strong>description</strong>: How many reads to analyze from the start of the file(s). Any n &lt; 1 to parse whole file(s).; <strong>common</strong>: true</li>
</ul>
<h3 id="outputs_3">Outputs</h3>
<ul>
<li><code>inferred_encoding</code> (String)</li>
<li><code>encoding_file</code> (File)</li>
</ul>
<h2 id="junction_annotation">junction_annotation</h2>
<dl>
<dt>description</dt>
<dd>Annotates junctions found in an RNA-Seq BAM as known, novel, or partially novel</dd>
<dt>external_help</dt>
<dd>https://stjudecloud.github.io/ngsderive/subcommands/junction_annotation/</dd>
<dt>outputs</dt>
<dd>{'junction_summary': 'TSV file containing the <code>ngsderive junction-annotation</code> summary', 'junctions': 'TSV file containing a detailed list of annotated junctions'}</dd>
</dl>
<h3 id="inputs_4">Inputs</h3>
<h4 id="required_4">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file to annotate junctions for</li>
<li><code>bam_index</code> (File, <strong>required</strong>): BAM index file corresponding to the input BAM</li>
<li><code>gene_model</code> (File, <strong>required</strong>): Gene model as a GFF/GTF file</li>
</ul>
<h4 id="defaults_4">Defaults</h4>
<ul>
<li><code>fuzzy_junction_match_range</code> (Int, default=0); <strong>description</strong>: Consider found splices within <code>+-k</code> bases of a known splice event annotated; <strong>common</strong>: true</li>
<li><code>min_intron</code> (Int, default=50); <strong>description</strong>: Minimum size of intron to be considered a splice; <strong>common</strong>: true</li>
<li><code>min_mapq</code> (Int, default=30); <strong>description</strong>: Minimum MAPQ to consider for supporting reads; <strong>common</strong>: true</li>
<li><code>min_reads</code> (Int, default=2); <strong>description</strong>: Filter any junctions that don't have at least <code>min_reads</code> reads supporting them; <strong>common</strong>: true</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>prefix</code> (String, default=basename(bam,".bam")): Prefix for the summary TSV and junction files. The extensions <code>.junction_summary.tsv</code> and <code>.junctions.tsv</code> will be added.</li>
</ul>
<h3 id="outputs_4">Outputs</h3>
<ul>
<li><code>junction_summary</code> (File)</li>
<li><code>junctions</code> (File)</li>
</ul>
<h2 id="endedness">endedness</h2>
<dl>
<dt>description</dt>
<dd>Derives the endedness of the input BAM file. Reports evidence for final result.</dd>
<dt>outputs</dt>
<dd>{'endedness_file': 'TSV file containing the <code>ngsderive endedness</code> report'}</dd>
</dl>
<h3 id="inputs_5">Inputs</h3>
<h4 id="required_5">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file to derive endedness from</li>
</ul>
<h4 id="defaults_5">Defaults</h4>
<ul>
<li><code>calc_rpt</code> (Boolean, default=false); <strong>description</strong>: Calculate and output Reads-Per-Template. This will produce a more sophisticated estimate for endedness, but uses substantially more memory (can reach up to 200% of BAM size in memory consumption for some inputs).; <strong>common</strong>: true</li>
<li><code>lenient</code> (Boolean, default=false); <strong>description</strong>: Return a zero exit code on unknown results; <strong>common</strong>: true</li>
<li><code>modify_disk_size_gb</code> (Int, default=0): Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB.</li>
<li><code>modify_memory_gb</code> (Int, default=0): Add to or subtract from dynamic memory allocation. Default memory is determined by value of <code>calc_rpt</code> and the size of the input. Specified in GB.</li>
<li><code>num_reads</code> (Int, default=-1); <strong>description</strong>: How many reads to analyze from the start of the file. Any n &lt; 1 to parse whole file.; <strong>common</strong>: true</li>
<li><code>outfile_name</code> (String, default=basename(bam,".bam") + ".endedness.tsv"): Name for the endedness TSV file</li>
<li><code>paired_deviance</code> (Float, default=0.0); <strong>description</strong>: Distance from 0.5 split between number of f+l- reads and f-l+ reads allowed to be called 'Paired-End'. Default of <code>0.0</code> only appropriate if the whole file is being processed.; <strong>common</strong>: true</li>
<li><code>round_rpt</code> (Boolean, default=false); <strong>description</strong>: Round RPT to the nearest INT before comparing to expected values. Appropriate if using <code>--num-reads</code> &gt; 0.; <strong>common</strong>: true</li>
<li><code>split_by_rg</code> (Boolean, default=false); <strong>description</strong>: Contain one entry per read group; <strong>common</strong>: true</li>
</ul>
<h3 id="outputs_5">Outputs</h3>
<ul>
<li><code>endedness_file</code> (File)</li>
</ul>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../multiqc/" class="btn btn-neutral float-left" title="Multiqc"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../picard/" class="btn btn-neutral float-right" title="Picard">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
      <p>© 2020-Present St. Jude Children's Research Hospital</p>
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/stjudecloud/workflows" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../multiqc/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../picard/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "../..";</script>
    <script src="../../js/theme_extra.js"></script>
    <script src="../../js/theme.js"></script>
      <script src="../../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
