<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="author" content="St. Jude Cloud" />
      <link rel="shortcut icon" href="../../img/favicon.ico" />
    <title>Dnaseq standard fastq - St. Jude Cloud Bioinformatics Workflows</title>
    <link rel="stylesheet" href="../../css/theme.css" />
    <link rel="stylesheet" href="../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Dnaseq standard fastq";
        var mkdocs_page_input_path = "workflows/dnaseq-standard-fastq.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../.." class="icon icon-home"> St. Jude Cloud Bioinformatics Workflows
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../build_for_dnanexus/">Building WDL workflows for DNAnexus</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Tasks</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/bwa/">Bwa</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/cellranger/">Cellranger</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/deeptools/">Deeptools</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/estimate/">Estimate</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/fastqc/">Fastqc</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/fq/">Fq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/gatk4/">Gatk4</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/htseq/">Htseq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/kraken2/">Kraken2</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/librarian/">librarian</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/md5sum/">Md5sum</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/mosdepth/">Mosdepth</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/multiqc/">Multiqc</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/ngsderive/">Ngsderive</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/picard/">Picard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/qualimap/">Qualimap</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/read_group/">Read group</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/sambamba/">Sambamba</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/samtools/">Samtools</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/star/">Star</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/util/">Utilities</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Workflows</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../10x-bam-to-fastqs/">Cell Ranger BAM to FASTQs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ESTIMATE/">ESTIMATE</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../bam-to-fastqs/">Bam to fastqs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../bwa-db-build/">Bwa db build</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../dnaseq-core/">Dnaseq core</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="#">Dnaseq standard fastq</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#inputs">Inputs</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#required">Required</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#optional">Optional</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#defaults">Defaults</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#outputs">Outputs</a>
    </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../dnaseq-standard/">Dnaseq standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../flag_filter/">FlagFilter</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../gatk-reference/">Gatk reference</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../make-qc-reference/">Make qc reference</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../markdups-post/">MarkDuplicates Post</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../quality-check-standard/">Quality check standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../rnaseq-variant-calling/">Rnaseq variant calling</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../samtools-merge/">Samtools merge</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../scrnaseq-standard/">Scrnaseq standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../star-db-build/">Star db build</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../..">St. Jude Cloud Bioinformatics Workflows</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../.." class="icon icon-home" aria-label="Docs"></a></li>
          <li class="breadcrumb-item">Workflows</li>
      <li class="breadcrumb-item active">Dnaseq standard fastq</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/stjudecloud/workflows/edit/master/docs/workflows/dnaseq-standard-fastq.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <p><strong>WARNING:</strong> this workflow is experimental! Use at your own risk!</p>
<h2 id="dnaseq_standard_fastq_experimental">dnaseq_standard_fastq_experimental</h2>
<dl>
<dt>description</dt>
<dd>Aligns DNA reads using bwa</dd>
<dt>outputs</dt>
<dd>{'harmonized_bam': 'Harmonized DNA-Seq BAM, aligned with bwa', 'harmonized_bam_index': 'Index for the harmonized DNA-Seq BAM file'}</dd>
<dt>allowNestedInputs</dt>
<dd>true</dd>
</dl>
<h3 id="inputs">Inputs</h3>
<h4 id="required">Required</h4>
<ul>
<li><code>bwa_db</code> (File, <strong>required</strong>): Gzipped tar archive of the bwa reference files. Files should be at the root of the archive.</li>
<li><code>prefix</code> (String, <strong>required</strong>): Prefix for the BAM file. The extension <code>.bam</code> will be added.</li>
<li><code>read_groups</code> (Array[ReadGroup], <strong>required</strong>); <strong>description</strong>: An Array of structs defining read groups to include in the harmonized BAM. Must correspond to input FASTQs. Each read group ID must be contained in the basename of a FASTQ file or pair of FASTQ files if Paired-End. This requirement means the length of <code>read_groups</code> must equal the length of <code>read_one_fastqs_gz</code> and the length of <code>read_two_fastqs_gz</code> if non-zero. Only the <code>ID</code> field is required, and it must be unique for each read group defined. See data_structures/read_group.wdl for help formatting your input JSON.; <strong>external_help</strong>: https://samtools.github.io/hts-specs/SAMv1.pdf</li>
<li><code>read_one_fastqs_gz</code> (Array[File], <strong>required</strong>): Input gzipped FASTQ format file(s) with 1st read in pair to align</li>
<li><code>read_two_fastqs_gz</code> (Array[File], <strong>required</strong>): Input gzipped FASTQ format file(s) with 2nd read in pair to align</li>
<li><code>fqlint._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>parse_input._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>subsample._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>dnaseq_core_experimental.bwa_aln_pe._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>dnaseq_core_experimental.bwa_mem._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>dnaseq_core_experimental.index._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>dnaseq_core_experimental.read_group_to_string._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>dnaseq_core_experimental.read_ones._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>dnaseq_core_experimental.read_twos._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>dnaseq_core_experimental.sort._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>dnaseq_core_experimental.rg_merge.basic_merge._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>dnaseq_core_experimental.rg_merge.final_merge._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>dnaseq_core_experimental.rg_merge.inner_merge._runtime</code> (Any, <strong>required</strong>)</li>
</ul>
<h4 id="optional">Optional</h4>
<ul>
<li><code>dnaseq_core_experimental.sample_override</code> (String?)</li>
<li><code>dnaseq_core_experimental.rg_merge.basic_merge.new_header</code> (File?)</li>
<li><code>dnaseq_core_experimental.rg_merge.final_merge.new_header</code> (File?)</li>
<li><code>dnaseq_core_experimental.rg_merge.inner_merge.new_header</code> (File?)</li>
</ul>
<h4 id="defaults">Defaults</h4>
<ul>
<li><code>aligner</code> (String, default="mem"); <strong>description</strong>: BWA aligner to use; <strong>choices</strong>: ['mem', 'aln']</li>
<li><code>reads_per_file</code> (Int, default=10000000): Controls the number of reads per FASTQ file for internal split to run BWA in parallel.</li>
<li><code>subsample_n_reads</code> (Int, default=-1): Only process a random sampling of <code>n</code> reads. Any <code>n</code>&lt;=<code>0</code> for processing entire input.</li>
<li><code>use_all_cores</code> (Boolean, default=false): Use all cores? Recommended for cloud environments.</li>
<li><code>validate_input</code> (Boolean, default=true): Ensure input FASTQs ares well-formed before beginning harmonization?</li>
<li><code>fqlint.disable_validator_codes</code> (Array[String], default=[])</li>
<li><code>fqlint.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>fqlint.modify_memory_gb</code> (Int, default=0)</li>
<li><code>fqlint.paired_read_validation_level</code> (String, default="high")</li>
<li><code>fqlint.panic</code> (Boolean, default=true)</li>
<li><code>fqlint.single_read_validation_level</code> (String, default="high")</li>
<li><code>subsample.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>subsample.prefix</code> (String, default=sub(basename(read_one_fastq),"([_\.][rR][12])?(\.subsampled)?\.(fastq|fq)(\.gz)?$","")): Prefix for the BAM file. The extension <code>.bam</code> will be added.</li>
<li><code>subsample.probability</code> (Float, default=1.0)</li>
<li><code>dnaseq_core_experimental.bwa_aln_pe.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>dnaseq_core_experimental.bwa_aln_pe.ncpu</code> (Int, default=4)</li>
<li><code>dnaseq_core_experimental.bwa_mem.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>dnaseq_core_experimental.bwa_mem.ncpu</code> (Int, default=4)</li>
<li><code>dnaseq_core_experimental.index.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>dnaseq_core_experimental.index.ncpu</code> (Int, default=2)</li>
<li><code>dnaseq_core_experimental.index.use_all_cores</code> (Boolean, default=false): Use all cores? Recommended for cloud environments.</li>
<li><code>dnaseq_core_experimental.read_ones.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>dnaseq_core_experimental.read_ones.ncpu</code> (Int, default=2)</li>
<li><code>dnaseq_core_experimental.read_ones.prefix</code> (String, default=sub(basename(fastq),"(fastq|fq)\.gz$","")): Prefix for the BAM file. The extension <code>.bam</code> will be added.</li>
<li><code>dnaseq_core_experimental.read_twos.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>dnaseq_core_experimental.read_twos.ncpu</code> (Int, default=2)</li>
<li><code>dnaseq_core_experimental.read_twos.prefix</code> (String, default=sub(basename(fastq),"(fastq|fq)\.gz$","")): Prefix for the BAM file. The extension <code>.bam</code> will be added.</li>
<li><code>dnaseq_core_experimental.rg_merge.max_length</code> (Int, default=100)</li>
<li><code>dnaseq_core_experimental.sort.memory_gb</code> (Int, default=25)</li>
<li><code>dnaseq_core_experimental.sort.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>dnaseq_core_experimental.sort.prefix</code> (String, default=basename(bam,".bam") + ".sorted"): Prefix for the BAM file. The extension <code>.bam</code> will be added.</li>
<li><code>dnaseq_core_experimental.sort.sort_order</code> (String, default="coordinate")</li>
<li><code>dnaseq_core_experimental.sort.validation_stringency</code> (String, default="SILENT")</li>
<li><code>dnaseq_core_experimental.rg_merge.basic_merge.combine_rg</code> (Boolean, default=true)</li>
<li><code>dnaseq_core_experimental.rg_merge.basic_merge.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>dnaseq_core_experimental.rg_merge.basic_merge.name_sorted</code> (Boolean, default=false)</li>
<li><code>dnaseq_core_experimental.rg_merge.basic_merge.ncpu</code> (Int, default=2)</li>
<li><code>dnaseq_core_experimental.rg_merge.basic_merge.region</code> (String, default="")</li>
<li><code>dnaseq_core_experimental.rg_merge.final_merge.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>dnaseq_core_experimental.rg_merge.final_merge.name_sorted</code> (Boolean, default=false)</li>
<li><code>dnaseq_core_experimental.rg_merge.final_merge.ncpu</code> (Int, default=2)</li>
<li><code>dnaseq_core_experimental.rg_merge.final_merge.region</code> (String, default="")</li>
<li><code>dnaseq_core_experimental.rg_merge.inner_merge.combine_rg</code> (Boolean, default=true)</li>
<li><code>dnaseq_core_experimental.rg_merge.inner_merge.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>dnaseq_core_experimental.rg_merge.inner_merge.name_sorted</code> (Boolean, default=false)</li>
<li><code>dnaseq_core_experimental.rg_merge.inner_merge.ncpu</code> (Int, default=2)</li>
<li><code>dnaseq_core_experimental.rg_merge.inner_merge.region</code> (String, default="")</li>
</ul>
<h3 id="outputs">Outputs</h3>
<ul>
<li><code>harmonized_bam</code> (File)</li>
<li><code>harmonized_bam_index</code> (File)</li>
</ul>
<h2 id="parse_input">parse_input</h2>
<dl>
<dt>description</dt>
<dd>Parses and validates the <code>dnaseq_standard</code> workflow's provided inputs</dd>
<dt>outputs</dt>
<dd>{'check': 'Dummy output to indicate success and to enable call-caching'}</dd>
</dl>
<h3 id="inputs_1">Inputs</h3>
<h4 id="required_1">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>aligner</code> (String, <strong>required</strong>); <strong>description</strong>: BWA aligner to use; <strong>choices</strong>: ['mem', 'aln']</li>
<li><code>array_lengths</code> (Array[Int], <strong>required</strong>): Expected to be an array of length 3, containing the lengths of <code>read_one_fastqs_gz</code>, <code>read_two_fastqs_gz</code>, and <code>read_groups</code></li>
</ul>
<h3 id="outputs_1">Outputs</h3>
<ul>
<li><code>check</code> (String)</li>
</ul>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../dnaseq-core/" class="btn btn-neutral float-left" title="Dnaseq core"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../dnaseq-standard/" class="btn btn-neutral float-right" title="Dnaseq standard">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
      <p>Â© 2020-Present St. Jude Children's Research Hospital</p>
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/stjudecloud/workflows" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../dnaseq-core/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../dnaseq-standard/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "../..";</script>
    <script src="../../js/theme_extra.js"></script>
    <script src="../../js/theme.js"></script>
      <script src="../../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
