<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="author" content="St. Jude Cloud" />
      <link rel="shortcut icon" href="../../img/favicon.ico" />
    <title>Rnaseq variant calling - St. Jude Cloud Bioinformatics Workflows</title>
    <link rel="stylesheet" href="../../css/theme.css" />
    <link rel="stylesheet" href="../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Rnaseq variant calling";
        var mkdocs_page_input_path = "workflows/rnaseq-variant-calling.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../.." class="icon icon-home"> St. Jude Cloud Bioinformatics Workflows
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../build_for_dnanexus/">Building WDL workflows for DNAnexus</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Tasks</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/arriba/">Arriba</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/bwa/">Bwa</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/deeptools/">Deeptools</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/estimate/">Estimate</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/fastqc/">Fastqc</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/fq/">Fq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/gatk4/">Gatk4</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/htseq/">Htseq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/kraken2/">Kraken2</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/librarian/">librarian</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/md5sum/">Md5sum</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/mosdepth/">Mosdepth</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/multiqc/">Multiqc</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/ngsderive/">Ngsderive</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/picard/">Picard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/qualimap/">Qualimap</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/read_group/">Read group</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/sambamba/">Sambamba</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/samtools/">Samtools</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/star/">Star</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/util/">Utilities</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Workflows</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../ESTIMATE/">ESTIMATE</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../bam-to-fastqs/">Bam to fastqs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../bwa-db-build/">Bwa db build</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../dnaseq-core/">Dnaseq core</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../dnaseq-standard-fastq/">Dnaseq standard fastq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../dnaseq-standard/">Dnaseq standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../flag_filter/">FlagFilter</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../gatk-reference/">Gatk reference</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../make-qc-reference/">Make qc reference</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../markdups-post/">MarkDuplicates Post</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../quality-check-standard/">Quality check standard</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="#">Rnaseq variant calling</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#inputs">Inputs</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#required">Required</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#defaults">Defaults</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#outputs">Outputs</a>
    </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../samtools-merge/">Samtools merge</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../star-db-build/">Star db build</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../..">St. Jude Cloud Bioinformatics Workflows</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../.." class="icon icon-home" aria-label="Docs"></a></li>
          <li class="breadcrumb-item">Workflows</li>
      <li class="breadcrumb-item active">Rnaseq variant calling</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/stjudecloud/workflows/edit/master/docs/workflows/rnaseq-variant-calling.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h2 id="rnaseq_variant_calling">rnaseq_variant_calling</h2>
<dl>
<dt>description</dt>
<dd>Call short germline variants from RNA-Seq data. Produces a VCF file of variants. Based on GATK RNA-Seq short variant calling best practices pipeline.</dd>
<dt>outputs</dt>
<dd>{'recalibrated_bam': 'BAM that has undergone recalibration of base quality scores', 'recalibrated_bam_index': 'Index file for recalibrated BAM file', 'variant_filtered_vcf': 'VCF file after variant filters have been applied', 'variant_filtered_vcf_index': 'Index for filtered variant VCF file'}</dd>
</dl>
<h3 id="inputs">Inputs</h3>
<h4 id="required">Required</h4>
<ul>
<li><code>bam</code> (File, <strong>required</strong>): BAM file of aligned RNA-Seq reads</li>
<li><code>bam_index</code> (File, <strong>required</strong>): Index file for BAM file</li>
<li><code>calling_interval_list</code> (File, <strong>required</strong>): Interval list of regions from which to call variants. Used for parallelization.</li>
<li><code>dbSNP_vcf</code> (File, <strong>required</strong>): dbSNP VCF file</li>
<li><code>dbSNP_vcf_index</code> (File, <strong>required</strong>): Index file for dbSNP VCF file</li>
<li><code>dict</code> (File, <strong>required</strong>): Sequence dictionary for reference FASTA file</li>
<li><code>fasta</code> (File, <strong>required</strong>): Reference FASTA file</li>
<li><code>fasta_index</code> (File, <strong>required</strong>): Index file for reference FASTA file</li>
<li><code>known_vcf_indexes</code> (Array[File], <strong>required</strong>): Array of index files for known indels VCF files</li>
<li><code>known_vcfs</code> (Array[File], <strong>required</strong>): Array of known indels VCF files</li>
<li><code>apply_bqsr._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>base_recalibrator._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>haplotype_caller._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>mark_duplicates._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>merge_vcfs._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>scatter_interval_list._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>split_n_cigar_reads._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>variant_filtration._runtime</code> (Any, <strong>required</strong>)</li>
</ul>
<h4 id="defaults">Defaults</h4>
<ul>
<li><code>bam_is_dup_marked</code> (Boolean, default=false): Whether the input BAM file has duplicates marked.</li>
<li><code>prefix</code> (String, default=basename(bam,".bam")): Prefix for the output files.</li>
<li><code>scatter_count</code> (Int, default=6): Number of intervals to scatter over. This should typically be set to 5-20. Higher values will increase parallelism and speed up the workflow, but increase overhead in provisioning resources.</li>
<li><code>apply_bqsr.memory_gb</code> (Int, default=25)</li>
<li><code>apply_bqsr.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>apply_bqsr.ncpu</code> (Int, default=4)</li>
<li><code>apply_bqsr.use_original_quality_scores</code> (Boolean, default=false)</li>
<li><code>base_recalibrator.memory_gb</code> (Int, default=25)</li>
<li><code>base_recalibrator.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>base_recalibrator.ncpu</code> (Int, default=4)</li>
<li><code>base_recalibrator.outfile_name</code> (String, default=basename(bam,".bam") + ".recal.txt")</li>
<li><code>base_recalibrator.use_original_quality_scores</code> (Boolean, default=false)</li>
<li><code>haplotype_caller.memory_gb</code> (Int, default=25)</li>
<li><code>haplotype_caller.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>haplotype_caller.ncpu</code> (Int, default=4)</li>
<li><code>haplotype_caller.prefix</code> (String, default=basename(bam,".bam")): Prefix for the output files.</li>
<li><code>haplotype_caller.stand_call_conf</code> (Int, default=20)</li>
<li><code>haplotype_caller.use_soft_clipped_bases</code> (Boolean, default=false)</li>
<li><code>mark_duplicates.clear_dt</code> (Boolean, default=true)</li>
<li><code>mark_duplicates.duplicate_scoring_strategy</code> (String, default="SUM_OF_BASE_QUALITIES")</li>
<li><code>mark_duplicates.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>mark_duplicates.modify_memory_gb</code> (Int, default=0)</li>
<li><code>mark_duplicates.optical_distance</code> (Int, default=0)</li>
<li><code>mark_duplicates.prefix</code> (String, default=basename(bam,".bam") + ".MarkDuplicates"): Prefix for the output files.</li>
<li><code>mark_duplicates.read_name_regex</code> (String, default="^[!-9;-?A-~:]+:([!-9;-?A-~]+):([0-9]+):([0-9]+)$")</li>
<li><code>mark_duplicates.remove_duplicates</code> (Boolean, default=false)</li>
<li><code>mark_duplicates.remove_sequencing_duplicates</code> (Boolean, default=false)</li>
<li><code>mark_duplicates.tagging_policy</code> (String, default="All")</li>
<li><code>mark_duplicates.validation_stringency</code> (String, default="SILENT")</li>
<li><code>merge_vcfs.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>scatter_interval_list.sort</code> (Boolean, default=true)</li>
<li><code>scatter_interval_list.subdivision_mode</code> (String, default="BALANCING_WITHOUT_INTERVAL_SUBDIVISION_WITH_OVERFLOW")</li>
<li><code>scatter_interval_list.unique</code> (Boolean, default=true)</li>
<li><code>split_n_cigar_reads.memory_gb</code> (Int, default=25)</li>
<li><code>split_n_cigar_reads.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>split_n_cigar_reads.ncpu</code> (Int, default=8)</li>
<li><code>split_n_cigar_reads.prefix</code> (String, default=basename(bam,".bam") + ".split"): Prefix for the output files.</li>
<li><code>variant_filtration.cluster</code> (Int, default=3)</li>
<li><code>variant_filtration.filter_expressions</code> (Array[String], default=["FS &gt; 30.0", "QD &lt; 2.0"])</li>
<li><code>variant_filtration.filter_names</code> (Array[String], default=["FS", "QD"])</li>
<li><code>variant_filtration.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>variant_filtration.ncpu</code> (Int, default=1)</li>
<li><code>variant_filtration.window</code> (Int, default=35)</li>
</ul>
<h3 id="outputs">Outputs</h3>
<ul>
<li><code>recalibrated_bam</code> (File)</li>
<li><code>recalibrated_bam_index</code> (File)</li>
<li><code>variant_filtered_vcf</code> (File)</li>
<li><code>variant_filtered_vcf_index</code> (File)</li>
</ul>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../quality-check-standard/" class="btn btn-neutral float-left" title="Quality check standard"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../samtools-merge/" class="btn btn-neutral float-right" title="Samtools merge">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
      <p>© 2020-Present St. Jude Children's Research Hospital</p>
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/stjudecloud/workflows" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../quality-check-standard/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../samtools-merge/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "../..";</script>
    <script src="../../js/theme_extra.js"></script>
    <script src="../../js/theme.js"></script>
      <script src="../../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
