<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="author" content="St. Jude Cloud" />
      <link rel="shortcut icon" href="../../img/favicon.ico" />
    <title>Quality check standard - St. Jude Cloud Bioinformatics Workflows</title>
    <link rel="stylesheet" href="../../css/theme.css" />
    <link rel="stylesheet" href="../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Quality check standard";
        var mkdocs_page_input_path = "workflows/quality-check-standard.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../.." class="icon icon-home"> St. Jude Cloud Bioinformatics Workflows
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../build_for_dnanexus/">Building WDL workflows for DNAnexus</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Tasks</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/arriba/">Arriba</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/bwa/">Bwa</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/cellranger/">Cellranger</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/deeptools/">Deeptools</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/estimate/">Estimate</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/fastqc/">Fastqc</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/fq/">Fq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/gatk4/">Gatk4</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/htseq/">Htseq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/kraken2/">Kraken2</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/librarian/">librarian</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/md5sum/">Md5sum</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/mosdepth/">Mosdepth</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/multiqc/">Multiqc</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/ngsderive/">Ngsderive</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/picard/">Picard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/qualimap/">Qualimap</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/read_group/">Read group</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/sambamba/">Sambamba</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/samtools/">Samtools</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/star/">Star</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/util/">Utilities</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Workflows</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../10x-bam-to-fastqs/">Cell Ranger BAM to FASTQs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ESTIMATE/">ESTIMATE</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../bam-to-fastqs/">Bam to fastqs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../bwa-db-build/">Bwa db build</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../dnaseq-core/">Dnaseq core</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../dnaseq-standard-fastq/">Dnaseq standard fastq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../dnaseq-standard/">Dnaseq standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../flag_filter/">FlagFilter</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../gatk-reference/">Gatk reference</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../make-qc-reference/">Make qc reference</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../markdups-post/">MarkDuplicates Post</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="#">Quality check standard</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#inputs">Inputs</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#required">Required</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#optional">Optional</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#defaults">Defaults</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#outputs">Outputs</a>
    </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../rnaseq-variant-calling/">Rnaseq variant calling</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../samtools-merge/">Samtools merge</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../scrnaseq-standard/">Scrnaseq standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../star-db-build/">Star db build</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../..">St. Jude Cloud Bioinformatics Workflows</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../.." class="icon icon-home" aria-label="Docs"></a></li>
          <li class="breadcrumb-item">Workflows</li>
      <li class="breadcrumb-item active">Quality check standard</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/stjudecloud/workflows/edit/master/docs/workflows/quality-check-standard.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h2 id="quality_check">quality_check</h2>
<dl>
<dt>description</dt>
<dd>Performs comprehensive quality checks, aggregating all analyses and metrics into a final MultiQC report.</dd>
<dt>help</dt>
<dd>Assumes that input BAM is position-sorted.</dd>
<dt>external_help</dt>
<dd>https://multiqc.info/</dd>
<dt>outputs</dt>
<dd>{'bam_checksum': 'STDOUT of the <code>md5sum</code> command run on the input BAM that has been redirected to a file', 'validate_sam_file': 'Validation report produced by <code>picard ValidateSamFile</code>. Validation warnings and errors are logged.', 'flagstat_report': '<code>samtools flagstat</code> STDOUT redirected to a file. If <code>mark_duplicates</code> is <code>true</code>, then this result will be generated from the duplicate marked BAM.', 'fastqc_results': 'A gzipped tar archive of all FastQC output files', 'instrument_file': 'TSV file containing the <code>ngsderive isntrument</code> report for the input BAM file', 'read_length_file': 'TSV file containing the <code>ngsderive readlen</code> report for the input BAM file', 'inferred_encoding': 'TSV file containing the <code>ngsderive encoding</code> report for the input BAM file', 'alignment_metrics': {'description': 'The text file output of <code>picard CollectAlignmentSummaryMetrics</code>', 'external_help': 'http://broadinstitute.github.io/picard/picard-metric-definitions.html#AlignmentSummaryMetrics'}, 'alignment_metrics_pdf': 'The PDF file output of <code>picard CollectAlignmentSummaryMetrics</code>', 'insert_size_metrics': {'description': 'The text file output of <code>picard CollectInsertSizeMetrics</code>. If <code>mark_duplicates</code> is <code>true</code>, then this result will be generated from the duplicate marked BAM.', 'external_help': 'http://broadinstitute.github.io/picard/picard-metric-definitions.html#InsertSizeMetrics'}, 'insert_size_metrics_pdf': 'The PDF file output of <code>picard CollectInsertSizeMetrics</code>. If <code>mark_duplicates</code> is <code>true</code>, then this result will be generated from the duplicate marked BAM.', 'quality_score_distribution_txt': 'The text file output of <code>picard QualityScoreDistribution</code>', 'quality_score_distribution_pdf': 'The PDF file output of <code>picard QualityScoreDistribution</code>', 'phred_scores': 'Headered TSV file containing PHRED score statistics', 'kraken_report': {'description': 'A Kraken2 summary report', 'external_help': 'https://github.com/DerrickWood/kraken2/blob/master/docs/MANUAL.markdown#sample-report-output-format'}, 'mosdepth_global_dist': 'The <code>$prefix.mosdepth.global.dist.txt</code> file contains a cumulative distribution indicating the proportion of total bases that were covered for at least a given coverage value. It does this for each chromosome, and for the whole genome.', 'mosdepth_global_summary': 'A summary of mean depths per chromosome', 'mosdepth_region_dist': 'The <code>$prefix.mosdepth.region.dist.txt</code> file contains a cumulative distribution indicating the proportion of total bases in the region(s) defined by the <code>coverage_bed</code> that were covered for at least a given coverage value. There will be one file in this array for each <code>coverage_beds</code> input file.', 'mosdepth_region_summary': 'A summary of mean depths per chromosome and within specified regions per chromosome. There will be one file in this array for each <code>coverage_beds</code> input file.', 'multiqc_report': 'A gzipped tar archive of all MultiQC output files', 'orig_read_count': 'A TSV report containing the original read count before subsampling. Only present if <code>subsample_n_reads &gt; 0</code>.', 'kraken_sequences': {'description': 'Detailed Kraken2 output that has been gzipped. Only present if <code>store_kraken_sequences == true</code>.', 'external_help': 'https://github.com/DerrickWood/kraken2/blob/master/docs/MANUAL.markdown#standard-kraken-output-format'}, 'comparative_kraken_report': 'Kraken2 summary report for only the alternatively filtered reads. Only present if <code>run_comparative_kraken == true</code>.', 'comparative_kraken_sequences': 'Detailed Kraken2 output for only the alternatively filtered reads. Only present if <code>run_comparative_kraken == true &amp;&amp; store_kraken_sequences == true</code>.', 'mark_duplicates_metrics': {'description': 'The METRICS_FILE result of <code>picard MarkDuplicates</code>. Only present if <code>mark_duplicates == true &amp;&amp; optical_distance &gt; 0</code>.', 'external_help': 'http://broadinstitute.github.io/picard/picard-metric-definitions.html#DuplicationMetrics'}, 'mosdepth_dups_marked_global_dist': 'The <code>$prefix.mosdepth.global.dist.txt</code> file contains a cumulative distribution indicating the proportion of total bases that were covered for at least a given coverage value. It does this for each chromosome, and for the whole genome. This file is produced from analyzing the duplicate marked BAM. Only present if <code>mark_duplicates == true</code>.', 'mosdepth_dups_marked_global_summary': 'A summary of mean depths per chromosome. This file is produced from analyzing the duplicate marked BAM. Only present if <code>mark_duplicates == true</code>.', 'mosdepth_dups_marked_region_dist': 'The <code>$prefix.mosdepth.region.dist.txt</code> file contains a cumulative distribution indicating the proportion of total bases in the region(s) defined by the <code>coverage_bed</code> that were covered for at least a given coverage value. There will be one file in this array for each <code>coverage_beds</code> input file. This file is produced from analyzing the duplicate marked BAM. Only present if <code>mark_duplicates == true</code>.', 'mosdepth_dups_marked_region_summary': 'A summary of mean depths per chromosome and within specified regions per chromosome. There will be one file in this array for each <code>coverage_beds</code> input file. This file is produced from analyzing the duplicate marked BAM. Only present if <code>mark_duplicates == true</code>.', 'inferred_strandedness': 'TSV file containing the <code>ngsderive strandedness</code> report. Only present if <code>rna == true</code>.', 'qualimap_rnaseq_results': 'Gzipped tar archive of all QualiMap output files. Only present if <code>rna == true</code>.', 'junction_summary': 'TSV file containing the <code>ngsderive junction-annotation</code> summary. Only present if <code>rna == true</code>', 'junctions': 'TSV file containing a detailed list of annotated junctions. Only present if <code>rna == true</code>.', 'librarian_report': 'A tar archive containing the <code>librarian</code> report and raw data. Only present if <code>run_librarian == true</code>.', 'IntermediateFiles': 'Any and all files produced as intermediate during pipeline processing. Only output if <code>output_intermediate_files == true</code>.'}</dd>
<dt>allowNestedInputs</dt>
<dd>true</dd>
</dl>
<h3 id="inputs">Inputs</h3>
<h4 id="required">Required</h4>
<ul>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file to quality check</li>
<li><code>bam_index</code> (File, <strong>required</strong>): BAM index file corresponding to the input BAM</li>
<li><code>kraken_db</code> (File, <strong>required</strong>): Kraken2 database. Can be generated with <code>../reference/make-qc-reference.wdl</code>. Must be a tarball without a root directory.</li>
<li><code>alt_filtered_fastq._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>alt_filtered_fqlint._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>bam_to_fastq._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>collect_alignment_summary_metrics._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>collect_insert_size_metrics._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>comparative_kraken._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>compression_integrity._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>compute_checksum._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>encoding._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>endedness._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>fastqc._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>flagstat._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>fqlint._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>global_phred_scores._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>instrument._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>junction_annotation._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>kraken._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>librarian._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>markdups._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>multiqc._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>parse_input._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>qualimap_rnaseq._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>quality_score_distribution._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>quickcheck._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>read_length._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>regions_coverage._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>strandedness._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>subsample._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>subsample_index._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>validate_bam._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>wg_coverage._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>comparative_kraken_filter_validator.validate_exclude_if_all._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>comparative_kraken_filter_validator.validate_exclude_if_any._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>comparative_kraken_filter_validator.validate_include_if_all._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>comparative_kraken_filter_validator.validate_include_if_any._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>kraken_filter_validator.validate_exclude_if_all._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>kraken_filter_validator.validate_exclude_if_any._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>kraken_filter_validator.validate_include_if_all._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>kraken_filter_validator.validate_include_if_any._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>markdups_post.collect_insert_size_metrics._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>markdups_post.flagstat._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>markdups_post.regions_coverage._runtime</code> (Any, <strong>required</strong>)</li>
<li><code>markdups_post.wg_coverage._runtime</code> (Any, <strong>required</strong>)</li>
</ul>
<h4 id="optional">Optional</h4>
<ul>
<li><code>gtf</code> (File?): GTF features file. Gzipped or uncompressed. <strong>Required</strong> for RNA-Seq data.</li>
<li><code>validate_bam.reference_fasta</code> (File?)</li>
<li><code>wg_coverage.coverage_bed</code> (File?)</li>
<li><code>markdups_post.wg_coverage.coverage_bed</code> (File?)</li>
</ul>
<h4 id="defaults">Defaults</h4>
<ul>
<li><code>comparative_filter</code> (FlagFilter, default={"include_if_all": "0x0", "exclude_if_any": "0x904", "include_if_any": "0x0", "exclude_if_all": "0x0"}): Filter to apply to the input BAM while performing a second FASTQ conversion, before running Kraken2 another time. This is a <code>FlagFilter</code> object (see ../../data_structures/flag_filter.wdl for more information). By default, it will <strong>remove unmapped, secondary, and supplementary reads</strong> from the created FASTQs. <strong>WARNING</strong> These filters can be tricky to configure; please read documentation thoroughly before changing the defaults.</li>
<li><code>coverage_beds</code> (Array[File], default=[]): An array of 3 column BEDs which are passed to the <code>-b</code> flag of mosdepth, in order to restrict coverage analysis to select regions. Any regional analysis enabled by this option is <em>in addition</em> to whole genome coverage, which is calculated regardless of this setting. An exon BED and a Coding Sequence BED are examples of regions you may wish to restrict coverage analysis to. Those two BEDs can be created with the workflow in <code>../reference/make-qc-reference.wdl</code>.</li>
<li><code>coverage_labels</code> (Array[String], default=[]): An array of equal length to <code>coverage_beds</code> which determines the prefix label applied to the output files. If omitted, defaults of <code>regions1</code>, <code>regions2</code>, etc. will be used. If using the BEDs created by <code>../reference/make-qc-reference.wdl</code>, the labels ["exon", "CDS"] are appropriate. Make sure to provide the coverage BEDs <strong>in the same order</strong> as the labels.</li>
<li><code>extra_multiqc_inputs</code> (Array[File], default=[]): An array of additional files to pass directly into MultiQC</li>
<li><code>mark_duplicates</code> (Boolean, default=rna): Mark duplicates before select analyses? Default behavior is to set this to the value of the <code>rna</code> parameter. This is because DNA files are often duplicate marked already, and RNA-Seq files are usually <em>not</em> duplicate marked. If set to <code>true</code>, a BAM will be generated and passed to selected downstream analyses. For more details about what analyses are run, review <code>./markdups-post.wdl</code>. <strong>WARNING, this duplicate marked BAM is <em>not</em> ouput by default.</strong> If you would like to output this file, set <code>output_intermediate_files = true</code>.</li>
<li><code>multiqc_config</code> (File, default="https://raw.githubusercontent.com/stjudecloud/workflows/main/workflows/qc/inputs/multiqc_config_hg38.yaml"): YAML file for configuring MultiQC</li>
<li><code>optical_distance</code> (Int, default=0): Maximum distance between read coordinates to consider them optical duplicates instead of library duplicates (e.g. PCR duplicates). If <code>mark_duplicates == false</code>, this parameter is ignored. If <code>0</code>, then <em>optical</em> duplicate marking is disabled and only traditional duplicate marking will be performed. Suggested settings of 100 for unpatterned versions of the Illumina platform (e.g. HiSeq) or 2500 for patterned flowcell models (e.g. NovaSeq). Calculation of distance depends on coordinate data embedded in the read names, typically produced by Illumina sequencing machines. Optical duplicate detection will not work on non-standard names without a custom regex for tile-data extraction. Review the <code>mark_duplicates</code> task in <code>../../tools/picard.wdl</code> for more information.</li>
<li><code>output_intermediate_files</code> (Boolean, default=false): Output intermediate files? FASTQs; if <code>rna == true</code> a collated BAM; if <code>mark_duplicates == true</code> a duplicate marked BAM, various accessory files like indexes and md5sums; if subsampling was requested <em>and</em> performed then a sampled BAM and associated index. <strong>WARNING, these files can be large.</strong></li>
<li><code>prefix</code> (String, default=basename(bam,".bam")): Prefix for all results files</li>
<li><code>rna</code> (Boolean, default=false): Is the sequenced molecule RNA? Enabling this option adds RNA-Seq specific analyses to the workflow. If <code>true</code>, a GTF file must be provided. If <code>false</code>, the GTF file is ignored.</li>
<li><code>run_comparative_kraken</code> (Boolean, default=false): Run Kraken2 a second time with different FASTQ filtering? If <code>true</code>, <code>comparative_filter</code> is used in a second run of BAM-&gt;FASTQ conversion, resulting in differently filtered FASTQs analyzed by Kraken2. If <code>false</code>, <code>comparative_filter</code> is ignored.</li>
<li><code>run_librarian</code> (Boolean, default=rna); <strong>description</strong>: Run the <code>librarian</code> tool to generate a report of the likely Illumina library prep kit used to generate the data. <strong>WARNING</strong> this tool is not guaranteed to work on all data, and may produce nonsensical results. <code>librarian</code> was trained on a limited set of GEO read data (Gene Expression Oriented). This means the input data should be Paired-End, of mouse or human origin, read length should be &gt;50bp, and derived from a library prep kit that is in the <code>librarian</code> database. By default, this tool is run when <code>rna == true</code>.; <strong>external_help</strong>: https://f1000research.com/articles/11-1122/v2</li>
<li><code>standard_filter</code> (FlagFilter, default={"include_if_all": "0x0", "exclude_if_any": "0x900", "include_if_any": "0x0", "exclude_if_all": "0x0"}): Filter to apply to the input BAM while converting to FASTQ, before running Kraken2 and <code>librarian</code> (if <code>run_librarian == true</code>). This is a <code>FlagFilter</code> object (see ../../data_structures/flag_filter.wdl for more information). By default, it will <strong>remove secondary and supplementary reads</strong> from the created FASTQs. <strong>WARNING:</strong> These filters can be tricky to configure; please read documentation thoroughly before changing the defaults. <strong>WARNING:</strong> If you have set <code>run_librarian</code> to <code>true</code>, we <strong>strongly</strong> recommend leaving this filter at the default value. <code>librarian</code> is trained on a specific set of reads, and changing this filter may produce nonsensical results.</li>
<li><code>store_kraken_sequences</code> (Boolean, default=false): Store the Kraken2 sequences output? This will apply to all runs of Kraken2 (see <code>parameter_meta.run_comparative_kraken</code>). <strong>WARNING</strong> these files can be very large.</li>
<li><code>subsample_n_reads</code> (Int, default=-1): Only process a random sampling of approximately <code>n</code> reads. Any <code>n &lt;= 0</code> for processing entire input. Subsampling is done probabalistically so the exact number of reads in the output will have some variation.</li>
<li><code>use_all_cores</code> (Boolean, default=false): Use all cores? Recommended for cloud environments.</li>
<li><code>alt_filtered_fastq.append_read_number</code> (Boolean, default=true)</li>
<li><code>alt_filtered_fastq.collated</code> (Boolean, default=false)</li>
<li><code>alt_filtered_fastq.fail_on_unexpected_reads</code> (Boolean, default=false)</li>
<li><code>alt_filtered_fastq.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>alt_filtered_fastq.modify_memory_gb</code> (Int, default=0)</li>
<li><code>alt_filtered_fastq.ncpu</code> (Int, default=2)</li>
<li><code>alt_filtered_fastq.output_singletons</code> (Boolean, default=false)</li>
<li><code>alt_filtered_fqlint.disable_validator_codes</code> (Array[String], default=[])</li>
<li><code>alt_filtered_fqlint.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>alt_filtered_fqlint.modify_memory_gb</code> (Int, default=0)</li>
<li><code>alt_filtered_fqlint.paired_read_validation_level</code> (String, default="high")</li>
<li><code>alt_filtered_fqlint.panic</code> (Boolean, default=true)</li>
<li><code>alt_filtered_fqlint.single_read_validation_level</code> (String, default="high")</li>
<li><code>bam_to_fastq.append_read_number</code> (Boolean, default=true)</li>
<li><code>bam_to_fastq.collated</code> (Boolean, default=false)</li>
<li><code>bam_to_fastq.fail_on_unexpected_reads</code> (Boolean, default=false)</li>
<li><code>bam_to_fastq.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>bam_to_fastq.modify_memory_gb</code> (Int, default=0)</li>
<li><code>bam_to_fastq.ncpu</code> (Int, default=2)</li>
<li><code>bam_to_fastq.output_singletons</code> (Boolean, default=false)</li>
<li><code>collect_alignment_summary_metrics.memory_gb</code> (Int, default=8)</li>
<li><code>collect_alignment_summary_metrics.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>collect_alignment_summary_metrics.validation_stringency</code> (String, default="SILENT")</li>
<li><code>collect_insert_size_metrics.memory_gb</code> (Int, default=8)</li>
<li><code>collect_insert_size_metrics.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>collect_insert_size_metrics.validation_stringency</code> (String, default="SILENT")</li>
<li><code>comparative_kraken.min_base_quality</code> (Int, default=0)</li>
<li><code>comparative_kraken.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>comparative_kraken.modify_memory_gb</code> (Int, default=0)</li>
<li><code>comparative_kraken.ncpu</code> (Int, default=4)</li>
<li><code>comparative_kraken.use_names</code> (Boolean, default=true)</li>
<li><code>compression_integrity.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>compute_checksum.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>encoding.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>endedness.calc_rpt</code> (Boolean, default=false)</li>
<li><code>endedness.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>endedness.modify_memory_gb</code> (Int, default=0)</li>
<li><code>endedness.num_reads</code> (Int, default=-1)</li>
<li><code>endedness.paired_deviance</code> (Float, default=0.0)</li>
<li><code>endedness.round_rpt</code> (Boolean, default=false)</li>
<li><code>endedness.split_by_rg</code> (Boolean, default=false)</li>
<li><code>fastqc.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>fastqc.ncpu</code> (Int, default=4)</li>
<li><code>flagstat.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>flagstat.ncpu</code> (Int, default=2)</li>
<li><code>flagstat.use_all_cores</code> (Boolean, default=false): Use all cores? Recommended for cloud environments.</li>
<li><code>fqlint.disable_validator_codes</code> (Array[String], default=[])</li>
<li><code>fqlint.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>fqlint.modify_memory_gb</code> (Int, default=0)</li>
<li><code>fqlint.paired_read_validation_level</code> (String, default="high")</li>
<li><code>fqlint.panic</code> (Boolean, default=true)</li>
<li><code>fqlint.single_read_validation_level</code> (String, default="high")</li>
<li><code>global_phred_scores.fast_mode</code> (Boolean, default=true)</li>
<li><code>global_phred_scores.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>instrument.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>instrument.num_reads</code> (Int, default=10000)</li>
<li><code>junction_annotation.fuzzy_junction_match_range</code> (Int, default=0)</li>
<li><code>junction_annotation.min_intron</code> (Int, default=50)</li>
<li><code>junction_annotation.min_mapq</code> (Int, default=30)</li>
<li><code>junction_annotation.min_reads</code> (Int, default=2)</li>
<li><code>junction_annotation.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>kraken.min_base_quality</code> (Int, default=0)</li>
<li><code>kraken.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>kraken.modify_memory_gb</code> (Int, default=0)</li>
<li><code>kraken.ncpu</code> (Int, default=4)</li>
<li><code>kraken.use_names</code> (Boolean, default=true)</li>
<li><code>librarian.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>librarian.prefix</code> (String, default=sub(basename(read_one_fastq),"([_\.][rR][12])?(\.subsampled)?\.(fastq|fq)(\.gz)?$","") + ".librarian"): Prefix for all results files</li>
<li><code>markdups.clear_dt</code> (Boolean, default=true)</li>
<li><code>markdups.duplicate_scoring_strategy</code> (String, default="SUM_OF_BASE_QUALITIES")</li>
<li><code>markdups.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>markdups.modify_memory_gb</code> (Int, default=0)</li>
<li><code>markdups.read_name_regex</code> (String, default="^[!-9;-?A-~:]+:([!-9;-?A-~]+):([0-9]+):([0-9]+)$")</li>
<li><code>markdups.remove_duplicates</code> (Boolean, default=false)</li>
<li><code>markdups.remove_sequencing_duplicates</code> (Boolean, default=false)</li>
<li><code>markdups.tagging_policy</code> (String, default="All")</li>
<li><code>markdups.validation_stringency</code> (String, default="SILENT")</li>
<li><code>multiqc.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>qualimap_rnaseq.memory_gb</code> (Int, default=16)</li>
<li><code>qualimap_rnaseq.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>quality_score_distribution.memory_gb</code> (Int, default=8)</li>
<li><code>quality_score_distribution.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>quality_score_distribution.validation_stringency</code> (String, default="SILENT")</li>
<li><code>quickcheck.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>read_length.majority_vote_cutoff</code> (Float, default=0.7)</li>
<li><code>read_length.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>read_length.num_reads</code> (Int, default=-1)</li>
<li><code>regions_coverage.min_mapping_quality</code> (Int, default=20)</li>
<li><code>regions_coverage.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>regions_coverage.use_fast_mode</code> (Boolean, default=true)</li>
<li><code>strandedness.min_mapq</code> (Int, default=30)</li>
<li><code>strandedness.min_reads_per_gene</code> (Int, default=10)</li>
<li><code>strandedness.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>strandedness.num_genes</code> (Int, default=1000)</li>
<li><code>strandedness.split_by_rg</code> (Boolean, default=false)</li>
<li><code>subsample.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>subsample.ncpu</code> (Int, default=2)</li>
<li><code>subsample_index.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>subsample_index.ncpu</code> (Int, default=2)</li>
<li><code>validate_bam.index_validation_stringency_less_exhaustive</code> (Boolean, default=false)</li>
<li><code>validate_bam.max_errors</code> (Int, default=2147483647)</li>
<li><code>validate_bam.memory_gb</code> (Int, default=16)</li>
<li><code>validate_bam.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>validate_bam.succeed_on_warnings</code> (Boolean, default=true)</li>
<li><code>validate_bam.validation_stringency</code> (String, default="LENIENT")</li>
<li><code>wg_coverage.min_mapping_quality</code> (Int, default=20)</li>
<li><code>wg_coverage.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>wg_coverage.use_fast_mode</code> (Boolean, default=true)</li>
<li><code>markdups_post.collect_insert_size_metrics.memory_gb</code> (Int, default=8)</li>
<li><code>markdups_post.collect_insert_size_metrics.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>markdups_post.collect_insert_size_metrics.validation_stringency</code> (String, default="SILENT")</li>
<li><code>markdups_post.flagstat.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>markdups_post.flagstat.ncpu</code> (Int, default=2)</li>
<li><code>markdups_post.flagstat.use_all_cores</code> (Boolean, default=false): Use all cores? Recommended for cloud environments.</li>
<li><code>markdups_post.regions_coverage.min_mapping_quality</code> (Int, default=20)</li>
<li><code>markdups_post.regions_coverage.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>markdups_post.regions_coverage.use_fast_mode</code> (Boolean, default=true)</li>
<li><code>markdups_post.wg_coverage.min_mapping_quality</code> (Int, default=20)</li>
<li><code>markdups_post.wg_coverage.modify_disk_size_gb</code> (Int, default=0)</li>
<li><code>markdups_post.wg_coverage.use_fast_mode</code> (Boolean, default=true)</li>
</ul>
<h3 id="outputs">Outputs</h3>
<ul>
<li><code>bam_checksum</code> (File)</li>
<li><code>validate_sam_file</code> (File)</li>
<li><code>flagstat_report</code> (File)</li>
<li><code>fastqc_results</code> (File)</li>
<li><code>instrument_file</code> (File)</li>
<li><code>read_length_file</code> (File)</li>
<li><code>inferred_encoding</code> (File)</li>
<li><code>inferred_endedness</code> (File)</li>
<li><code>alignment_metrics</code> (File)</li>
<li><code>alignment_metrics_pdf</code> (File)</li>
<li><code>insert_size_metrics</code> (File)</li>
<li><code>insert_size_metrics_pdf</code> (File)</li>
<li><code>quality_score_distribution_txt</code> (File)</li>
<li><code>quality_score_distribution_pdf</code> (File)</li>
<li><code>phred_scores</code> (File)</li>
<li><code>kraken_report</code> (File)</li>
<li><code>mosdepth_global_dist</code> (File)</li>
<li><code>mosdepth_global_summary</code> (File)</li>
<li><code>mosdepth_region_dist</code> (Array[File])</li>
<li><code>mosdepth_region_summary</code> (Array[File])</li>
<li><code>multiqc_report</code> (File)</li>
<li><code>orig_read_count</code> (File?)</li>
<li><code>kraken_sequences</code> (File?)</li>
<li><code>comparative_kraken_report</code> (File?)</li>
<li><code>comparative_kraken_sequences</code> (File?)</li>
<li><code>mosdepth_dups_marked_global_dist</code> (File?)</li>
<li><code>mosdepth_dups_marked_global_summary</code> (File?)</li>
<li><code>mosdepth_dups_marked_region_summary</code> (Array[File]?)</li>
<li><code>mosdepth_dups_marked_region_dist</code> (Array[File?]?)</li>
<li><code>mark_duplicates_metrics</code> (File?)</li>
<li><code>inferred_strandedness</code> (File?)</li>
<li><code>qualimap_rnaseq_results</code> (File?)</li>
<li><code>junction_summary</code> (File?)</li>
<li><code>junctions</code> (File?)</li>
<li><code>librarian_report</code> (File?)</li>
<li><code>intermediate_files</code> (IntermediateFiles?)</li>
</ul>
<h2 id="parse_input">parse_input</h2>
<dl>
<dt>description</dt>
<dd>Parses and validates the <code>quality_check</code> workflow's provided inputs</dd>
<dt>outputs</dt>
<dd>{'check': 'Dummy output to indicate success and to enable call-caching', 'labels': 'An array of labels to use on the result coverage files associated with each coverage BED'}</dd>
</dl>
<h3 id="inputs_1">Inputs</h3>
<h4 id="required_1">Required</h4>
<ul>
<li><code>_runtime</code> (Any, <strong>required</strong>)</li>
<li><code>coverage_beds_len</code> (Int, <strong>required</strong>): Length of the provided <code>coverage_beds</code> array</li>
<li><code>coverage_labels</code> (Array[String], <strong>required</strong>): An array of equal length to <code>coverage_beds_len</code> which determines the prefix label applied to coverage output files. If an empty array is supplied, defaults of <code>regions1</code>, <code>regions2</code>, etc. will be used.</li>
<li><code>gtf_provided</code> (Boolean, <strong>required</strong>): Was a GTF supplied by the user? Must be <code>true</code> if <code>rna == true</code>.</li>
<li><code>rna</code> (Boolean, <strong>required</strong>): Is the sequenced molecule RNA?</li>
</ul>
<h3 id="outputs_1">Outputs</h3>
<ul>
<li><code>labels</code> (Array[String])</li>
</ul>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../markdups-post/" class="btn btn-neutral float-left" title="MarkDuplicates Post"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../rnaseq-variant-calling/" class="btn btn-neutral float-right" title="Rnaseq variant calling">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
      <p>© 2020-Present St. Jude Children's Research Hospital</p>
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/stjudecloud/workflows" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../markdups-post/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../rnaseq-variant-calling/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "../..";</script>
    <script src="../../js/theme_extra.js"></script>
    <script src="../../js/theme.js"></script>
      <script src="../../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
