<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="author" content="St. Jude Cloud" />
      <link rel="shortcut icon" href="../../img/favicon.ico" />
    <title>Quality Check Standard - St. Jude Cloud Bioinformatics Workflows</title>
    <link rel="stylesheet" href="../../css/theme.css" />
    <link rel="stylesheet" href="../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Quality Check Standard";
        var mkdocs_page_input_path = "workflows/quality-check-standard.md";
        var mkdocs_page_url = null;
      </script>
    
    <script src="../../js/jquery-3.6.0.min.js" defer></script>
    <!--[if lt IE 9]>
      <script src="../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
      <script>hljs.initHighlightingOnLoad();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../.." class="icon icon-home"> St. Jude Cloud Bioinformatics Workflows
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../build_for_dnanexus/">Building WDL workflows for DNAnexus</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Tasks</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/bwa/">BWA</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/cellranger/">Cellranger</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/deeptools/">Deeptools</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/estimate/">ESTIMATE</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/fastq_screen/">FastQ Screen</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/fastqc/">FastQC</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/fq/">FQ</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/htseq/">HTSeq</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/md5sum/">Md5sum</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/mosdepth/">Mosdepth</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/multiqc/">MultiQC</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/ngsderive/">ngsderive</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/picard/">Picard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/qualimap/">QualiMap</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/samtools/">SAMtools</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/star/">STAR</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../tasks/util/">Utilities</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Workflows</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../10x-bam-to-fastqs/">Cell Ranger Bam to FastQs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../ESTIMATE/">ESTIMATE</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../bam-to-fastqs/">Bam to FastQs</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../bwa-db-build/">BWA DB build</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../calc-gene-lengths/">Calculate Gene Lengths</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../chipseq-bwa-db-build/">BWA DB build</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../make-qc-reference/">Make QC Reference</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="./">Quality Check Standard</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#licensing">LICENSING</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#mit-license">MIT License</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#quality_check">quality_check</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#inputs">Inputs</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#required">Required</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#optional">Optional</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#defaults">Defaults</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#outputs">Outputs</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#parse_input">parse_input</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#inputs_1">Inputs</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#required_1">Required</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#optional_1">Optional</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#outputs_1">Outputs</a>
    </li>
        </ul>
    </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../rnaseq-standard-fastq/">RNA-Seq Standard from FastQ</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../rnaseq-standard/">RNA-Seq Standard</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../rnaseq-star-db-build/">RNASeq STAR DB build</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../scrnaseq-standard/">Scrnaseq standard</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../..">St. Jude Cloud Bioinformatics Workflows</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../.." class="icon icon-home" aria-label="Docs"></a> &raquo;</li>
          <li>Workflows &raquo;</li>
      <li>Quality Check Standard</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/stjudecloud/workflows/edit/master/docs/workflows/quality-check-standard.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="quality-check-standard">Quality Check Standard</h1>
<p>This workflow runs a variety of quality checking software on any BAM file.
It can be WGS, WES, or Transcriptome data. The results are aggregated and
run through <a href="https://multiqc.info/">MultiQC</a>.</p>
<h2 id="licensing">LICENSING</h2>
<h4 id="mit-license">MIT License</h4>
<p>Copyright 2020-Present St. Jude Children's Research Hospital</p>
<p>Permission is hereby granted, free of charge, to any person obtaining a copy of this
software and associated documentation files (the "Software"), to deal in the Software
without restriction, including without limitation the rights to use, copy, modify, merge,
publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons
to whom the Software is furnished to do so, subject to the following conditions:</p>
<p>The above copyright notice and this permission notice shall be included in all copies or
substantial portions of the Software.</p>
<p>THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.</p>
<h2 id="quality_check">quality_check</h2>
<h3 id="inputs">Inputs</h3>
<h4 id="required">Required</h4>
<ul>
<li><code>bam</code> (File, <strong>required</strong>): Input BAM format file to quality check</li>
<li><code>bam_index</code> (File, <strong>required</strong>): BAM index file corresponding to the input BAM</li>
<li><code>molecule</code> (String, <strong>required</strong>): 'DNA' or 'RNA'</li>
<li><code>reference_fasta</code> (File, <strong>required</strong>): Reference genome in FASTA format</li>
</ul>
<h4 id="optional">Optional</h4>
<ul>
<li><code>fastq_screen_db</code> (File?): Database for FastQ Screen. <strong>Required</strong> for WGS and WES data. Can be generated using <code>make-qc-reference.wdl</code>. Must untar directly to genome directories.</li>
<li><code>gtf</code> (File?): GTF features file. <strong>Required</strong> for RNA-Seq data</li>
<li><code>star_log</code> (File?): Log file generated by the RNA-Seq aligner STAR</li>
<li><code>collect_alignment_summary_metrics.disk_size_gb</code> (Int?)</li>
<li><code>collect_gc_bias_metrics.disk_size_gb</code> (Int?)</li>
<li><code>collect_insert_size_metrics.disk_size_gb</code> (Int?)</li>
<li><code>fastq_screen.sample_name</code> (String?)</li>
<li><code>picard_sort.disk_size_gb</code> (Int?)</li>
<li><code>qualimap_rnaseq.disk_size_gb</code> (Int?)</li>
<li><code>quality_score_distribution.disk_size_gb</code> (Int?)</li>
<li><code>wg_coverage.coverage_bed</code> (File?)</li>
</ul>
<h4 id="defaults">Defaults</h4>
<ul>
<li><code>coverage_beds</code> (Array[File], default=[]): An array of 3 column BEDs which are passed to the <code>-b</code> flag of mosdepth, in order to restrict coverage analysis to select regions</li>
<li><code>coverage_labels</code> (Array[String], default=[]): An array of equal length to <code>coverage_beds</code> which determines the prefix label applied to the output files. If omitted, defaults of <code>regions1</code>, <code>regions2</code>, etc. will be used.</li>
<li><code>max_retries</code> (Int, default=1): Number of times to retry failed steps</li>
<li><code>paired_end</code> (Boolean, default=true): Whether the data is paired end</li>
<li><code>phred_encoding</code> (String, default=""): Encoding format used for PHRED quality scores. Must be empty, 'sanger', or 'illumina1.3'. Only needed for WGS/WES. If missing, will be inferred</li>
<li><code>strandedness</code> (String, default=""): empty, 'Stranded-Reverse', 'Stranded-Forward', or 'Unstranded'. Only needed for RNA-Seq data. If missing, will be inferred</li>
<li><code>bam_to_fastq.memory_gb</code> (Int, default=40)</li>
<li><code>bam_to_fastq.prefix</code> (String, default=basename(bam,".bam"))</li>
<li><code>collect_alignment_summary_metrics.memory_gb</code> (Int, default=8)</li>
<li><code>collect_gc_bias_metrics.memory_gb</code> (Int, default=8)</li>
<li><code>collect_insert_size_metrics.memory_gb</code> (Int, default=8)</li>
<li><code>compute_checksum.memory_gb</code> (Int, default=5)</li>
<li><code>fastq_screen.num_reads</code> (Int, default=0)</li>
<li><code>fastqc.memory_gb</code> (Int, default=5)</li>
<li><code>fastqc.ncpu</code> (Int, default=1)</li>
<li><code>fqlint.memory_gb</code> (Int, default=8)</li>
<li><code>junction_annotation.fuzzy_junction_match_range</code> (Int, default=0)</li>
<li><code>junction_annotation.memory_gb</code> (Int, default=35)</li>
<li><code>junction_annotation.min_intron</code> (Int, default=50)</li>
<li><code>junction_annotation.min_mapq</code> (Int, default=30)</li>
<li><code>junction_annotation.min_reads</code> (Int, default=2)</li>
<li><code>junction_annotation.prefix</code> (String, default=basename(bam,".bam"))</li>
<li><code>multiqc.disk_size</code> (Int, default=20)</li>
<li><code>multiqc.memory_gb</code> (Int, default=5)</li>
<li><code>ngsderive_encoding.memory_gb</code> (Int, default=5)</li>
<li><code>ngsderive_encoding.num_reads</code> (Int, default=-1)</li>
<li><code>ngsderive_read_length.memory_gb</code> (Int, default=5)</li>
<li><code>ngsderive_strandedness.memory_gb</code> (Int, default=5)</li>
<li><code>picard_sort.memory_gb</code> (Int, default=25)</li>
<li><code>picard_sort.output_filename</code> (String, default=basename(bam,".bam") + ".sorted.bam")</li>
<li><code>qualimap_rnaseq.memory_gb</code> (Int, default=16)</li>
<li><code>quality_score_distribution.memory_gb</code> (Int, default=8)</li>
<li><code>regions_coverage.memory_gb</code> (Int, default=8)</li>
<li><code>regions_coverage.min_mapping_quality</code> (Int, default=20)</li>
<li><code>regions_coverage.use_fast_mode</code> (Boolean, default=true)</li>
<li><code>samtools_flagstat.memory_gb</code> (Int, default=5)</li>
<li><code>samtools_flagstat.outfilename</code> (String, default=basename(bam,".bam") + ".flagstat.txt")</li>
<li><code>samtools_subsample.desired_reads</code> (Int, default=500000)</li>
<li><code>samtools_subsample.detect_nproc</code> (Boolean, default=false)</li>
<li><code>samtools_subsample.ncpu</code> (Int, default=1)</li>
<li><code>samtools_subsample.outname</code> (String, default=basename(bam,".bam") + ".subsampled.bam")</li>
<li><code>validate_bam.index_validation_stringency_less_exhaustive</code> (Boolean, default=false)</li>
<li><code>validate_bam.max_errors</code> (Int, default=2147483647)</li>
<li><code>validate_bam.memory_gb</code> (Int, default=8)</li>
<li><code>validate_bam.output_filename</code> (String, default=basename(bam,".bam") + ".ValidateSamFile.txt")</li>
<li><code>validate_bam.succeed_on_warnings</code> (Boolean, default=true)</li>
<li><code>wg_coverage.memory_gb</code> (Int, default=8)</li>
<li><code>wg_coverage.min_mapping_quality</code> (Int, default=20)</li>
<li><code>wg_coverage.use_fast_mode</code> (Boolean, default=true)</li>
</ul>
<h3 id="outputs">Outputs</h3>
<ul>
<li><code>bam_checksum</code> (File)</li>
<li><code>validate_sam_file</code> (File)</li>
<li><code>flagstat</code> (File)</li>
<li><code>fastqc_results</code> (File)</li>
<li><code>instrument_file</code> (File)</li>
<li><code>read_length_file</code> (File)</li>
<li><code>inferred_encoding</code> (File)</li>
<li><code>alignment_metrics</code> (File)</li>
<li><code>alignment_metrics_pdf</code> (File)</li>
<li><code>gc_bias_metrics</code> (File)</li>
<li><code>gc_bias_metrics_pdf</code> (File)</li>
<li><code>insert_size_metrics</code> (File)</li>
<li><code>insert_size_metrics_pdf</code> (File)</li>
<li><code>quality_score_distribution_txt</code> (File)</li>
<li><code>quality_score_distribution_pdf</code> (File)</li>
<li><code>multiqc_zip</code> (File)</li>
<li><code>mosdepth_global_dist</code> (File)</li>
<li><code>mosdepth_global_summary</code> (File)</li>
<li><code>mosdepth_region_dist</code> (Array[File])</li>
<li><code>mosdepth_region_summary</code> (Array[File])</li>
<li><code>fastq_screen_results</code> (File?)</li>
<li><code>inferred_strandedness</code> (File?)</li>
<li><code>qualimap_rnaseq_results</code> (File?)</li>
<li><code>junction_summary</code> (File?)</li>
<li><code>junctions</code> (File?)</li>
</ul>
<h2 id="parse_input">parse_input</h2>
<h3 id="inputs_1">Inputs</h3>
<h4 id="required_1">Required</h4>
<ul>
<li><code>coverage_beds_len</code> (Int, <strong>required</strong>)</li>
<li><code>coverage_labels</code> (Array[String], <strong>required</strong>)</li>
<li><code>input_fq_format</code> (String, <strong>required</strong>)</li>
<li><code>input_molecule</code> (String, <strong>required</strong>)</li>
<li><code>input_strand</code> (String, <strong>required</strong>)</li>
</ul>
<h4 id="optional_1">Optional</h4>
<ul>
<li><code>input_gtf</code> (File?)</li>
</ul>
<h3 id="outputs_1">Outputs</h3>
<ul>
<li><code>check</code> (String)</li>
<li><code>labels</code> (Array[String])</li>
</ul>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../make-qc-reference/" class="btn btn-neutral float-left" title="Make QC Reference"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../rnaseq-standard-fastq/" class="btn btn-neutral float-right" title="RNA-Seq Standard from FastQ">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
      <p>© 2020-Present St. Jude Children's Research Hospital</p>
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/stjudecloud/workflows" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../make-qc-reference/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../rnaseq-standard-fastq/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script>var base_url = '../..';</script>
    <script src="../../js/theme_extra.js" defer></script>
    <script src="../../js/theme.js" defer></script>
      <script src="../../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
